# AMR
_date: 1/28/2024 5:23PM_  
[_Mendeley paper collection_](https://www.mendeley.com/reference-manager/library/collections/0920c7ef-c6a5-49de-92f5-e77c01a8b5da/all-references/)


## Prediction in the lit

### Reviews

- [Antimicrobial Resistance Prediction in PATRIC and RAST](https://www.nature.com/articles/srep27930)
- [Machine Learning for Antimicrobial Resistance Prediction: Current Practice, Limitations, and Clinical Perspective](https://journals.asm.org/doi/full/10.1128/cmr.00179-21)
- [A review of artificial intelligence applications for antimicrobial resistance](https://mednexus.org/doi/full/10.1016/j.bsheal.2020.08.003#sec-2)

### Approaches

- Rule based approach is to check ARGs if available in the genome, based on info already found in databases:
    - Many species not well documented
    - High rate of FN (genes that are not regarded as ARGs but are actually ARGs)


- use of WGS for AMR prediction
    - k-mer methods (but too large and redundant) - [review](https://mednexus.org/doi/full/10.1016/j.bsheal.2020.08.003) - creates a binary matrix of k-mers and then uses ML to predict AMR


#### Prediction of antibiotic resistance in Escherichia coli from large-scale pan-genome data, 2018

This is Moradigaravand et al., 2018  
Their data is available on ENA and can be downloaded + phenotypic resistance in an excel sheet

In this study they presented a comparison between ML approaches and rule-based from CARD, by implementing different ML models

- organisms: _E. coli_
- data: 1,936 _E. coli_ strains
- models: LR, RF, GBDT, DNN
- metrics: precision, recall, accuracy, f1 score
- features: gene presence/absence, population structure info, year of isolation 

overall, 44 models were trained, measured feature importance as the number of times it was used for gradient boosted decision trees

Then added SNPs+INDELs info while retraining GBDT model, improvement for 8/11 drugs

<figure align='center'>
    <img src='image.png' width=80% height=30%>
    <figcaption>performance of GB</figcaption>
</figure>

GBDT ouperforms other models with 91% avg accuracy (for 11/11 drugs)


They started by constructing a pan-genome of _E. coli_ via this pipeline: (aready have this pipeline implemented - directing outputs and automating on large amount of files)

WGS fasta -> Prokka -> GFF annotation -> Roary -> Pan-genome

Performed variant calling using snp-sites from the core genome alignment

pan genome data: https://github.com/DaneshMoradigaravand/PanPred 

_note: compared all 5 models with each other only - including rule based from CARD/ResFinder_

#### Prediction of antimicrobial resistance based on whole-genome sequencing and machine learning, 2024

- organims: _E. coli_
- data: made available on ENA database, accessions + the labels (S/R) in [this excel sheet]() for 11 drugs (1936-ish), taken from another study
- model: LR, SVM, RF and CNN


#### PARMAP: A Pan-Genome-Based Computational Framework for Predicting Antimicrobial Resistance, 2020

- organisms: _Neisseria gonorrhoeae_, _M. tuberculosis_, _E. coli_
- data: 
    - 1597 _N. gonorrhoeae_ strains, taken from [another study (Schubert, 2019)](https://www.nature.com/articles/s41564-018-0309-1) 
    - 1,447 _M. tuberculosis_ strains, taken from PATRIC
    - 1,936 _E. coli_, taken from [another study (Moradigaravand et al., 2018)](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1006258)
- statistical test: Fischer exact test (FET) to identify gene allel features that are significantly associated with AMR
- models: LR, SVC, GDBT, LR
- model evaluation: val set approach (80/20)
- metric: ROC curve, AUC

Pan genome analysis with the use of ML.

- created pangenome out of core + accessory
- feature extractio: gene allel-strain (GS) matrix to show gene presence/absence
- applied PCA for dimensionality reduction
- projected on 2D via UMAP


#### Genome-wide discovery of epistatic loci affecting antibiotic resistance in Neisseria gonorrhoeae using evolutionary couplings, 2019

This is Schubert, 2019 for _N. gonorrhoeae_
[copy](https://sci-hub.se/10.1038/s41564-018-0309-1)

- organisms: _N. gonorrhoeae_
- data: 1597 _N. gonorrhoeae_ strains
    - Minimum inhibitory concentration of penicillin (PEN), tetracycline (TET), cefixime (CFX), and ciprofloxacin (CIPRO) measured in the exploratory (n = 1,102) and confirmatory dataset (n = 495) and their NCBI SRA identifiers.



### Available data

Some databases that might be useful:

| Name      | Description                                                    | Website                                                     |
|-----------|----------------------------------------------------------------|-------------------------------------------------------------|
| PATRIC    | Bacterial genomes with AMR phenotypes and minimum inhibition concentration        | [https://patricbrc.org/](https://patricbrc.org/)           |
| CARD      | ARG and their resistance mechanisms                                               | [https://card.mcmaster.ca/](https://card.mcmaster.ca/)     |
| ARDB      | ARG information                                                                   | [https://ardb.cbcb.umd.edu/](https://ardb.cbcb.umd.edu/)   |
| BacMet    | Antibacterial biocide and metal resistance genes                                  | [http://bacmet.biomedicine.gu.se](http://bacmet.biomedicine.gu.se) |
| ARG-ANNOT | ARG and point mutations                                                           | [http://www.mediterranee-infection.com/article.php?laref=282&titer=arg-annot](http://www.mediterranee-infection.com/article.php?laref=282&titer=arg-annot) |
| ResFinder | ARGs and chromosomal point mutations in total or partial DNA sequence of bacteria | [http://genepi.food.dtu.dk/resfinder](http://genepi.food.dtu.dk/resfinder) |


Works that used data directly from PATRIC:
- [Global pathogenomic analysis identifies known and candidate genetic antimicrobial resistance determinants in twelve species](https://www.nature.com/articles/s41467-023-43549-9#MOESM5)
- [Machine learning and structural analysis of Mycobacterium tuberculosis pan-genome identifies genetic signatures of antibiotic resistance](https://www.nature.com/articles/s41467-018-06634-y)
- [Machine learning with random subspace ensembles identifies antimicrobial resistance determinants from pan-genomes of three pathogens](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007608)
- [A pan-genome-based machine learning approach for predicting antimicrobial resistance activities of the Escherichia coli strains](https://academic.oup.com/bioinformatics/article/34/13/i89/5045729)


Datasets that might be useful for training models:

- CRyPTIC consortium
- _E. coli_ dataset from Moradigaravand et al., 2018
- _N. gonorrhoeae_ dataset from Schubert, 2019 

PATRIC


### Notable lines:

"ML models typically treat genes as independent predictors, with no consideration of structural and functional linkages; they also may not be accurate when new mutational variants of known AMR genes emerge" - [Machine Learning for Antimicrobial Resistance Prediction: Current Practice, Limitations, and Clinical Perspective](https://journals.asm.org/doi/full/10.1128/cmr.00179-21)

"To narrow the scope of this problem, several recent studies have sequenced hundreds of genomes for a given species and extensively characterized the variations in known AMR genes that confer the resistance phenotype. This is a successful approach because in many cases AMR can be the result of a single gene or a small set of variations... the main drawback is that by focusing on a small set of genes with known phenotypes any potentially novel AMR determinants are missed." - [Antimicrobial Resistance Prediction in PATRIC and RAST](https://www.nature.com/articles/srep27930)

"Recent ML-aided GWAS have identified genetic and metabolic mechanisms behind resistance in M. tuberculosis, and demonstrated ML’s competitiveness compared to typical statistical testing at recovering known AMR genes for multiple pathogens" - [Global pathogenomic analysis identifies known and candidate genetic antimicrobial resistance determinants in twelve species](https://www.nature.com/articles/s41467-023-43549-9)

"The absence of a ‘gold-standard’ dataset to compare results from our pipeline to means that we must compare to imperfect standards, such as existing testing methods with lower resolution (PCR) and use synthetic sequencing data to compare targets not covered by PCR in our laboratory" - [An ISO-certified genomics workflow for identification and surveillance of antimicrobial resistance](https://www.nature.com/articles/s41467-022-35713-4)

"This disconnect between statistically identified and mechanistically established genetic determinants of AMR highlights the current gap in knowledge in AMR genetics and remains a challenge for the real world adoption of ML-based systems for rapidly predicting AMR and informing treatment strategies" - [Global pathogenomic analysis identifies known and candidate genetic antimicrobial resistance determinants in twelve species](https://www.nature.com/articles/s41467-023-43549-9)

"Another option potentially leading to a more holistic set
of potential SNPs would be to use an artificial pseudo-pan-genome
incorporating many genomes of a particular species as a reference
within the SNP detection workflow." - [Prediction of antimicrobial resistance based on whole-genome sequencing and machine learning](https://academic.oup.com/bioinformatics/article/38/2/325/6382301)

"Whilst such rule-based models are highly accurate for some common pathogens with well-characterized resistance mechanisms (e.g. _Mycobacterium tuberculosis_ and _Staphylococcus aureus_) [7], they cannot be employed to detect resistance caused by unknown mechanisms in other major pathogenic strains, and require regular curation to remain effective" - [Prediction of antibiotic resistance in Escherichia coli from large-scale pan-genome data](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1006258#sec018)


"Therefore, it is very important to prioritize candidate AMR-associated genes in a population of strains (based on their likelihood of contributing to AMR) in order to identify new factors that are likely to be involved in AMR in N. gonorrhoeae"  
"Recent studies have shown that AMR can be predicted
by using pan-genome information, but the performance
differs greatly in different species" - [PARMAP: A Pan-Genome-Based Computational Framework for Predicting Antimicrobial Resistance](https://www.frontiersin.org/journals/microbiology/articles/10.3389/fmicb.2020.578795/full)


------------

## Ideas

### I- pangenome network based approach for AMR prediction 

- Variant calling pipleine directly out of WGS data (fasta/fastq?) - SNPs as features

Network:
- nodes: genes, pan genome (core + accessory) (look for pan genome construction of of gene clusters, CD-Hit..) but thats a huge network
- links: ppi, automate extracting data from String/BioGRID 

### II- GWAS network based approach to AMR prediction

### III- epistatic interactions network

### IV- multi-omic based approach for AMR prediction



## Terms

- **ARG**: Antimicrobial Resistance Gene
- **AMR**: Antimicrobial Resistance
- **Pan-genome**: the entire gene repertoire of a bacterial species, including core genes (present in all strains) and accessory genes (present in some strains)
- **GWAS**: Genome-wide association study, a study that involves rapidly scanning markers across the complete sets of DNA, or genomes, of many people to find genetic variations associated with a particular trait. Typically used in human genetics, but also in bacteria, used for AMR prediction.
- **TWAS**: Transcriptome-wide association study, a study that involves rapidly scanning markers across the complete sets of RNA, or transcriptomes, of many people to find genetic variations associated with a particular trait. Involves gene expression data.
- **MIC**: Minimum Inhibitory Concentration, an experimental measure of the effectiveness of an antimicrobial drug against a particular microbe (number that can be translated to resistance/susceptibility via EUCAST guidelines - ECOFFs table)
- **FET**: Fischer exact test, a statistical test used to determine if there are nonrandom associations between two categorical variables

<figure align='center'>
    <a href="https://www.medrxiv.org/content/10.1101/2021.02.24.21252386v1.full.pdf">
        <img src="assets/images/ECOFF-CRyPTIC.png" width=60%>
    </a>
    <figcaption>ECOFF values adopted by CRyPTIC</figcaption>
</figure>