{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rayane/capstone-project/src/model/../LOR_calculation.py:128: SyntaxWarning: invalid escape sequence '\\o'\n",
      "  '''\n",
      "/home/rayane/capstone-project/src/model/../LOR_calculation.py:216: SyntaxWarning: invalid escape sequence '\\o'\n",
      "  '''\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '../data/presence_matrices/Campylobacter_coli_presence_absence_T.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[26], line 24\u001b[0m\n\u001b[1;32m     22\u001b[0m sys\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mappend(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m../..\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     23\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mcluster_analysis\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;241m*\u001b[39m\n\u001b[0;32m---> 24\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mLOR_calculation\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;241m*\u001b[39m\n\u001b[1;32m     26\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mwarnings\u001b[39;00m\n",
      "File \u001b[0;32m~/capstone-project/src/model/../LOR_calculation.py:35\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[38;5;66;03m# --\u001b[39;00m\n\u001b[1;32m     34\u001b[0m presence_path\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m../data/presence_matrices/Campylobacter_coli_presence_absence_T.csv\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;66;03m#add the method that generates presence to this script\u001b[39;00m\n\u001b[0;32m---> 35\u001b[0m presence_df\u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpresence_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     36\u001b[0m \u001b[38;5;66;03m# presence_df.index = presence_df.index.astype(str)\u001b[39;00m\n\u001b[1;32m     38\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m gene \u001b[38;5;129;01min\u001b[39;00m presence_df\u001b[38;5;241m.\u001b[39mindex:\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.12/site-packages/pandas/util/_decorators.py:211\u001b[0m, in \u001b[0;36mdeprecate_kwarg.<locals>._deprecate_kwarg.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    209\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    210\u001b[0m         kwargs[new_arg_name] \u001b[38;5;241m=\u001b[39m new_arg_value\n\u001b[0;32m--> 211\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.12/site-packages/pandas/util/_decorators.py:331\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    325\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m>\u001b[39m num_allow_args:\n\u001b[1;32m    326\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    327\u001b[0m         msg\u001b[38;5;241m.\u001b[39mformat(arguments\u001b[38;5;241m=\u001b[39m_format_argument_list(allow_args)),\n\u001b[1;32m    328\u001b[0m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[1;32m    329\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mfind_stack_level(),\n\u001b[1;32m    330\u001b[0m     )\n\u001b[0;32m--> 331\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.12/site-packages/pandas/io/parsers/readers.py:950\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    935\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    936\u001b[0m     dialect,\n\u001b[1;32m    937\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    946\u001b[0m     defaults\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdelimiter\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m\"\u001b[39m},\n\u001b[1;32m    947\u001b[0m )\n\u001b[1;32m    948\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 950\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.12/site-packages/pandas/io/parsers/readers.py:605\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    602\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    604\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 605\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    607\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    608\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1442\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1439\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1441\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1442\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1735\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1733\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1734\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1735\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1736\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1737\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1738\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1739\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1740\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1741\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1742\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1743\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1744\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1745\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1746\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.12/site-packages/pandas/io/common.py:856\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    851\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    852\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    853\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    854\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    855\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 856\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    857\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    858\u001b[0m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    859\u001b[0m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    860\u001b[0m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    861\u001b[0m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    862\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    863\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    864\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    865\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '../data/presence_matrices/Campylobacter_coli_presence_absence_T.csv'"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "\n",
    "\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.linear_model import SGDClassifier\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.utils import resample\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.pipeline import make_pipeline\n",
    "from sklearn.base import clone\n",
    "from sklearn.metrics import roc_auc_score\n",
    "\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from statsmodels.formula.api import ols\n",
    "\n",
    "import sys\n",
    "sys.path.append(\"..\")\n",
    "sys.path.append(\"../..\")\n",
    "from cluster_analysis import *\n",
    "from LOR_calculation import *\n",
    "\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Campylobacter coli"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Cluster 0',\n",
       " 'Cluster 1',\n",
       " 'Cluster 4',\n",
       " 'Cluster 5',\n",
       " 'Cluster 7',\n",
       " 'Cluster 8',\n",
       " 'Cluster 10',\n",
       " 'Cluster 12',\n",
       " 'Cluster 16',\n",
       " 'Cluster 17',\n",
       " 'Cluster 20',\n",
       " 'Cluster 23',\n",
       " 'Cluster 26',\n",
       " 'Cluster 27',\n",
       " 'Cluster 29',\n",
       " 'Cluster 30',\n",
       " 'Cluster 36',\n",
       " 'Cluster 38',\n",
       " 'Cluster 40',\n",
       " 'Cluster 44',\n",
       " 'Cluster 47',\n",
       " 'Cluster 49',\n",
       " 'Cluster 54',\n",
       " 'Cluster 57',\n",
       " 'Cluster 59',\n",
       " 'Cluster 60',\n",
       " 'Cluster 61',\n",
       " 'Cluster 63',\n",
       " 'Cluster 64',\n",
       " 'Cluster 71',\n",
       " 'Cluster 73',\n",
       " 'Cluster 75',\n",
       " 'Cluster 76',\n",
       " 'Cluster 78',\n",
       " 'Cluster 80',\n",
       " 'Cluster 82',\n",
       " 'Cluster 83',\n",
       " 'Cluster 84',\n",
       " 'Cluster 87',\n",
       " 'Cluster 89',\n",
       " 'Cluster 90',\n",
       " 'Cluster 92',\n",
       " 'Cluster 93',\n",
       " 'Cluster 94',\n",
       " 'Cluster 96',\n",
       " 'Cluster 97',\n",
       " 'Cluster 98',\n",
       " 'Cluster 99',\n",
       " 'Cluster 102',\n",
       " 'Cluster 103',\n",
       " 'Cluster 105',\n",
       " 'Cluster 106',\n",
       " 'Cluster 107',\n",
       " 'Cluster 111',\n",
       " 'Cluster 117',\n",
       " 'Cluster 118',\n",
       " 'Cluster 119',\n",
       " 'Cluster 120',\n",
       " 'Cluster 123',\n",
       " 'Cluster 124',\n",
       " 'Cluster 127',\n",
       " 'Cluster 131',\n",
       " 'Cluster 132',\n",
       " 'Cluster 135',\n",
       " 'Cluster 139',\n",
       " 'Cluster 140',\n",
       " 'Cluster 142',\n",
       " 'Cluster 143',\n",
       " 'Cluster 144',\n",
       " 'Cluster 148',\n",
       " 'Cluster 150',\n",
       " 'Cluster 151',\n",
       " 'Cluster 153',\n",
       " 'Cluster 155',\n",
       " 'Cluster 157',\n",
       " 'Cluster 160',\n",
       " 'Cluster 161',\n",
       " 'Cluster 163',\n",
       " 'Cluster 164',\n",
       " 'Cluster 165',\n",
       " 'Cluster 166',\n",
       " 'Cluster 171',\n",
       " 'Cluster 172',\n",
       " 'Cluster 174',\n",
       " 'Cluster 175',\n",
       " 'Cluster 176',\n",
       " 'Cluster 177',\n",
       " 'Cluster 179',\n",
       " 'Cluster 185',\n",
       " 'Cluster 187',\n",
       " 'Cluster 189',\n",
       " 'Cluster 190',\n",
       " 'Cluster 191',\n",
       " 'Cluster 193',\n",
       " 'Cluster 202',\n",
       " 'Cluster 203',\n",
       " 'Cluster 206',\n",
       " 'Cluster 216',\n",
       " 'Cluster 219',\n",
       " 'Cluster 220',\n",
       " 'Cluster 221',\n",
       " 'Cluster 222',\n",
       " 'Cluster 224',\n",
       " 'Cluster 229',\n",
       " 'Cluster 231',\n",
       " 'Cluster 234',\n",
       " 'Cluster 236',\n",
       " 'Cluster 237',\n",
       " 'Cluster 238',\n",
       " 'Cluster 239',\n",
       " 'Cluster 243',\n",
       " 'Cluster 245',\n",
       " 'Cluster 246',\n",
       " 'Cluster 248',\n",
       " 'Cluster 250',\n",
       " 'Cluster 251',\n",
       " 'Cluster 256',\n",
       " 'Cluster 259',\n",
       " 'Cluster 272',\n",
       " 'Cluster 273',\n",
       " 'Cluster 275',\n",
       " 'Cluster 282',\n",
       " 'Cluster 283',\n",
       " 'Cluster 284',\n",
       " 'Cluster 285',\n",
       " 'Cluster 286',\n",
       " 'Cluster 288',\n",
       " 'Cluster 294',\n",
       " 'Cluster 295',\n",
       " 'Cluster 296',\n",
       " 'Cluster 298',\n",
       " 'Cluster 299',\n",
       " 'Cluster 302',\n",
       " 'Cluster 303',\n",
       " 'Cluster 305',\n",
       " 'Cluster 306',\n",
       " 'Cluster 308',\n",
       " 'Cluster 310',\n",
       " 'Cluster 311',\n",
       " 'Cluster 314',\n",
       " 'Cluster 316',\n",
       " 'Cluster 320',\n",
       " 'Cluster 323',\n",
       " 'Cluster 326',\n",
       " 'Cluster 329',\n",
       " 'Cluster 330',\n",
       " 'Cluster 338',\n",
       " 'Cluster 340',\n",
       " 'Cluster 343',\n",
       " 'Cluster 345',\n",
       " 'Cluster 347',\n",
       " 'Cluster 348',\n",
       " 'Cluster 349',\n",
       " 'Cluster 350',\n",
       " 'Cluster 351',\n",
       " 'Cluster 352',\n",
       " 'Cluster 357',\n",
       " 'Cluster 359',\n",
       " 'Cluster 361',\n",
       " 'Cluster 362',\n",
       " 'Cluster 363',\n",
       " 'Cluster 364',\n",
       " 'Cluster 365',\n",
       " 'Cluster 366',\n",
       " 'Cluster 367',\n",
       " 'Cluster 370',\n",
       " 'Cluster 371',\n",
       " 'Cluster 372',\n",
       " 'Cluster 373',\n",
       " 'Cluster 376',\n",
       " 'Cluster 378',\n",
       " 'Cluster 380',\n",
       " 'Cluster 381',\n",
       " 'Cluster 382',\n",
       " 'Cluster 383',\n",
       " 'Cluster 385',\n",
       " 'Cluster 386',\n",
       " 'Cluster 387',\n",
       " 'Cluster 388',\n",
       " 'Cluster 390',\n",
       " 'Cluster 391',\n",
       " 'Cluster 393',\n",
       " 'Cluster 394',\n",
       " 'Cluster 395',\n",
       " 'Cluster 396',\n",
       " 'Cluster 399',\n",
       " 'Cluster 401',\n",
       " 'Cluster 402',\n",
       " 'Cluster 406',\n",
       " 'Cluster 407',\n",
       " 'Cluster 409',\n",
       " 'Cluster 411',\n",
       " 'Cluster 413',\n",
       " 'Cluster 414',\n",
       " 'Cluster 415',\n",
       " 'Cluster 417',\n",
       " 'Cluster 418',\n",
       " 'Cluster 420',\n",
       " 'Cluster 422',\n",
       " 'Cluster 425',\n",
       " 'Cluster 432',\n",
       " 'Cluster 433',\n",
       " 'Cluster 434',\n",
       " 'Cluster 435',\n",
       " 'Cluster 445',\n",
       " 'Cluster 446',\n",
       " 'Cluster 447',\n",
       " 'Cluster 449',\n",
       " 'Cluster 450',\n",
       " 'Cluster 451',\n",
       " 'Cluster 454',\n",
       " 'Cluster 455',\n",
       " 'Cluster 456',\n",
       " 'Cluster 470',\n",
       " 'Cluster 471',\n",
       " 'Cluster 473',\n",
       " 'Cluster 474',\n",
       " 'Cluster 475',\n",
       " 'Cluster 476',\n",
       " 'Cluster 477',\n",
       " 'Cluster 479',\n",
       " 'Cluster 480',\n",
       " 'Cluster 482',\n",
       " 'Cluster 483',\n",
       " 'Cluster 484',\n",
       " 'Cluster 488',\n",
       " 'Cluster 492',\n",
       " 'Cluster 494',\n",
       " 'Cluster 495',\n",
       " 'Cluster 499',\n",
       " 'Cluster 502',\n",
       " 'Cluster 503',\n",
       " 'Cluster 505',\n",
       " 'Cluster 506',\n",
       " 'Cluster 509',\n",
       " 'Cluster 515',\n",
       " 'Cluster 519',\n",
       " 'Cluster 521',\n",
       " 'Cluster 528',\n",
       " 'Cluster 531',\n",
       " 'Cluster 533',\n",
       " 'Cluster 535',\n",
       " 'Cluster 536',\n",
       " 'Cluster 537',\n",
       " 'Cluster 542',\n",
       " 'Cluster 543',\n",
       " 'Cluster 545',\n",
       " 'Cluster 546',\n",
       " 'Cluster 547',\n",
       " 'Cluster 548',\n",
       " 'Cluster 556',\n",
       " 'Cluster 557',\n",
       " 'Cluster 558',\n",
       " 'Cluster 559',\n",
       " 'Cluster 564',\n",
       " 'Cluster 566',\n",
       " 'Cluster 568',\n",
       " 'Cluster 570',\n",
       " 'Cluster 573',\n",
       " 'Cluster 574',\n",
       " 'Cluster 575',\n",
       " 'Cluster 576',\n",
       " 'Cluster 578',\n",
       " 'Cluster 579',\n",
       " 'Cluster 580',\n",
       " 'Cluster 584',\n",
       " 'Cluster 587',\n",
       " 'Cluster 588',\n",
       " 'Cluster 590',\n",
       " 'Cluster 591',\n",
       " 'Cluster 592',\n",
       " 'Cluster 593',\n",
       " 'Cluster 596',\n",
       " 'Cluster 603',\n",
       " 'Cluster 606',\n",
       " 'Cluster 608',\n",
       " 'Cluster 612',\n",
       " 'Cluster 613',\n",
       " 'Cluster 617',\n",
       " 'Cluster 619',\n",
       " 'Cluster 620',\n",
       " 'Cluster 628',\n",
       " 'Cluster 629',\n",
       " 'Cluster 630',\n",
       " 'Cluster 631',\n",
       " 'Cluster 632',\n",
       " 'Cluster 634',\n",
       " 'Cluster 638',\n",
       " 'Cluster 640',\n",
       " 'Cluster 641',\n",
       " 'Cluster 642',\n",
       " 'Cluster 648',\n",
       " 'Cluster 652',\n",
       " 'Cluster 653',\n",
       " 'Cluster 654',\n",
       " 'Cluster 656',\n",
       " 'Cluster 657',\n",
       " 'Cluster 664',\n",
       " 'Cluster 665',\n",
       " 'Cluster 668',\n",
       " 'Cluster 671',\n",
       " 'Cluster 672',\n",
       " 'Cluster 675',\n",
       " 'Cluster 678',\n",
       " 'Cluster 679',\n",
       " 'Cluster 683',\n",
       " 'Cluster 684',\n",
       " 'Cluster 688',\n",
       " 'Cluster 691',\n",
       " 'Cluster 695',\n",
       " 'Cluster 696',\n",
       " 'Cluster 697',\n",
       " 'Cluster 699',\n",
       " 'Cluster 700',\n",
       " 'Cluster 704',\n",
       " 'Cluster 705',\n",
       " 'Cluster 713',\n",
       " 'Cluster 714',\n",
       " 'Cluster 716',\n",
       " 'Cluster 717',\n",
       " 'Cluster 718',\n",
       " 'Cluster 719',\n",
       " 'Cluster 721',\n",
       " 'Cluster 722',\n",
       " 'Cluster 723',\n",
       " 'Cluster 726',\n",
       " 'Cluster 728',\n",
       " 'Cluster 729',\n",
       " 'Cluster 733',\n",
       " 'Cluster 735',\n",
       " 'Cluster 736',\n",
       " 'Cluster 740',\n",
       " 'Cluster 741',\n",
       " 'Cluster 742',\n",
       " 'Cluster 749',\n",
       " 'Cluster 750',\n",
       " 'Cluster 751',\n",
       " 'Cluster 754',\n",
       " 'Cluster 755',\n",
       " 'Cluster 756',\n",
       " 'Cluster 757',\n",
       " 'Cluster 762',\n",
       " 'Cluster 763',\n",
       " 'Cluster 764',\n",
       " 'Cluster 766',\n",
       " 'Cluster 768',\n",
       " 'Cluster 769',\n",
       " 'Cluster 770',\n",
       " 'Cluster 772',\n",
       " 'Cluster 775',\n",
       " 'Cluster 776',\n",
       " 'Cluster 777',\n",
       " 'Cluster 780',\n",
       " 'Cluster 781',\n",
       " 'Cluster 782',\n",
       " 'Cluster 783',\n",
       " 'Cluster 787',\n",
       " 'Cluster 788',\n",
       " 'Cluster 789',\n",
       " 'Cluster 790',\n",
       " 'Cluster 791',\n",
       " 'Cluster 792',\n",
       " 'Cluster 793',\n",
       " 'Cluster 795',\n",
       " 'Cluster 796',\n",
       " 'Cluster 797',\n",
       " 'Cluster 801',\n",
       " 'Cluster 802',\n",
       " 'Cluster 804',\n",
       " 'Cluster 808',\n",
       " 'Cluster 809',\n",
       " 'Cluster 815',\n",
       " 'Cluster 816',\n",
       " 'Cluster 817',\n",
       " 'Cluster 818',\n",
       " 'Cluster 820',\n",
       " 'Cluster 822',\n",
       " 'Cluster 823',\n",
       " 'Cluster 824',\n",
       " 'Cluster 826',\n",
       " 'Cluster 828',\n",
       " 'Cluster 829',\n",
       " 'Cluster 832',\n",
       " 'Cluster 833',\n",
       " 'Cluster 838',\n",
       " 'Cluster 839',\n",
       " 'Cluster 840',\n",
       " 'Cluster 841',\n",
       " 'Cluster 843',\n",
       " 'Cluster 844',\n",
       " 'Cluster 845',\n",
       " 'Cluster 846',\n",
       " 'Cluster 847',\n",
       " 'Cluster 848',\n",
       " 'Cluster 849',\n",
       " 'Cluster 850',\n",
       " 'Cluster 851',\n",
       " 'Cluster 852',\n",
       " 'Cluster 853',\n",
       " 'Cluster 854',\n",
       " 'Cluster 856',\n",
       " 'Cluster 860',\n",
       " 'Cluster 861',\n",
       " 'Cluster 862',\n",
       " 'Cluster 863',\n",
       " 'Cluster 864',\n",
       " 'Cluster 865',\n",
       " 'Cluster 868',\n",
       " 'Cluster 870',\n",
       " 'Cluster 871',\n",
       " 'Cluster 872',\n",
       " 'Cluster 876',\n",
       " 'Cluster 877',\n",
       " 'Cluster 878',\n",
       " 'Cluster 880',\n",
       " 'Cluster 883',\n",
       " 'Cluster 885',\n",
       " 'Cluster 886',\n",
       " 'Cluster 892',\n",
       " 'Cluster 893',\n",
       " 'Cluster 894',\n",
       " 'Cluster 899',\n",
       " 'Cluster 900',\n",
       " 'Cluster 901',\n",
       " 'Cluster 902',\n",
       " 'Cluster 903',\n",
       " 'Cluster 907',\n",
       " 'Cluster 908',\n",
       " 'Cluster 909',\n",
       " 'Cluster 914',\n",
       " 'Cluster 915',\n",
       " 'Cluster 916',\n",
       " 'Cluster 924',\n",
       " 'Cluster 925',\n",
       " 'Cluster 929',\n",
       " 'Cluster 933',\n",
       " 'Cluster 934',\n",
       " 'Cluster 935',\n",
       " 'Cluster 937',\n",
       " 'Cluster 940',\n",
       " 'Cluster 943',\n",
       " 'Cluster 944',\n",
       " 'Cluster 949',\n",
       " 'Cluster 952',\n",
       " 'Cluster 955',\n",
       " 'Cluster 956',\n",
       " 'Cluster 957',\n",
       " 'Cluster 962',\n",
       " 'Cluster 963',\n",
       " 'Cluster 964',\n",
       " 'Cluster 965',\n",
       " 'Cluster 966',\n",
       " 'Cluster 971',\n",
       " 'Cluster 972',\n",
       " 'Cluster 974',\n",
       " 'Cluster 975',\n",
       " 'Cluster 980',\n",
       " 'Cluster 981',\n",
       " 'Cluster 982',\n",
       " 'Cluster 984',\n",
       " 'Cluster 985',\n",
       " 'Cluster 995',\n",
       " 'Cluster 996',\n",
       " 'Cluster 997',\n",
       " 'Cluster 998',\n",
       " 'Cluster 999',\n",
       " 'Cluster 1000',\n",
       " 'Cluster 1001',\n",
       " 'Cluster 1002',\n",
       " 'Cluster 1003',\n",
       " 'Cluster 1004',\n",
       " 'Cluster 1006',\n",
       " 'Cluster 1009',\n",
       " 'Cluster 1011',\n",
       " 'Cluster 1017',\n",
       " 'Cluster 1018',\n",
       " 'Cluster 1019',\n",
       " 'Cluster 1020',\n",
       " 'Cluster 1022',\n",
       " 'Cluster 1023',\n",
       " 'Cluster 1024',\n",
       " 'Cluster 1025',\n",
       " 'Cluster 1027',\n",
       " 'Cluster 1031',\n",
       " 'Cluster 1032',\n",
       " 'Cluster 1033',\n",
       " 'Cluster 1035',\n",
       " 'Cluster 1036',\n",
       " 'Cluster 1037',\n",
       " 'Cluster 1043',\n",
       " 'Cluster 1044',\n",
       " 'Cluster 1046',\n",
       " 'Cluster 1047',\n",
       " 'Cluster 1050',\n",
       " 'Cluster 1051',\n",
       " 'Cluster 1055',\n",
       " 'Cluster 1056',\n",
       " 'Cluster 1057',\n",
       " 'Cluster 1066',\n",
       " 'Cluster 1067',\n",
       " 'Cluster 1068',\n",
       " 'Cluster 1075',\n",
       " 'Cluster 1076',\n",
       " 'Cluster 1077',\n",
       " 'Cluster 1078',\n",
       " 'Cluster 1079',\n",
       " 'Cluster 1082',\n",
       " 'Cluster 1087',\n",
       " 'Cluster 1088',\n",
       " 'Cluster 1089',\n",
       " 'Cluster 1090',\n",
       " 'Cluster 1096',\n",
       " 'Cluster 1097',\n",
       " 'Cluster 1098',\n",
       " 'Cluster 1101',\n",
       " 'Cluster 1103',\n",
       " 'Cluster 1104',\n",
       " 'Cluster 1106',\n",
       " 'Cluster 1107',\n",
       " 'Cluster 1110',\n",
       " 'Cluster 1113',\n",
       " 'Cluster 1114',\n",
       " 'Cluster 1115',\n",
       " 'Cluster 1116',\n",
       " 'Cluster 1117',\n",
       " 'Cluster 1119',\n",
       " 'Cluster 1125',\n",
       " 'Cluster 1126',\n",
       " 'Cluster 1127',\n",
       " 'Cluster 1128',\n",
       " 'Cluster 1131',\n",
       " 'Cluster 1132',\n",
       " 'Cluster 1133',\n",
       " 'Cluster 1135',\n",
       " 'Cluster 1141',\n",
       " 'Cluster 1142',\n",
       " 'Cluster 1143',\n",
       " 'Cluster 1145',\n",
       " 'Cluster 1146',\n",
       " 'Cluster 1147',\n",
       " 'Cluster 1148',\n",
       " 'Cluster 1153',\n",
       " 'Cluster 1157',\n",
       " 'Cluster 1158',\n",
       " 'Cluster 1159',\n",
       " 'Cluster 1160',\n",
       " 'Cluster 1161',\n",
       " 'Cluster 1162',\n",
       " 'Cluster 1163',\n",
       " 'Cluster 1168',\n",
       " 'Cluster 1171',\n",
       " 'Cluster 1173',\n",
       " 'Cluster 1174',\n",
       " 'Cluster 1180',\n",
       " 'Cluster 1181',\n",
       " 'Cluster 1182',\n",
       " 'Cluster 1185',\n",
       " 'Cluster 1186',\n",
       " 'Cluster 1187',\n",
       " 'Cluster 1188',\n",
       " 'Cluster 1189',\n",
       " 'Cluster 1191',\n",
       " 'Cluster 1197',\n",
       " 'Cluster 1200',\n",
       " 'Cluster 1201',\n",
       " 'Cluster 1205',\n",
       " 'Cluster 1211',\n",
       " 'Cluster 1212',\n",
       " 'Cluster 1213',\n",
       " 'Cluster 1214',\n",
       " 'Cluster 1215',\n",
       " 'Cluster 1216',\n",
       " 'Cluster 1217',\n",
       " 'Cluster 1219',\n",
       " 'Cluster 1222',\n",
       " 'Cluster 1223',\n",
       " 'Cluster 1224',\n",
       " 'Cluster 1225',\n",
       " 'Cluster 1228',\n",
       " 'Cluster 1229',\n",
       " 'Cluster 1232',\n",
       " 'Cluster 1233',\n",
       " 'Cluster 1234',\n",
       " 'Cluster 1235',\n",
       " 'Cluster 1236',\n",
       " 'Cluster 1238',\n",
       " 'Cluster 1242',\n",
       " 'Cluster 1243',\n",
       " 'Cluster 1244',\n",
       " 'Cluster 1245',\n",
       " 'Cluster 1246',\n",
       " 'Cluster 1249',\n",
       " 'Cluster 1250',\n",
       " 'Cluster 1251',\n",
       " 'Cluster 1254',\n",
       " 'Cluster 1255',\n",
       " 'Cluster 1257',\n",
       " 'Cluster 1260',\n",
       " 'Cluster 1262',\n",
       " 'Cluster 1263',\n",
       " 'Cluster 1264',\n",
       " 'Cluster 1267',\n",
       " 'Cluster 1270',\n",
       " 'Cluster 1274',\n",
       " 'Cluster 1283',\n",
       " 'Cluster 1284',\n",
       " 'Cluster 1285',\n",
       " 'Cluster 1288',\n",
       " 'Cluster 1290',\n",
       " 'Cluster 1291',\n",
       " 'Cluster 1293',\n",
       " 'Cluster 1294',\n",
       " 'Cluster 1295',\n",
       " 'Cluster 1298',\n",
       " 'Cluster 1299',\n",
       " 'Cluster 1302',\n",
       " 'Cluster 1303',\n",
       " 'Cluster 1307',\n",
       " 'Cluster 1308',\n",
       " 'Cluster 1310',\n",
       " 'Cluster 1311',\n",
       " 'Cluster 1312',\n",
       " 'Cluster 1313',\n",
       " 'Cluster 1314',\n",
       " 'Cluster 1315',\n",
       " 'Cluster 1317',\n",
       " 'Cluster 1318',\n",
       " 'Cluster 1319',\n",
       " 'Cluster 1321',\n",
       " 'Cluster 1322',\n",
       " 'Cluster 1325',\n",
       " 'Cluster 1326',\n",
       " 'Cluster 1327',\n",
       " 'Cluster 1328',\n",
       " 'Cluster 1337',\n",
       " 'Cluster 1338',\n",
       " 'Cluster 1339',\n",
       " 'Cluster 1346',\n",
       " 'Cluster 1350',\n",
       " 'Cluster 1353',\n",
       " 'Cluster 1356',\n",
       " 'Cluster 1357',\n",
       " 'Cluster 1358',\n",
       " 'Cluster 1359',\n",
       " 'Cluster 1364',\n",
       " 'Cluster 1365',\n",
       " 'Cluster 1366',\n",
       " 'Cluster 1367',\n",
       " 'Cluster 1368',\n",
       " 'Cluster 1371',\n",
       " 'Cluster 1374',\n",
       " 'Cluster 1375',\n",
       " 'Cluster 1376',\n",
       " 'Cluster 1377',\n",
       " 'Cluster 1378',\n",
       " 'Cluster 1379',\n",
       " 'Cluster 1380',\n",
       " 'Cluster 1381',\n",
       " 'Cluster 1383',\n",
       " 'Cluster 1384',\n",
       " 'Cluster 1386',\n",
       " 'Cluster 1387',\n",
       " 'Cluster 1388',\n",
       " 'Cluster 1390',\n",
       " 'Cluster 1392',\n",
       " 'Cluster 1395',\n",
       " 'Cluster 1396',\n",
       " 'Cluster 1397',\n",
       " 'Cluster 1398',\n",
       " 'Cluster 1399',\n",
       " 'Cluster 1400',\n",
       " 'Cluster 1407',\n",
       " 'Cluster 1408',\n",
       " 'Cluster 1409',\n",
       " 'Cluster 1410',\n",
       " 'Cluster 1412',\n",
       " 'Cluster 1413',\n",
       " 'Cluster 1414',\n",
       " 'Cluster 1415',\n",
       " 'Cluster 1418',\n",
       " 'Cluster 1419',\n",
       " 'Cluster 1424',\n",
       " 'Cluster 1425',\n",
       " 'Cluster 1426',\n",
       " 'Cluster 1427',\n",
       " 'Cluster 1428',\n",
       " 'Cluster 1434',\n",
       " 'Cluster 1435',\n",
       " 'Cluster 1436',\n",
       " 'Cluster 1437',\n",
       " 'Cluster 1443',\n",
       " 'Cluster 1446',\n",
       " 'Cluster 1447',\n",
       " 'Cluster 1448',\n",
       " 'Cluster 1454',\n",
       " 'Cluster 1455',\n",
       " 'Cluster 1460',\n",
       " 'Cluster 1461',\n",
       " 'Cluster 1462',\n",
       " 'Cluster 1466',\n",
       " 'Cluster 1467',\n",
       " 'Cluster 1468',\n",
       " 'Cluster 1474',\n",
       " 'Cluster 1475',\n",
       " 'Cluster 1476',\n",
       " 'Cluster 1477',\n",
       " 'Cluster 1483',\n",
       " 'Cluster 1485',\n",
       " 'Cluster 1488',\n",
       " 'Cluster 1496',\n",
       " 'Cluster 1497',\n",
       " 'Cluster 1498',\n",
       " 'Cluster 1499',\n",
       " 'Cluster 1503',\n",
       " 'Cluster 1504',\n",
       " 'Cluster 1506',\n",
       " 'Cluster 1509',\n",
       " 'Cluster 1510',\n",
       " 'Cluster 1511',\n",
       " 'Cluster 1514',\n",
       " 'Cluster 1515',\n",
       " 'Cluster 1519',\n",
       " 'Cluster 1520',\n",
       " 'Cluster 1524',\n",
       " 'Cluster 1526',\n",
       " 'Cluster 1527',\n",
       " 'Cluster 1528',\n",
       " 'Cluster 1529',\n",
       " 'Cluster 1531',\n",
       " 'Cluster 1532',\n",
       " 'Cluster 1533',\n",
       " 'Cluster 1536',\n",
       " 'Cluster 1538',\n",
       " 'Cluster 1539',\n",
       " 'Cluster 1540',\n",
       " 'Cluster 1541',\n",
       " 'Cluster 1542',\n",
       " 'Cluster 1543',\n",
       " 'Cluster 1545',\n",
       " 'Cluster 1550',\n",
       " 'Cluster 1551',\n",
       " 'Cluster 1552',\n",
       " 'Cluster 1553',\n",
       " 'Cluster 1554',\n",
       " 'Cluster 1555',\n",
       " 'Cluster 1558',\n",
       " 'Cluster 1559',\n",
       " 'Cluster 1560',\n",
       " 'Cluster 1563',\n",
       " 'Cluster 1564',\n",
       " 'Cluster 1565',\n",
       " 'Cluster 1566',\n",
       " 'Cluster 1567',\n",
       " 'Cluster 1569',\n",
       " 'Cluster 1572',\n",
       " 'Cluster 1574',\n",
       " 'Cluster 1575',\n",
       " 'Cluster 1576',\n",
       " 'Cluster 1577',\n",
       " 'Cluster 1580',\n",
       " 'Cluster 1581',\n",
       " 'Cluster 1587',\n",
       " 'Cluster 1588',\n",
       " 'Cluster 1590',\n",
       " 'Cluster 1595',\n",
       " 'Cluster 1596',\n",
       " 'Cluster 1597',\n",
       " 'Cluster 1601',\n",
       " 'Cluster 1602',\n",
       " 'Cluster 1603',\n",
       " 'Cluster 1608',\n",
       " 'Cluster 1616',\n",
       " 'Cluster 1618',\n",
       " 'Cluster 1619',\n",
       " 'Cluster 1620',\n",
       " 'Cluster 1624',\n",
       " 'Cluster 1625',\n",
       " 'Cluster 1630',\n",
       " 'Cluster 1631',\n",
       " 'Cluster 1632',\n",
       " 'Cluster 1638',\n",
       " 'Cluster 1639',\n",
       " 'Cluster 1640',\n",
       " 'Cluster 1641',\n",
       " 'Cluster 1642',\n",
       " 'Cluster 1646',\n",
       " 'Cluster 1648',\n",
       " 'Cluster 1651',\n",
       " 'Cluster 1652',\n",
       " 'Cluster 1653',\n",
       " 'Cluster 1654',\n",
       " 'Cluster 1658',\n",
       " 'Cluster 1659',\n",
       " 'Cluster 1660',\n",
       " 'Cluster 1661',\n",
       " 'Cluster 1668',\n",
       " 'Cluster 1669',\n",
       " 'Cluster 1670',\n",
       " 'Cluster 1671',\n",
       " 'Cluster 1672',\n",
       " 'Cluster 1673',\n",
       " 'Cluster 1674',\n",
       " 'Cluster 1675',\n",
       " 'Cluster 1678',\n",
       " 'Cluster 1679',\n",
       " 'Cluster 1680',\n",
       " 'Cluster 1681',\n",
       " 'Cluster 1684',\n",
       " 'Cluster 1686',\n",
       " 'Cluster 1687',\n",
       " 'Cluster 1688',\n",
       " 'Cluster 1692',\n",
       " 'Cluster 1695',\n",
       " 'Cluster 1696',\n",
       " 'Cluster 1697',\n",
       " 'Cluster 1698',\n",
       " 'Cluster 1699',\n",
       " 'Cluster 1700',\n",
       " 'Cluster 1701',\n",
       " 'Cluster 1706',\n",
       " 'Cluster 1710',\n",
       " 'Cluster 1711',\n",
       " 'Cluster 1713',\n",
       " 'Cluster 1719',\n",
       " 'Cluster 1727',\n",
       " 'Cluster 1728',\n",
       " 'Cluster 1729',\n",
       " 'Cluster 1730',\n",
       " 'Cluster 1731',\n",
       " 'Cluster 1733',\n",
       " 'Cluster 1739',\n",
       " 'Cluster 1740',\n",
       " 'Cluster 1742',\n",
       " 'Cluster 1750',\n",
       " 'Cluster 1751',\n",
       " 'Cluster 1752',\n",
       " 'Cluster 1753',\n",
       " 'Cluster 1755',\n",
       " 'Cluster 1757',\n",
       " 'Cluster 1761',\n",
       " 'Cluster 1762',\n",
       " 'Cluster 1763',\n",
       " 'Cluster 1767',\n",
       " 'Cluster 1770',\n",
       " 'Cluster 1771',\n",
       " 'Cluster 1772',\n",
       " 'Cluster 1773',\n",
       " 'Cluster 1774',\n",
       " 'Cluster 1776',\n",
       " 'Cluster 1777',\n",
       " 'Cluster 1778',\n",
       " 'Cluster 1782',\n",
       " 'Cluster 1783',\n",
       " 'Cluster 1784',\n",
       " 'Cluster 1785',\n",
       " 'Cluster 1792',\n",
       " 'Cluster 1793',\n",
       " 'Cluster 1794',\n",
       " 'Cluster 1795',\n",
       " 'Cluster 1796',\n",
       " 'Cluster 1801',\n",
       " 'Cluster 1808',\n",
       " 'Cluster 1809',\n",
       " 'Cluster 1810',\n",
       " 'Cluster 1813',\n",
       " 'Cluster 1817',\n",
       " 'Cluster 1818',\n",
       " 'Cluster 1819',\n",
       " 'Cluster 1820',\n",
       " 'Cluster 1822',\n",
       " 'Cluster 1825',\n",
       " 'Cluster 1829',\n",
       " 'Cluster 1830',\n",
       " 'Cluster 1831',\n",
       " 'Cluster 1834',\n",
       " 'Cluster 1835',\n",
       " 'Cluster 1836',\n",
       " 'Cluster 1838',\n",
       " 'Cluster 1839',\n",
       " 'Cluster 1840',\n",
       " 'Cluster 1841',\n",
       " 'Cluster 1847',\n",
       " 'Cluster 1850',\n",
       " 'Cluster 1851',\n",
       " 'Cluster 1856',\n",
       " 'Cluster 1857',\n",
       " 'Cluster 1858',\n",
       " 'Cluster 1863',\n",
       " 'Cluster 1864',\n",
       " 'Cluster 1869',\n",
       " 'Cluster 1874',\n",
       " 'Cluster 1875',\n",
       " 'Cluster 1884',\n",
       " 'Cluster 1891',\n",
       " 'Cluster 1892',\n",
       " 'Cluster 1893',\n",
       " 'Cluster 1904',\n",
       " 'Cluster 1905',\n",
       " 'Cluster 1906',\n",
       " 'Cluster 1907',\n",
       " 'Cluster 1908',\n",
       " 'Cluster 1917',\n",
       " 'Cluster 1918',\n",
       " 'Cluster 1919',\n",
       " 'Cluster 1920',\n",
       " 'Cluster 1924',\n",
       " 'Cluster 1925',\n",
       " 'Cluster 1926',\n",
       " 'Cluster 1927',\n",
       " 'Cluster 1928',\n",
       " 'Cluster 1929',\n",
       " 'Cluster 1932',\n",
       " 'Cluster 1939',\n",
       " 'Cluster 1940',\n",
       " 'Cluster 1941',\n",
       " 'Cluster 1942',\n",
       " 'Cluster 1948',\n",
       " 'Cluster 1949',\n",
       " 'Cluster 1950',\n",
       " 'Cluster 1951',\n",
       " 'Cluster 1954',\n",
       " 'Cluster 1956',\n",
       " 'Cluster 1957',\n",
       " 'Cluster 1960',\n",
       " 'Cluster 1961',\n",
       " 'Cluster 1967',\n",
       " 'Cluster 1968',\n",
       " 'Cluster 1973',\n",
       " 'Cluster 1977',\n",
       " 'Cluster 1978',\n",
       " 'Cluster 1979',\n",
       " 'Cluster 1980',\n",
       " 'Cluster 1981',\n",
       " 'Cluster 1982',\n",
       " 'Cluster 1986',\n",
       " 'Cluster 1987',\n",
       " 'Cluster 1988',\n",
       " 'Cluster 1989',\n",
       " 'Cluster 1994',\n",
       " 'Cluster 1996',\n",
       " 'Cluster 1997',\n",
       " 'Cluster 1998',\n",
       " 'Cluster 2000',\n",
       " 'Cluster 2008',\n",
       " 'Cluster 2009',\n",
       " 'Cluster 2010',\n",
       " 'Cluster 2011',\n",
       " 'Cluster 2012',\n",
       " 'Cluster 2023',\n",
       " 'Cluster 2025',\n",
       " 'Cluster 2029',\n",
       " 'Cluster 2030',\n",
       " 'Cluster 2031',\n",
       " 'Cluster 2038',\n",
       " 'Cluster 2039',\n",
       " 'Cluster 2040',\n",
       " 'Cluster 2049',\n",
       " 'Cluster 2053',\n",
       " 'Cluster 2054',\n",
       " 'Cluster 2055',\n",
       " 'Cluster 2056',\n",
       " 'Cluster 2058',\n",
       " 'Cluster 2059',\n",
       " 'Cluster 2060',\n",
       " 'Cluster 2062',\n",
       " 'Cluster 2067',\n",
       " 'Cluster 2068',\n",
       " 'Cluster 2069',\n",
       " 'Cluster 2074',\n",
       " 'Cluster 2075',\n",
       " 'Cluster 2076',\n",
       " 'Cluster 2077',\n",
       " 'Cluster 2079',\n",
       " 'Cluster 2080',\n",
       " 'Cluster 2081',\n",
       " 'Cluster 2082',\n",
       " 'Cluster 2085',\n",
       " 'Cluster 2086',\n",
       " 'Cluster 2087',\n",
       " 'Cluster 2088',\n",
       " 'Cluster 2089',\n",
       " 'Cluster 2094',\n",
       " 'Cluster 2095',\n",
       " 'Cluster 2102',\n",
       " 'Cluster 2104',\n",
       " 'Cluster 2105',\n",
       " 'Cluster 2107',\n",
       " 'Cluster 2109',\n",
       " 'Cluster 2113',\n",
       " 'Cluster 2114',\n",
       " 'Cluster 2115',\n",
       " 'Cluster 2116',\n",
       " 'Cluster 2117',\n",
       " 'Cluster 2118',\n",
       " 'Cluster 2120',\n",
       " 'Cluster 2131',\n",
       " 'Cluster 2132',\n",
       " 'Cluster 2133',\n",
       " 'Cluster 2134',\n",
       " 'Cluster 2135',\n",
       " ...]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pan_df=get_cluster_pan_gene_class('../../pangenome-repo/Pangenome-Analysis-Workflow/codes/Campylobacter_coli/Campylobacter_coli_pangenome.csv')\n",
    "\n",
    "pan_df = pan_df[pan_df['pan_gene_class']!='unique']\n",
    "non_unique_genes = pan_df.index.tolist()\n",
    "non_unique_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "species='Campylobacter_coli'; drug='ciprofloxacin'\n",
    "\n",
    "X_df = pd.read_csv('../../data/presence_matrices/Campylobacter_coli_presence_absence.csv', index_col=0)\n",
    "y_df= pd.read_csv('../../data/phenotypes/Campylobacter_coli_ciprofloxacin.csv', index_col=0)\n",
    "\n",
    "for gene in X_df.columns:\n",
    "    \n",
    "    if X_df[gene].std() == 0:\n",
    "        X_df.drop(gene, axis=1, inplace=True)\n",
    "\n",
    "X_df = X_df.sort_index()\n",
    "y_df = y_df.sort_index()\n",
    "\n",
    "y_indices=list(y_df.index)\n",
    "X_indices=list(X_df.index)\n",
    "\n",
    "intersection = [i for i in y_indices if i in X_indices]\n",
    "y_df = y_df.loc[intersection]\n",
    "X_df = X_df.loc[intersection]\n",
    "\n",
    "X = X_df.values\n",
    "y = y_df.values\n",
    "\n",
    "labeled_matrix = pd.concat([X_df, y_df], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Acinetobacter baumannii"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "import re\n",
    "import seaborn as sns\n",
    "\n",
    "genome_ids_file=\"../../pangenome-repo/Pangenome-Analysis-Workflow/test/A/Acinetobacter_baumannii_genome_ids.csv\"\n",
    "clstr_freq_file=\"../../pangenome-repo/Pangenome-Analysis-Workflow/codes/Acinetobacter_baumannii/Acinetobacter_baumannii_cluster_frequencies.csv\"\n",
    "clstr_file=\"../../pangenome-repo/Pangenome-Analysis-Workflow/codes/Acinetobacter_baumannii/Acinetobacter_baumannii.fasta.clstr\"\n",
    "clstr_fasta_file=\"../../pangenome-repo/Pangenome-Analysis-Workflow/codes/Acinetobacter_baumannii/Acinetobacter_baumannii.fasta\"\n",
    "\n",
    "samples_df = pd.read_csv(genome_ids_file, dtype=str)\n",
    "samples_list=samples_df['genome.genome_ids'].tolist()\n",
    "\n",
    "clstr_freq_df=pd.read_csv(clstr_freq_file, index_col=1)\n",
    "clstr_freq_df=clstr_freq_df.drop(clstr_freq_df.columns[0], axis=1)\n",
    "clstr_list=clstr_freq_df.index.tolist()\n",
    "\n",
    "df = pd.DataFrame(index=samples_list, columns=clstr_list)\n",
    "df = df.fillna(0)\n",
    "\n",
    "with open(clstr_file) as f:\n",
    "    for line in f:\n",
    "        if line.startswith(\">Cluster\"):\n",
    "            #if its a > line get the cluster id and start counting its occurences in which samples by saving it to the local var\n",
    "            cluster_id=line[1:].strip()\n",
    "        else:\n",
    "            #else its a line designating the sample genome, one that the last cluster_id is present in\n",
    "            # get the 1st group matching here \\d\\t\\d+aa, >fig\\|([\\d\\.\\d]+).+\n",
    "\n",
    "            genome_id=re.match(r\"\\d+\\t\\d+aa, >fig\\|([\\d\\.\\d]+).+\", line).group(1)\n",
    "            genome_id=genome_id[:-1]  #removing the trailing .\n",
    "            df.loc[genome_id, cluster_id]+=1\n",
    "\n",
    "df = df.T\n",
    "df.to_csv('../../data/presence_matrices/Acinetobacter_baumannii_GxS.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "pan_df=get_cluster_pan_gene_class('../../pangenome-repo/Pangenome-Analysis-Workflow/codes/Acinetobacter_baumannii/Acinetobacter_baumannii_pangenome.csv')\n",
    "\n",
    "pan_df = pan_df[pan_df['pan_gene_class']!='unique']\n",
    "non_unique_genes = pan_df.index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "species='Acinetobacter_baumannii'; drug='amikacin'\n",
    "\n",
    "X_df = pd.read_csv('../../data/presence_matrices/Acinetobacter_baumannii_GxS.csv', index_col=0); X_df=X_df.T\n",
    "X_df.index = X_df.index.astype('float')\n",
    "# X_df.index = X_df.index.astype('object')\n",
    "\n",
    "# -- remove from X_df all genes that are not in the non_unique_genes\n",
    "X_df = X_df[non_unique_genes]\n",
    "\n",
    "y_df= pd.read_csv('../../data/phenotypes/Acinetobacter_baumannii_amikacin.csv', index_col=0)\n",
    "y_df.index = y_df.index.astype('float')\n",
    "\n",
    "for gene in X_df.columns:\n",
    "    \n",
    "    if X_df[gene].std() == 0:\n",
    "        X_df.drop(gene, axis=1, inplace=True)\n",
    "\n",
    "X_df = X_df.sort_index()\n",
    "y_df = y_df.sort_index()\n",
    "\n",
    "\n",
    "y_indices=list(y_df.index)\n",
    "X_indices=list(X_df.index)\n",
    "\n",
    "intersection = [i for i in y_indices if i in X_indices]\n",
    "y_df = y_df.loc[intersection]\n",
    "X_df = X_df.loc[intersection]\n",
    "\n",
    "# -- removing duplicate rows? --          to check th e nature of duplicates it could be an error through type conversion\n",
    "X_df = X_df.loc[~X_df.index.duplicated(keep='first')]\n",
    "\n",
    "X = X_df.values\n",
    "y = y_df.values\n",
    "\n",
    "labeled_matrix = pd.concat([X_df, y_df], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(937, 5467)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SVM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_models = 200\n",
    "t = 0.4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "E: 398210, V: 5311 when 200 models are used, with corr threshold=0.4; species=Campylobacter_coli, drug=ciprofloxacin.\n"
     ]
    }
   ],
   "source": [
    "models=[]\n",
    "for i in range(n_models):\n",
    "    X_boot, y_boot = resample(X, y, random_state=i)\n",
    "    model=SGDClassifier(loss='hinge', penalty= None, max_iter=1000, tol=1e-3)\n",
    "    model.fit(X_boot, y_boot.ravel())\n",
    "    models.append(model)\n",
    "\n",
    "weights=np.zeros((n_models, X.shape[1]))\n",
    "for i in range(n_models):\n",
    "    weights[i]=models[i].coef_\n",
    "weights_df = pd.DataFrame(weights, columns=X_df.columns)\n",
    "\n",
    "corr_SVM = weights_df.corr()\n",
    "\n",
    "G=nx.Graph()\n",
    "\n",
    "for i in range(X.shape[1]):\n",
    "    for j in range(i+1, X.shape[1]):\n",
    "        node_i=X_df.columns[i]; node_j=X_df.columns[j]\n",
    "        correlation = corr_SVM.iloc[i, j]\n",
    "        if correlation > t or corr_SVM.iloc[j, i] < -t:\n",
    "            G.add_edge(node_i, node_j, weight=correlation)\n",
    "\n",
    "# G=set_cluster_attributes(G)\n",
    "\n",
    "nx.write_graphml(G, f'../../data/graphs-6/{n_models}_randomized_SVM_{t}signed_corr_{species}_{drug}.graphml')\n",
    "\n",
    "edges_list = list(G.edges)\n",
    "print(f\"E: {len(edges_list)}, V: {G.number_of_nodes()} when {n_models} models are used, with corr threshold={t}; species={species}, drug={drug}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ANOVA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_edges=[]\n",
    "\n",
    "edges_list_2=[(i.replace(' ', '_'), j.replace(' ', '_')) for i, j in edges_list]\n",
    "labeled_matrix_2 = labeled_matrix.rename(columns={col: col.replace(' ', '_') for col in labeled_matrix.columns})\n",
    "p_values={}\n",
    "\n",
    "count_nan=0\n",
    "save_nan_testing=[]\n",
    "\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(\"ignore\")\n",
    "    for pair in edges_list_2:\n",
    "        a, b = pair\n",
    "        # print(\" --- looking into the interaction between\", a, \"and\", b)\n",
    "        formula = f'SIR ~ {a} + {b} + {a}:{b}'\n",
    "        model = ols(formula, data=labeled_matrix_2).fit()\n",
    "        anova_table = sm.stats.anova_lm(model, typ=2)\n",
    "\n",
    "        if not np.isnan(anova_table['PR(>F)'][2]):\n",
    "            p_values[anova_table.index[2]]=anova_table['PR(>F)'][2]\n",
    "        else:\n",
    "            save_nan_testing.append(pair)\n",
    "            count_nan+=1\n",
    "        \n",
    "print(f\"{count_nan} interactions have nan p-values - removed\")\n",
    "\n",
    "# mst correction - using fdr\n",
    "\n",
    "gene_gene_pair = list(p_values.keys())\n",
    "p_values_list = list(p_values.values())\n",
    "\n",
    "reject, corrected_p_values= fdrcorrection(p_values_list)\n",
    "\n",
    "results_df = pd.DataFrame({\n",
    "    'gene_gene': gene_gene_pair,\n",
    "    'p-value': p_values_list,\n",
    "    'corr_p_value': corrected_p_values,\n",
    "    'reject_H0': reject\n",
    "})\n",
    "\n",
    "count_rejected=0\n",
    "for i in reject:\n",
    "    if i:\n",
    "        count_rejected+=1\n",
    "\n",
    "print(f\"Number of rejected H0: {count_rejected} out of {len(reject)}\")\n",
    "\n",
    "new_edge_pairs=results_df[results_df['reject_H0'] == True]['gene_gene'].values\n",
    "\n",
    "edge_pairs_tuples = [tuple(pair.split(':')) for pair in new_edge_pairs]\n",
    "edge_pairs_tuples\n",
    "edge_pairs_tuples=[(i.replace('_', ' '), j.replace('_', ' ')) for i, j in edge_pairs_tuples]\n",
    "\n",
    "G_sig=nx.Graph()\n",
    "G_sig.add_edges_from(edge_pairs_tuples)\n",
    "\n",
    "# G_sig=set_cluster_attributes(G_sig)\n",
    "\n",
    "nx.write_graphml(G_sig, f'../../data/graphs-6/corrected_{n_models}_randomized_SVM_{t}signed_corr_{species}_{drug}.graphml')\n",
    "\n",
    "print(f\"final number of nodes: {G_sig.number_of_nodes()}, final number of edges: {G_sig.number_of_edges()}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "G_sig=nx.read_graphml('../../data/graphs-6/corrected_200_randomized_SVM_0.4signed_corr_Campylobacter_coli_ciprofloxacin.graphml')\n",
    "G_sig = set_cluster_attributes(G_sig)\n",
    "\n",
    "nx.write_graphml(G_sig,'../../data/graphs-6/corrected_200_randomized_SVM_0.4signed_corr_Campylobacter_coli_ciprofloxacin.graphml')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### LOR of R co-occurence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "LOR_df = compute_cooccurence_LOR(gene_pairs=edges_list)\n",
    "#try it on threshold 0.5"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
