{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "import os\n",
    "\n",
    "\n",
    "clstr_path=\"../pangenome-repo/Pangenome-Analysis-Workflow/codes/Campylobacter_coli/cdhit_test.fatsa.clstr\"\n",
    "freq_path=\"../pangenome-repo/Pangenome-Analysis-Workflow/codes/Campylobacter_coli/Campylobacter_coli_cluster_frequencies.csv\"\n",
    "pan_annot_path='../pangenome-repo/Pangenome-Analysis-Workflow/codes/Campylobacter_coli/Campylobacter_coli_pangenome.csv'\n",
    "clstr_fasta_path='../pangenome-repo/Pangenome-Analysis-Workflow/codes/Campylobacter_coli/Campylobacter_coli.fasta'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## essentials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import re\n",
    "import os\n",
    "import networkx as nx\n",
    "\n",
    "def generate_files_paths(pipeline_output_path):\n",
    "    '''\n",
    "    takes the output dir path of the pipeline for a species, and returns the 4 file paths inside of it needed to extract the data\n",
    "\n",
    "    Parameters:\n",
    "    -----------\n",
    "    pipeline_output_path: str, path to the pipeline output directory\n",
    "\n",
    "    Returns:\n",
    "    --------\n",
    "    clstr_path: str, path to the cluster file, cdhit output\n",
    "    freq_path: str, path to the cluster frequency file\n",
    "    pan_annot_path: str, path to the pangenome annotation file\n",
    "    clstr_fasta_path: str, path to the cluster fasta file, aslo cd hit output\n",
    "    '''\n",
    "\n",
    "    pipeline_output_path=pipeline_output_path.rstrip('/')\n",
    "    basename=os.path.basename(pipeline_output_path)\n",
    "\n",
    "    clstr_path=pipeline_output_path+'/'+basename+'.fasta.clstr'\n",
    "    freq_path=pipeline_output_path+'/'+basename+'_cluster_frequencies.csv'\n",
    "    pan_annot_path=pipeline_output_path+'/'+basename+'_pangenome.csv'\n",
    "    clstr_fasta_path=pipeline_output_path+'/'+basename+'.fasta'\n",
    "\n",
    "    return clstr_path, freq_path, pan_annot_path, clstr_fasta_path\n",
    "\n",
    "# clstr_path, freq_path, pan_annot_path, clstr_fasta_path= generate_files_paths(\"../pangenome-repo/Pangenome-Analysis-Workflow/codes/Campylobacter_coli/\")\n",
    "\n",
    "def get_cluster_frequency(freq_path):\n",
    "    '''\n",
    "    Takes the path to the cluster frequency file and returns a dataframe with the cluster frequencies\n",
    "\n",
    "    Parameters:\n",
    "    freq_path: str\n",
    "\n",
    "    Returns:\n",
    "    freq_df: pandas.DataFrame\n",
    "    '''\n",
    "\n",
    "    freq_df=pd.read_csv(freq_path, index_col=0)\n",
    "    freq_df.columns=['Cluster','frequency']\n",
    "    freq_df.index=freq_df['Cluster']\n",
    "    freq_df.drop('Cluster', axis=1, inplace=True)\n",
    "\n",
    "    return freq_df\n",
    "\n",
    "def get_cluster_representatives(clstr_path:str, save_csv:bool=False)->pd.DataFrame:\n",
    "    '''\n",
    "    Take a CD-HIT clstr output file and return a dictionary with cluster number as key and representative sequence as value.\n",
    "    Optionally saves the dictionary as a csv file.\n",
    "\n",
    "    parameters:\n",
    "    ----------\n",
    "    clstr_path: str, path to CD-HIT clstr file\n",
    "    save_csv: bool, if True saves the dictionary as a csv file (optional)\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    df: pd.DataFrame, data frame with cluster number as key and representative sequence as value\n",
    "\n",
    "    description:\n",
    "    ------------\n",
    "\n",
    "    CD-HIT cluster file parser:\n",
    "        .clstr file format is:\n",
    "\n",
    "        >Cluster 489\n",
    "        0\t552aa, >fig|195.2069.peg.1050... *\n",
    "        >Cluster 490\n",
    "        0\t551aa, >fig|195.2045.peg.1513... *\n",
    "        1\t530aa, >fig|195.2053.peg.1024... at 84.34%\n",
    "        2\t536aa, >fig|195.2186.peg.1791... at 94.40%\n",
    "        3\t541aa, >fig|195.2242.peg.678... at 94.82%\n",
    "        >Cluster 491\n",
    "        0\t501aa, >fig|195.2049.peg.1588... at 99.80%\n",
    "        1\t551aa, >fig|195.2166.peg.1287... *\n",
    "        2\t453aa, >fig|195.2206.peg.1626... at 99.56%\n",
    "        3\t453aa, >fig|195.2226.peg.1423... at 99.78%\n",
    "\n",
    "        Each cluster has a number, here representing a CDS; each entry starts with a >Cluster line followed by the cluster number\n",
    "        Under each cluster there are all entries for that cluster in each sample genome, based on a sequence identity (here 80% from pipeline)\n",
    "        One entry under a cluster is the representative sequence, marked with an asterisk (*), this is the entry we want to get its gene name from PATRIC \n",
    "        p.s., 195.2029.peg.1780 is the gene name for cluster 0\n",
    "\n",
    "        The cluster parser aims to match each cluster number to its representative sequence and get the gene name from PATRIC using api requests\n",
    "    '''\n",
    "    with open(clstr_path, 'r') as clstr_file:\n",
    "\n",
    "        found_representative=False #flag to know if we found a representative for the clstr number\n",
    "        cluster='' #initialize cluster number\n",
    "\n",
    "        cluster_dict={} #initialize dictionary to store cluster number and representative sequence\n",
    "\n",
    "        for line in clstr_file:\n",
    "\n",
    "            if line.startswith(\">\"): #if its a cluster number file, save the cluster number for the next iteration\n",
    "                cluster=line[1:].strip()\n",
    "                found_representative=False #reset to false in order to find rep\n",
    "\n",
    "            elif not found_representative:\n",
    "                if \"*\" in line:\n",
    "                    found_representative=True\n",
    "                    representative=line.split(\",\")[1].replace('... *', '').strip()[1:]\n",
    "                    representative=representative.split(\"|\")[1]\n",
    "                    # print(representative)\n",
    "                    cluster_dict[cluster]=representative\n",
    "\n",
    "            else:\n",
    "                continue\n",
    "\n",
    "    df=pd.DataFrame(cluster_dict.items(), columns=['cluster', 'gene_representative'])\n",
    "    df.set_index('cluster', inplace=True)\n",
    "    df.index.name='Cluster'\n",
    "    \n",
    "    return df\n",
    "\n",
    "def get_representative_products(clstr_fasta_path:str)->pd.DataFrame:\n",
    "    '''\n",
    "    Takes a CD-HIT fasta output and returns a dataframe that matches each CDS PATRIC ID with its product name\n",
    "\n",
    "    the fasta output is of this form:\n",
    "\n",
    "    >fig|195.2024.peg.83 Putative oxidoreductase ferredoxin-type protein, clusters with CPO\n",
    "    MNFSQISDACVKCGKCIPVCTIHEVNRDETTSPRGFLDLLAAYKEEKLELDKEAKKIFES\n",
    "    CFLCTNCVEVCPSKLRVDNVIEEVRYDIAKKFGIAWYKKIIFFFLRRRKILDLVAKLGYV\n",
    "    FQSCAFKIQSQDQNVGMKAKFSMPFVKKGRLLTSFNKKSFLNSNPDFIDNGGEKTVGFFV\n",
    "    GCLANYFYIDTANAVLKIAKEVKINVDLMKEQVCCGAPQFFTGDFKSVEILAKKNIEYFE\n",
    "    KKLEKLDAIIIPEATCSAMLKIDLEHFFNMQNEPEWAKRAQKISSRIYMASEYFYKFTNL\n",
    "    KELLESKKKLNYSITYHDPCHARKMQGVFKEPRELLKANYHFVEMSNPNACCGFGGVSMQ\n",
    "    TDYYDRALSVGLKKASMIDESKACVVSAECSACRMQISNALEQNSSKAIFASPLELIAKA\n",
    "    L\n",
    "    >fig|195.2024.peg.542 Septum-associated rare lipoprotein A\n",
    "    MKPYTINGKTYYPTVVSVGETADGIASWYGPGFHGKKTSNGETYNQNGLTAAHKTLPMNT\n",
    "    ILKVTNLNNNRQVTVRVNDRGPFVNNRIIDLSKGAASQIDMIAAGTAPVRLEVIGFGSAN\n",
    "    SGNNVVHSNINYGASGGIANNGQIYEGGNFMVQIGAFKNPSGAQTIASRYKTYRTYSSTI\n",
    "    RKSSVDGLSRVFLTGFRSEEEARDFAASGAFAGAFVVRE\n",
    "\n",
    "    The aim is to match 195.2024.peg.542 with Septum-associated rare lipoprotein A in teh dataframe\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - clstr_fasta_path: str, path of the fasta output\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    - df: pd.DataFrame, df of columns gene_representative and product_name\n",
    "    '''\n",
    "\n",
    "    dict={}\n",
    "\n",
    "    with open(clstr_fasta_path,'r') as f:\n",
    "        for line in f.readlines():\n",
    "            if line.startswith(\">\"):\n",
    "                pattern=\">fig.(\\d{3}\\.\\d{4}.peg.\\d+) (.+)\"\n",
    "                gene_representative=re.match(string=line, pattern=pattern).group(1)\n",
    "                product_name=re.match(string=line, pattern=pattern).group(2)\n",
    "                dict[gene_representative]=product_name\n",
    "    \n",
    "\n",
    "    df = pd.DataFrame(dict.items(), columns=['gene_representative', 'product_name'])\n",
    "    \n",
    "    return df\n",
    "\n",
    "def combine_cluster_product(clstr_rep_df:pd.DataFrame, rep_prod_df:pd.DataFrame)->pd.DataFrame:\n",
    "    '''\n",
    "    Combines the cluster representative dataframe with the product name dataframe\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - clstr_rep_df: pd.DataFrame, dataframe with cluster number and gene representative (get_cluster_representatives() output)\n",
    "    - rep_prod_df: pd.DataFrame, dataframe with gene representative and product name (get_representative_products() output)\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    - df: pd.DataFrame, dataframe with cluster number, gene representative and product name\n",
    "    '''\n",
    "    # if clstr rep df doesnt have a column named Cluster, create one out of index\n",
    "    if 'Cluster' not in clstr_rep_df.columns:\n",
    "        clstr_rep_df['Cluster']=clstr_rep_df.index\n",
    "\n",
    "    df=pd.merge(clstr_rep_df, rep_prod_df, on='gene_representative', how='left')\n",
    "    df.drop(columns='gene_representative', inplace=True)\n",
    "\n",
    "    df.set_index('Cluster', inplace=True)\n",
    "\n",
    "    return df\n",
    "\n",
    "def get_cluster_pan_gene_class(pan_annot_path:str)->pd.DataFrame:\n",
    "    '''\n",
    "    Takes a pangenome annotation file and returns a dataframe with cluster number and gene class\n",
    "    This annotation file is a <species>_pangenome.csv file output from the pangenome analysis pipeline\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - pan_annot_path: str, path to the pangenome annotation file\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    - pan_df: pd.DataFrame, dataframe with cluster number and gene class\n",
    "    '''\n",
    "    with open(pan_annot_path, 'r') as f:\n",
    "        pan_df=pd.read_csv(f, index_col=0)\n",
    "        pan_df.drop(pan_df.columns[[1]], axis=1, inplace=True)\n",
    "        pan_df.columns=['Cluster', 'pan_gene_class']\n",
    "\n",
    "    pan_df.set_index('Cluster', inplace=True)\n",
    "    return pan_df\n",
    "\n",
    "def get_cluster_representatives(clstr_path:str, save_csv:bool=False)->pd.DataFrame:\n",
    "    '''\n",
    "    Take a CD-HIT clstr output file and return a dictionary with cluster number as key and representative sequence as value.\n",
    "    Optionally saves the dictionary as a csv file.\n",
    "\n",
    "    parameters:\n",
    "    ----------\n",
    "    clstr_path: str, path to CD-HIT clstr file\n",
    "    save_csv: bool, if True saves the dictionary as a csv file (optional)\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    df: pd.DataFrame, data frame with cluster number as key and representative sequence as value\n",
    "\n",
    "    description:\n",
    "    ------------\n",
    "\n",
    "    CD-HIT cluster file parser:\n",
    "        .clstr file format is:\n",
    "\n",
    "        >Cluster 489\n",
    "        0\t552aa, >fig|195.2069.peg.1050... *\n",
    "        >Cluster 490\n",
    "        0\t551aa, >fig|195.2045.peg.1513... *\n",
    "        1\t530aa, >fig|195.2053.peg.1024... at 84.34%\n",
    "        2\t536aa, >fig|195.2186.peg.1791... at 94.40%\n",
    "        3\t541aa, >fig|195.2242.peg.678... at 94.82%\n",
    "        >Cluster 491\n",
    "        0\t501aa, >fig|195.2049.peg.1588... at 99.80%\n",
    "        1\t551aa, >fig|195.2166.peg.1287... *\n",
    "        2\t453aa, >fig|195.2206.peg.1626... at 99.56%\n",
    "        3\t453aa, >fig|195.2226.peg.1423... at 99.78%\n",
    "\n",
    "        Each cluster has a number, here representing a CDS; each entry starts with a >Cluster line followed by the cluster number\n",
    "        Under each cluster there are all entries for that cluster in each sample genome, based on a sequence identity (here 80% from pipeline)\n",
    "        One entry under a cluster is the representative sequence, marked with an asterisk (*), this is the entry we want to get its gene name from PATRIC \n",
    "        p.s., 195.2029.peg.1780 is the gene name for cluster 0\n",
    "\n",
    "        The cluster parser aims to match each cluster number to its representative sequence and get the gene name from PATRIC using api requests\n",
    "    '''\n",
    "    with open(clstr_path, 'r') as clstr_file:\n",
    "\n",
    "        found_representative=False #flag to know if we found a representative for the clstr number\n",
    "        cluster='' #initialize cluster number\n",
    "\n",
    "        cluster_dict={} #initialize dictionary to store cluster number and representative sequence\n",
    "\n",
    "        for line in clstr_file:\n",
    "\n",
    "            if line.startswith(\">\"): #if its a cluster number file, save the cluster number for the next iteration\n",
    "                cluster=line[1:].strip()\n",
    "                found_representative=False #reset to false in order to find rep\n",
    "\n",
    "            elif not found_representative:\n",
    "                if \"*\" in line:\n",
    "                    found_representative=True\n",
    "                    representative=line.split(\",\")[1].replace('... *', '').strip()[1:]\n",
    "                    representative=representative.split(\"|\")[1]\n",
    "                    # print(representative)\n",
    "                    cluster_dict[cluster]=representative\n",
    "\n",
    "            else:\n",
    "                continue\n",
    "\n",
    "    df=pd.DataFrame(cluster_dict.items(), columns=['cluster', 'gene_representative'])\n",
    "    df.set_index('cluster', inplace=True)\n",
    "    df.index.name='Cluster'\n",
    "\n",
    "    return df\n",
    "\n",
    "def get_cluster_attributes(clstr_path, freq_path, pan_annot_path, clstr_fasta_path):\n",
    "\n",
    "\n",
    "    clstr_patric_id_df=get_cluster_representatives(clstr_path)\n",
    "    clstr_gene_class_df= get_cluster_pan_gene_class(pan_annot_path)\n",
    "    _patric_id_product_df=get_representative_products(clstr_fasta_path)\n",
    "    clstr_product_df=combine_cluster_product(clstr_patric_id_df, _patric_id_product_df)\n",
    "    clstr_freq_df=get_cluster_frequency(freq_path)\n",
    "\n",
    "    df=pd.concat([clstr_patric_id_df, clstr_gene_class_df, clstr_product_df, clstr_freq_df], axis=1)\n",
    "\n",
    "    #if there exist a Cluster col remove\n",
    "    if 'Cluster' in df.columns:\n",
    "        df.drop('Cluster', axis=1, inplace=True)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def create_dicts(df:pd.DataFrame):\n",
    "    '''\n",
    "    takes a nxm data frame and returns m dictionaries of keys the index, and values each col repectively\n",
    "\n",
    "    param:\n",
    "    -----\n",
    "    - df: pd.DataFrame, index=Cluster, cols=gene_representative,\tpan_gene_class,\tproduct_name,\tfrequency\n",
    "\n",
    "    returns:\n",
    "    --------\n",
    "    - dicts: dict, dict of m dictionaries - key is the col name, value is the dictionary of the col\n",
    "    '''\n",
    "    dicts = {col: df[col].to_dict() for col in df.columns}\n",
    "    return dicts\n",
    "\n",
    "def set_node_attributes_by_cluster(df:pd.DataFrame, G:nx.graph):\n",
    "    '''\n",
    "    takes a dataframe and a graph and sets the node attributes for each cluster\n",
    "\n",
    "    param:\n",
    "    -----\n",
    "    - df: pd.DataFrame, index=Cluster, cols=gene_representative,\tpan_gene_class,\tproduct_name,\tfrequency\n",
    "    - G: nx.Graph, graph to set the node attributes\n",
    "\n",
    "    returns:\n",
    "    --------\n",
    "    - G: nx.Graph, graph with node attributes set\n",
    "    '''\n",
    "    dicts = create_dicts(df)\n",
    "\n",
    "    for key in dicts.keys():\n",
    "        nx.set_node_attributes(G, dicts[key], key)\n",
    "\n",
    "    return G\n",
    "\n",
    "_output_path='../pangenome-repo/Pangenome-Analysis-Workflow/codes/Campylobacter_coli/'\n",
    "\n",
    "def set_output_path(path:str):\n",
    "    '''\n",
    "    sets the output path for the pipeline\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - path: str, path to the output directory\n",
    "    '''\n",
    "    global _output_path\n",
    "    _output_path=path\n",
    "    return _output_path\n",
    "\n",
    "def set_cluster_attributes(G:nx.Graph, pipeline_output=_output_path)->nx.Graph:\n",
    "    '''\n",
    "    takes a Graph and the pipeline's output path and performs all of the following: \n",
    "\n",
    "    - get all the files paths required for the attributes\n",
    "    - generate df for each and concatenate them all in one of index Cluster\n",
    "    - create dictionaries out of it and perform nx.set_node_attributes using each on G\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    G: nx.Graph, network of nodes Clusters\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    G: nx.Graph, network of nodes Clusters, attributes: pan_gene_class, gene_representative, pan_gene_class, frequency\n",
    "    '''\n",
    "\n",
    "    p1, p2, p3, p4 = generate_files_paths(pipeline_output)\n",
    "    df = get_cluster_attributes(p1, p2, p3, p4)\n",
    "    G = set_node_attributes_by_cluster(df, G)\n",
    "\n",
    "    return G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "p1, p2, p3, p4 = generate_files_paths(pipeline_output_path='/home/raysas/projects/capstone/pangenome-repo/Pangenome-Analysis-Workflow/codes/Campylobacter_coli/')\n",
    "test_df = get_cluster_attributes(p1, p2, p3, p4)\n",
    "dicts= create_dicts(test_df)\n",
    "\n",
    "test_G=nx.read_graphml('../data/graphs-5/FDR_500randomized_SVM_0.6signed_corr_Campylobacter_coli_ciprofloxacin_significant.graphml')\n",
    "\n",
    "test_G_2=set_cluster_attributes(test_G)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "works :)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## cluster to PATRIC ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Frequency</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cluster 0</th>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1</th>\n",
       "      <td>239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 3</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4</th>\n",
       "      <td>268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8016</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8017</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8018</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8019</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8020</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8021 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Frequency\n",
       "Cluster                \n",
       "Cluster 0           103\n",
       "Cluster 1           239\n",
       "Cluster 2             3\n",
       "Cluster 3             1\n",
       "Cluster 4           268\n",
       "...                 ...\n",
       "Cluster 8016          1\n",
       "Cluster 8017          1\n",
       "Cluster 8018          1\n",
       "Cluster 8019          1\n",
       "Cluster 8020          1\n",
       "\n",
       "[8021 rows x 1 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_cluster_frequency(freq_path):\n",
    "    '''\n",
    "    Takes the path to the cluster frequency file and returns a dataframe with the cluster frequencies\n",
    "\n",
    "    Parameters:\n",
    "    freq_path: str\n",
    "\n",
    "    Returns:\n",
    "    freq_df: pandas.DataFrame\n",
    "    '''\n",
    "\n",
    "    freq_df=pd.read_csv(freq_path, index_col=0)\n",
    "    freq_df.columns=['Cluster','frequency']\n",
    "    freq_df.index=freq_df['Cluster']\n",
    "    freq_df.drop('Cluster', axis=1, inplace=True)\n",
    "\n",
    "    return freq_df\n",
    "\n",
    "f_df=get_cluster_frequency(freq_path)\n",
    "\n",
    "f_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_representative</th>\n",
       "      <th>pan_gene_class</th>\n",
       "      <th>product_name</th>\n",
       "      <th>Frequency</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cluster 0</th>\n",
       "      <td>195.2029.peg.1780</td>\n",
       "      <td>accessory</td>\n",
       "      <td>helicase, Snf2 family</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1</th>\n",
       "      <td>195.2024.peg.1611</td>\n",
       "      <td>accessory</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2</th>\n",
       "      <td>195.2136.peg.1150</td>\n",
       "      <td>unique</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 3</th>\n",
       "      <td>195.2145.peg.407</td>\n",
       "      <td>unique</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4</th>\n",
       "      <td>195.2024.peg.685</td>\n",
       "      <td>core</td>\n",
       "      <td>DNA-directed RNA polymerase beta' subunit (EC ...</td>\n",
       "      <td>268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8016</th>\n",
       "      <td>195.2201.peg.1720</td>\n",
       "      <td>unique</td>\n",
       "      <td>Endonuclease I precursor (EC 3.1.21.1)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8017</th>\n",
       "      <td>195.2201.peg.1791</td>\n",
       "      <td>unique</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8018</th>\n",
       "      <td>195.2210.peg.964</td>\n",
       "      <td>unique</td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8019</th>\n",
       "      <td>195.2230.peg.1815</td>\n",
       "      <td>unique</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8020</th>\n",
       "      <td>195.2314.peg.347</td>\n",
       "      <td>unique</td>\n",
       "      <td>Oxygen-insensitive NAD(P)H nitroreductase (EC ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8021 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_representative pan_gene_class  \\\n",
       "Cluster                                           \n",
       "Cluster 0      195.2029.peg.1780      accessory   \n",
       "Cluster 1      195.2024.peg.1611      accessory   \n",
       "Cluster 2      195.2136.peg.1150         unique   \n",
       "Cluster 3       195.2145.peg.407         unique   \n",
       "Cluster 4       195.2024.peg.685           core   \n",
       "...                          ...            ...   \n",
       "Cluster 8016   195.2201.peg.1720         unique   \n",
       "Cluster 8017   195.2201.peg.1791         unique   \n",
       "Cluster 8018    195.2210.peg.964         unique   \n",
       "Cluster 8019   195.2230.peg.1815         unique   \n",
       "Cluster 8020    195.2314.peg.347         unique   \n",
       "\n",
       "                                                   product_name  Frequency  \n",
       "Cluster                                                                     \n",
       "Cluster 0                                 helicase, Snf2 family        103  \n",
       "Cluster 1                                  hypothetical protein        239  \n",
       "Cluster 2                                  hypothetical protein          3  \n",
       "Cluster 3                                  hypothetical protein          1  \n",
       "Cluster 4     DNA-directed RNA polymerase beta' subunit (EC ...        268  \n",
       "...                                                         ...        ...  \n",
       "Cluster 8016             Endonuclease I precursor (EC 3.1.21.1)          1  \n",
       "Cluster 8017                               hypothetical protein          1  \n",
       "Cluster 8018                     conserved hypothetical protein          1  \n",
       "Cluster 8019                               hypothetical protein          1  \n",
       "Cluster 8020  Oxygen-insensitive NAD(P)H nitroreductase (EC ...          1  \n",
       "\n",
       "[8021 rows x 4 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_cluster_representatives(clstr_path:str, save_csv:bool=False)->pd.DataFrame:\n",
    "    '''\n",
    "    Take a CD-HIT clstr output file and return a dictionary with cluster number as key and representative sequence as value.\n",
    "    Optionally saves the dictionary as a csv file.\n",
    "\n",
    "    parameters:\n",
    "    ----------\n",
    "    clstr_path: str, path to CD-HIT clstr file\n",
    "    save_csv: bool, if True saves the dictionary as a csv file (optional)\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    df: pd.DataFrame, data frame with cluster number as key and representative sequence as value\n",
    "\n",
    "    description:\n",
    "    ------------\n",
    "\n",
    "    CD-HIT cluster file parser:\n",
    "        .clstr file format is:\n",
    "\n",
    "        >Cluster 489\n",
    "        0\t552aa, >fig|195.2069.peg.1050... *\n",
    "        >Cluster 490\n",
    "        0\t551aa, >fig|195.2045.peg.1513... *\n",
    "        1\t530aa, >fig|195.2053.peg.1024... at 84.34%\n",
    "        2\t536aa, >fig|195.2186.peg.1791... at 94.40%\n",
    "        3\t541aa, >fig|195.2242.peg.678... at 94.82%\n",
    "        >Cluster 491\n",
    "        0\t501aa, >fig|195.2049.peg.1588... at 99.80%\n",
    "        1\t551aa, >fig|195.2166.peg.1287... *\n",
    "        2\t453aa, >fig|195.2206.peg.1626... at 99.56%\n",
    "        3\t453aa, >fig|195.2226.peg.1423... at 99.78%\n",
    "\n",
    "        Each cluster has a number, here representing a CDS; each entry starts with a >Cluster line followed by the cluster number\n",
    "        Under each cluster there are all entries for that cluster in each sample genome, based on a sequence identity (here 80% from pipeline)\n",
    "        One entry under a cluster is the representative sequence, marked with an asterisk (*), this is the entry we want to get its gene name from PATRIC \n",
    "        p.s., 195.2029.peg.1780 is the gene name for cluster 0\n",
    "\n",
    "        The cluster parser aims to match each cluster number to its representative sequence and get the gene name from PATRIC using api requests\n",
    "    '''\n",
    "    with open(clstr_path, 'r') as clstr_file:\n",
    "\n",
    "        found_representative=False #flag to know if we found a representative for the clstr number\n",
    "        cluster='' #initialize cluster number\n",
    "\n",
    "        cluster_dict={} #initialize dictionary to store cluster number and representative sequence\n",
    "\n",
    "        for line in clstr_file:\n",
    "\n",
    "            if line.startswith(\">\"): #if its a cluster number file, save the cluster number for the next iteration\n",
    "                cluster=line[1:].strip()\n",
    "                found_representative=False #reset to false in order to find rep\n",
    "\n",
    "            elif not found_representative:\n",
    "                if \"*\" in line:\n",
    "                    found_representative=True\n",
    "                    representative=line.split(\",\")[1].replace('... *', '').strip()[1:]\n",
    "                    representative=representative.split(\"|\")[1]\n",
    "                    # print(representative)\n",
    "                    cluster_dict[cluster]=representative\n",
    "\n",
    "            else:\n",
    "                continue\n",
    "\n",
    "    df=pd.DataFrame(cluster_dict.items(), columns=['cluster', 'gene_representative'])\n",
    "    df.set_index('cluster', inplace=True)\n",
    "    df.index.name='Cluster'\n",
    "\n",
    "    return df\n",
    "\n",
    "def get_cluster_attributes(clstr_path, freq_path, pan_annot_path, clstr_fasta_path):\n",
    "\n",
    "\n",
    "    clstr_patric_id_df=get_cluster_representatives(clstr_path)\n",
    "    clstr_gene_class_df= get_cluster_pan_gene_class(pan_annot_path)\n",
    "    _patric_id_product_df=get_representative_products(clstr_fasta_path)\n",
    "    clstr_product_df=combine_cluster_product(clstr_patric_id_df, _patric_id_product_df)\n",
    "    clstr_freq_df=get_cluster_frequency(freq_path)\n",
    "\n",
    "    df=pd.concat([clstr_patric_id_df, clstr_gene_class_df, clstr_product_df, clstr_freq_df], axis=1)\n",
    "\n",
    "    #if there exist a Cluster col remove\n",
    "    if 'Cluster' in df.columns:\n",
    "        df.drop('Cluster', axis=1, inplace=True)\n",
    "    \n",
    "    return df\n",
    "\n",
    "get_cluster_attributes(clstr_path, freq_path, pan_annot_path, clstr_fasta_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_representative</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cluster 0</th>\n",
       "      <td>195.2029.peg.1780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1</th>\n",
       "      <td>195.2024.peg.1611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2</th>\n",
       "      <td>195.2136.peg.1150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 3</th>\n",
       "      <td>195.2145.peg.407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4</th>\n",
       "      <td>195.2024.peg.685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8016</th>\n",
       "      <td>195.2201.peg.1720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8017</th>\n",
       "      <td>195.2201.peg.1791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8018</th>\n",
       "      <td>195.2210.peg.964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8019</th>\n",
       "      <td>195.2230.peg.1815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8020</th>\n",
       "      <td>195.2314.peg.347</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8021 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_representative\n",
       "Cluster                         \n",
       "Cluster 0      195.2029.peg.1780\n",
       "Cluster 1      195.2024.peg.1611\n",
       "Cluster 2      195.2136.peg.1150\n",
       "Cluster 3       195.2145.peg.407\n",
       "Cluster 4       195.2024.peg.685\n",
       "...                          ...\n",
       "Cluster 8016   195.2201.peg.1720\n",
       "Cluster 8017   195.2201.peg.1791\n",
       "Cluster 8018    195.2210.peg.964\n",
       "Cluster 8019   195.2230.peg.1815\n",
       "Cluster 8020    195.2314.peg.347\n",
       "\n",
       "[8021 rows x 1 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clstr_rep_df=get_cluster_representatives(clstr_path, save_csv=True)\n",
    "clstr_rep_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PATRIC ID to gene name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<!DOCTYPE html>\n",
      "<html lang=\"en\">\n",
      "\t<head>\n",
      "\t\t\n",
      "\t\n",
      "\n",
      "\t\n",
      "\t  \n",
      "\t\t<link rel='stylesheet' href=\"/js/3.35.5/p3/resources/p3.css\" />\n",
      "\t  \n",
      "\t\t<link rel='stylesheet' href=\"/js/3.35.5/molstar/mol-bvbrc/molstar.css\" />\n",
      "\t  \n",
      "\t\t<link rel='stylesheet' href=\"/js/3.35.5/mauve_viewer/dist/mauve-viewer.css\" />\n",
      "\t  \n",
      "\t\t<link rel='stylesheet' href=\"/js/3.35.5/heatmap/dist/hotmap.css\" />\n",
      "\t  \n",
      "\t\n",
      "\n",
      "\n",
      "\t<meta http-equiv=\"Content-Type\" content=\"text/html; charset=utf-8\"/>\n",
      "\t<meta name=\"Keywords\" content=\"PATRIC, Pathosystems Resource Integration Center\"/>\n",
      "\t<meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">\n",
      "\t<meta name=\"google-site-verification\" content=\"TkjceNtwo7upCQl_gx8Zmhhp9tCmc08LitZGuvQ1Eb8\" />\n",
      "\t<link rel=\"manifest\" href=\"/public/manifest.json\">\n",
      "\t<link rel=\"icon\" sizes=\"16x16 32x32\" href=\"/patric/favicon.ico\">\n",
      "\t<link href=\"https://fonts.googleapis.com/css2?family=Roboto+Slab:wght@300;400&family=Roboto:wght@300;400;700&display=swap\" rel=\"stylesheet\">\n",
      "\n",
      "\t</head>\n",
      "\t<body onunload=\"\" class=\"claro patric \">\n",
      "\t\t<div id=\"ApplicationContainer\" data-dojo-type=\"dijit/layout/BorderContainer\" class=\"layoutContainer\" data-dojo-props=\"id:'ApplicationContainer',gutters:false,liveSplitters:false\">\n",
      "\t\t\t<div data-dojo-type=\"dijit/layout/ContentPane\" class=\"HeaderBar\" data-dojo-props=\"region:'top'\">\n",
      "    \n",
      "\n",
      "    <div id=\"bv-brc-header\" class=\"P3Header\" role=\"heading\" aria-level=\"1\">\n",
      "\n",
      "        <div id=\"bv-brc-header-container\">\n",
      "            <div class=\"nav-logo\">\n",
      "                <a class=\"nav-logo-link\" href=\"/\">\n",
      "                    <img class=\"bv-brc-logo\" src=\"/patric/images/bv-brc/bv-brc-header-logo.png\" alt=\"BV-BRC Logo\" style=\"height: 28px;\"/>\n",
      "                    <span><i style=\"color: #ffffff; font-size: 12px; padding-left: 2px;\">3.35.5 </i></i></span>\n",
      "                </a>\n",
      "            </div>\n",
      "\n",
      "            <div style=\"vertical-align: top; min-width: 435px; flex: 1\">\n",
      "                <div class=\"topMenuButton showOnLoad dijitHidden\" data-dojo-type=\"dijit/form/DropDownButton\">\n",
      "                    <span>ORGANISMS</span>\n",
      "                    <div data-dojo-type=\"p3/widget/TooltipDialog\">\n",
      "                        <div style=\"vertical-align: top;width:225px;display:inline-block;margin-right: 5px;\">\n",
      "                            <div class=\"menuHeader\">Bacterial Pathogens</div>\n",
      "                            <div class=\"menuItems\">\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/469#view_tab=overview\" title=\"Acinetobacter overview\">Acinetobacter</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Francisella overview\"\n",
      "                                           href=\"/view/Taxonomy/262#view_tab=overview\">Francisella</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/1386#view_tab=overview\" title=\"Bacillus overview\">Bacillus</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Helicobacter overview\"\n",
      "                                           href=\"/view/Taxonomy/209#view_tab=overview\">Helicobacter</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/773#view_tab=overview\" title=\"Bartonella overview\">Bartonella</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Listeria overview\"\n",
      "                                           href=\"/view/Taxonomy/1637#view_tab=overview\">Listeria</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/64895#view_tab=overview\" title=\"Borreliella overview\">Borreliella</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Mycobacterium overview\"\n",
      "                                           href=\"/view/Taxonomy/1763#view_tab=overview\">Mycobacterium</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/234#view_tab=overview\" title=\"Brucella overview\">Brucella</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Pseudomonas overview\"\n",
      "                                           href=\"/view/Taxonomy/286#view_tab=overview\">Pseudomonas</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/32008#view_tab=overview\" title=\"Burkholderia overview\">Burkholderia</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Rickettsia overview\"\n",
      "                                           href=\"/view/Taxonomy/780#view_tab=overview\">Rickettsia</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/194#view_tab=overview\" title=\"Campylobacter overview\">Campylobacter</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Salmonella overview\"\n",
      "                                           href=\"/view/Taxonomy/590#view_tab=overview\">Salmonella</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/810#view_tab=overview\" title=\"Chlamydia overview\">Chlamydia</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Shigella overview\"\n",
      "                                           href=\"/view/Taxonomy/620#view_tab=overview\">Shigella</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/1485#view_tab=overview\" title=\"Clostridium overview\">Clostridium</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Staphylococcus overview\"\n",
      "                                           href=\"/view/Taxonomy/1279#view_tab=overview\">Staphylococcus</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/776#view_tab=overview\" title=\"Coxiella overview\">Coxiella</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Streptococcus overview\"\n",
      "                                           href=\"/view/Taxonomy/1301#view_tab=overview\">Streptococcus</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/943#view_tab=overview\" title=\"Ehrlichia overview\">Ehrlichia</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Vibrio overview\"\n",
      "                                           href=\"/view/Taxonomy/662#view_tab=overview\">Vibrio</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/561#view_tab=overview\" title=\"Escherichia overview\">Escherichia</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Yersinia overview\"\n",
      "                                           href=\"/view/Taxonomy/629#view_tab=overview\">Yersinia</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <br>\n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Bacteria/2\" title=\"Overview page for all Bacteria genomes\">All Bacteria</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/2157\" title=\"Overview page for all Archea genomes\">All Archaea</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/10239\" title=\"Overview page for all Viruses\">All Viruses</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Host/?eq(taxon_lineage_ids,2759)#view_tab=genomes\"\n",
      "                                            title=\"Overview page for all Eukaryotic host genomes\">Eukaryotic Hosts</a></span>\n",
      "                                </div>\n",
      "                            </div>\n",
      "                        </div>\n",
      "\n",
      "                        <div style=\"vertical-align: top;width:341px;display:inline-block;margin-right: 5px;\">\n",
      "                            <div class=\"menuHeader\">Viral Families</div>\n",
      "\n",
      "                            <div class=\"menuItems\">\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/10508#view_tab=overview\" title=\"Adenoviridae overview\">Adenoviridae</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Orthomyxoviridae overview\"\n",
      "                                           href=\"/view/Taxonomy/11308#view_tab=overview\">Orthomyxoviridae</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/137992#view_tab=overview\" title=\"Asfarviridae overview\">Asfarviridae</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Paramyxoviridae overview\"\n",
      "                                           href=\"/view/Taxonomy/11158#view_tab=overview\">Paramyxoviridae</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/1980410#view_tab=overview\" title=\"Bunyavirales overview\">Bunyavirales</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Parvoviridae overview\"\n",
      "                                           href=\"/view/Taxonomy/10780#view_tab=overview\">Parvoviridae</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/11974#view_tab=overview\" title=\"Caliciviridae overview\">Caliciviridae</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Picornaviridae overview\"\n",
      "                                           href=\"/view/Taxonomy/12058#view_tab=overview\">Picornaviridae</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/11118#view_tab=overview\" title=\"Coronaviridae overview\">Coronaviridae</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Pneumoviridae overview\"\n",
      "                                           href=\"/view/Taxonomy/11244#view_tab=overview\">Pneumoviridae</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/11266#view_tab=overview\" title=\"Filoviridae overview\">Filoviridae</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Polyomaviridae overview\"\n",
      "                                           href=\"/view/Taxonomy/151341#view_tab=overview\">Polyomaviridae</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/11050#view_tab=overview\" title=\"Flaviviridae overview\">Flaviviridae</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Poxviridae overview\"\n",
      "                                           href=\"/view/Taxonomy/10240#view_tab=overview\">Poxviridae</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/10404#view_tab=overview\" title=\"Hepadnaviridae overview\">Hepadnaviridae</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Reovirales overview\"\n",
      "                                           href=\"/view/Taxonomy/2732541#view_tab=overview\">Reovirales</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/291484#view_tab=overview\" title=\"Hepeviridae overview\">Hepeviridae</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Rhabdoviridae overview\"\n",
      "                                           href=\"/view/Taxonomy/11270#view_tab=overview\">Rhabdoviridae</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/548681#view_tab=overview\" title=\"Herpesvirales overview\">Herpesvirales</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Togaviridae overview\"\n",
      "                                           href=\"/view/Taxonomy/11018#view_tab=overview\">Togaviridae</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                            </div>\n",
      "\n",
      "                            <div class=\"menuHeader\">Featured Viruses</div>\n",
      "\n",
      "                            <div class=\"menuItems\">\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/GenomeList/?eq(genome_name,phage)\" title=\"Bacteriophages overview\">Bacteriophages</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Influenza A virus overview\"\n",
      "                                           href=\"/view/Taxonomy/11320#view_tab=overview\">Influenza A virus</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/12637#view_tab=overview\" title=\"Dengue virus overview\">Dengue virus</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Mammarenavirus lassaense overview\"\n",
      "                                           href=\"/view/Taxonomy/3052310#view_tab=overview\">Mammarenavirus lassaense</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/186536#view_tab=overview\" title=\"Ebolavirus overview\">Ebolavirus</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Monkeypox virus overview\"\n",
      "                                           href=\"/view/Taxonomy/10244#view_tab=overview\">Monkeypox virus</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/12059#view_tab=overview\" title=\"Enterovirus overview\">Enterovirus</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"SARS-CoV-2 overview\"\n",
      "                                           href=\"/view/Taxonomy/2697049#view_tab=overview\">SARS-CoV-2</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                                <div>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        <a class=\"\" href=\"/view/Taxonomy/11102#view_tab=overview\" title=\"Hepacivirus overview\">Hepacivirus</a>\n",
      "                                    </span>\n",
      "                                    <span class=\"menuItem half\">\n",
      "                                        \n",
      "                                        <a class=\"\" title=\"Zika virus overview\"\n",
      "                                           href=\"/view/Taxonomy/64320#view_tab=overview\">Zika virus</a>\n",
      "                                        \n",
      "                                    </span>\n",
      "                                </div>\n",
      "                                \n",
      "                            </div>\n",
      "\n",
      "                        </div>\n",
      "                    </div>\n",
      "                </div>\n",
      "\n",
      "                <div class=\"topMenuButton showOnLoad dijitHidden\" data-dojo-type=\"dijit/form/DropDownButton\">\n",
      "                    <span>SEARCHES</span>\n",
      "\n",
      "                    <div data-dojo-type=\"p3/widget/TooltipDialog\">\n",
      "                        <div class=\"menuHeader\" style=\"padding-right: 20px;\">Advanced Search</div>\n",
      "                        <div><a class=\"navigationLink\" href=\"/searches/TaxaSearch\">Taxa</a></div>\n",
      "                        <div><a class=\"navigationLink\" href=\"/searches/GenomeSearch\">Genomes</a></div>\n",
      "                        <div><a class=\"navigationLink\" href=\"/searches/StrainSearch\">Strains</a></div>\n",
      "                        <div><a class=\"navigationLink\" href=\"/searches/GenomicFeatureSearch\">Features</a></div>\n",
      "                        <div><a class=\"navigationLink\" href=\"/searches/ProteinSearch\">Proteins</a></div>\n",
      "                        <div><a class=\"navigationLink\" href=\"/searches/SpecialtyGeneSearch\">Specialty Genes</a></div>\n",
      "                        <div><a class=\"navigationLink\" href=\"/searches/DomainAndMotifSearch\">Domains and Motifs</a></div>\n",
      "                        <div><a class=\"navigationLink\" href=\"/searches/EpitopeSearch\">Epitopes</a></div>\n",
      "                        <div><a class=\"navigationLink\" href=\"/searches/ProteinStructureSearch\">Protein Structures</a></div>\n",
      "                        <div><a class=\"navigationLink\" href=\"/searches/PathwaySearch\">Pathways</a></div>\n",
      "                        <div><a class=\"navigationLink\" href=\"/searches/SubsystemSearch\">Subsystems</a></div>\n",
      "                        <div><a class=\"navigationLink\" href=\"/searches/SurveillanceSearch\">Surveillance</a></div>\n",
      "                        <div><a class=\"navigationLink\" href=\"/searches/SerologySearch\">Serology</a></div>\n",
      "                    </div>\n",
      "                </div>\n",
      "\n",
      "                <div class=\"topMenuButton showOnLoad dijitHidden\" data-dojo-type=\"dijit/form/DropDownButton\">\n",
      "                    <span>TOOLS &amp; SERVICES</span>\n",
      "\n",
      "                    <div data-dojo-type=\"p3/widget/TooltipDialog\">\n",
      "                        <div>\n",
      "                            <div style=\"vertical-align: top;width: 180px;margin-right:5px;display:inline-block\">\n",
      "                                <div class=\"menuHeader\">Genomics</div>\n",
      "                                <div class=\"menuItems\">\n",
      "                                    <a class=\"navigationLink\" href=\"/app/Assembly2\">Genome Assembly <span style=\"font-style: italic;\">(B)</span></a>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/Annotation\">Genome Annotation</a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/ComprehensiveGenomeAnalysis\">Comprehensive Genome Analysis <span style=\"font-style: italic;\">(B)</span></a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/ComprehensiveSARS2Analysis\">SARS-CoV-2 Genome Analysis <span style=\"font-style: italic;\">(V)</span></a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/Homology\">BLAST</a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/PrimerDesign\">Primer Design</a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/GenomeDistance\">Similar Genome Finder</div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/GenomeAlignment\">Genome Alignment <span style=\"font-style: italic;\">(B)</span></a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/Variation\">Variation Analysis <span style=\"font-style: italic;\">(B)</span></a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/Tnseq\">Tn-Seq Analysis <span style=\"font-style: italic;\">(B)</span></a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/SubspeciesClassification\">Subspecies Classification <span style=\"font-style: italic;\">(V)</span></a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/SequenceSubmission\">Sequence Submission <span style=\"font-style: italic;\">(V)</span></a></div>\n",
      "                                </div>\n",
      "                                <div class=\"menuHeader\">Phylogenomics</div>\n",
      "                                <div class=\"menuItems\">\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/PhylogeneticTree\">Bacterial Genome Tree</a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/ViralGenomeTree\">Viral Genome Tree</a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/GeneTree\">Gene/Protein Tree</a></div>\n",
      "                                </div>\n",
      "                                <br>\n",
      "                                <div>\n",
      "                                    <a class=\"navigationLink\" href=\"/view/VariantLineage/#view_tab=overview\" style=\"font-weight: bold;\">SARS-CoV-2 Variant Tracker</span></a>\n",
      "                                </div>\n",
      "                            </div>\n",
      "\n",
      "                            <div style=\"vertical-align:top;width:200px;display:inline-block\">\n",
      "                                <div class=\"menuHeader\">Protein Tools</div>\n",
      "                                <div class=\"menuItems\">\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/MSA\">MSA and SNP Analysis</a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/MetaCATS\">Meta-CATS</a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/SeqComparison\">Proteome Comparison <span style=\"font-style: italic;\">(B)</span></a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/ComparativeSystems\">Protein Family Sorter</a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/ComparativeSystems\">Comparative Pathway</div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/ComparativeSystems\">Comparative Systems</a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/HASubtypeNumberingConversion\">HA Subtype Numbering Conversion <span style=\"font-style: italic;\">(V)</span></a></div>\n",
      "                                </div>\n",
      "                                <div class=\"menuHeader\">Metagenomics</div>\n",
      "                                <div class=\"menuItems\">\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/MetagenomicReadMapping\">Metagenomic Read Mapping</a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/TaxonomicClassification\">Taxonomic Classification</a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/app/MetagenomicBinning\">Metagenomic Binning</a></div>\n",
      "                                </div>\n",
      "                              <div class=\"menuHeader\">Transcriptomics</div>\n",
      "                              <div class=\"menuItems\">\n",
      "                                  <div><a class=\"navigationLink\" href=\"/app/Expression\">Expression Import</a></div>\n",
      "                                  <div><a class=\"navigationLink\" href=\"/app/Rnaseq\">RNA-Seq Analysis</a></div>\n",
      "                              </div>\n",
      "                              <p></p>\n",
      "                              <div class=\"menuHeader\">Utilities</div>\n",
      "                              <div class=\"menuItems\">\n",
      "                                  <div><a class=\"navigationLink\" href=\"/app/IDMapper\">ID Mapper</a></div>\n",
      "                                  <div><a class=\"navigationLink\" href=\"/app/FastqUtil\">Fastq Utilities</a></div>\n",
      "                              </div>\n",
      "                            </div>\n",
      "                        </div>\n",
      "                    </div>\n",
      "                </div>\n",
      "\n",
      "                <div class=\"topMenuButton showOnLoad dijitHidden\" data-dojo-type=\"dijit/form/DropDownButton\">\n",
      "                    <span>WORKSPACES</span>\n",
      "\n",
      "                    <div data-dojo-type=\"p3/widget/TooltipDialog\">\n",
      "                        <div>\n",
      "                            <div style=\"vertical-align: top;width:175px;display:inline-block\">\n",
      "                                <div class=\"HideWithAuth\">\n",
      "                                    Please <a class=\"loginLink\">sign in</a> to view your workspaces.\n",
      "                                </div>\n",
      "                                <div class=\"ShowWithAuth\">\n",
      "                                    <div><i class=\"fa icon-caret-down fa-1x noHoverIcon\"\n",
      "                                            style=\"color:#333;cursor:normal;\"></i>&nbsp;<a\n",
      "                                                class=\"navigationLink\" id=\"YourWorkspaceLink\">Workspaces</a></div>\n",
      "                                    <div id=\"YourWorkspaces\"></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/workspace/public\">Public Workspaces</a></div>\n",
      "                                    <div><a class=\"navigationLink\" href=\"/workspace/public/ARWattam@patricbrc.org/BV-BRC%20Workshop\">BV-BRC Workshop</a></div>\n",
      "                                </div>\n",
      "                            </div>\n",
      "                            <div style=\"vertical-align:top;display:inline-block\" class=\"ShowWithAuth\">\n",
      "                                <div>\n",
      "                                    <a class=\"navigationLink\" href=\"/job/\">My Jobs</a>\n",
      "                                </div>\n",
      "                                <div>\n",
      "                                    <a class=\"navigationLink\" href=\"/view/GenomeList/?eq(public,false)\">My Genomes</a>\n",
      "                                </div>\n",
      "                            </div>\n",
      "                        </div>\n",
      "                    </div>\n",
      "                </div>\n",
      "\n",
      "                <div class=\"topMenuButton showOnLoad dijitHidden\" data-dojo-type=\"dijit/form/DropDownButton\">\n",
      "                    <span>HELP</span>\n",
      "\n",
      "                    <div data-dojo-type=\"p3/widget/TooltipDialog\">\n",
      "                        <div style=\"font-weight: bold; font-size: 14px;\"><a class=\"navigationLinkOut\" href=\"https://www.bv-brc.org/docs/\">Documentation Website</a></div>\n",
      "\n",
      "                        <div style=\"margin-top: 5px; margin-bottom: 5px;\" class=\"menuHeader\">Updates and Reports</div>\n",
      "                        <div style=\"margin-left: 10px;\"><a class=\"navigationLinkOut\" href=\"https://www.bv-brc.org/docs/news/index.html\">News and Announcements</a></div>\n",
      "                        <div style=\"margin-left: 10px;\"><a class=\"navigationLinkOut\" href=\"https://www.bv-brc.org/docs/reports_and_analysis/index.html\">Reports and Analysis</a></div>\n",
      "\n",
      "                        <div style=\"margin-top: 5px; margin-bottom: 5px;\" class=\"menuHeader\">Help Resources</div>\n",
      "                        <div style=\"margin-left: 10px;\"><a class=\"navigationLinkOut\" href=\"https://www.bv-brc.org/docs/overview.html\">Overview</a></div>\n",
      "                        <div style=\"margin-left: 10px;\"><a class=\"navigationLinkOut\" href=\"https://www.bv-brc.org/docs/quick_start/index.html\">Quick Start</a></div>\n",
      "                        <div style=\"margin-left: 10px;\"><a class=\"navigationLinkOut\" href=\"https://www.bv-brc.org/docs/data_protocols/index.html\">Data Sources &amp; Protocols</a></div>\n",
      "                        <div style=\"margin-left: 10px;\"><a class=\"navigationLinkOut\" href=\"https://www.bv-brc.org/docs/quick_references/index.html\">Quick Reference Guides</a></div>\n",
      "                        <div style=\"margin-left: 10px;\"><a class=\"navigationLinkOut\" href=\"https://www.bv-brc.org/docs/tutorial/index.html\">Tutorials</a></div>\n",
      "                        <div style=\"margin-left: 10px;\"><a class=\"navigationLinkOut\" href=\"https://www.bv-brc.org/docs/common_tasks/index.html\">Common Tasks</a></div>\n",
      "                        <div style=\"margin-left: 10px;\"><a class=\"navigationLinkOut\" href=\"https://www.bv-brc.org/docs/cli_tutorial/index.html\">Command Line Interface</a></div>\n",
      "                        <div style=\"margin-left: 10px;\"><a class=\"navigationLinkOut\" href=\"https://www.bv-brc.org/docs/videos/index.html\">Instructional Videos</a></div>\n",
      "                        <div style=\"margin-left: 10px;\"><a class=\"navigationLinkOut\" href=\"https://www.bv-brc.org/docs/webinar/index.html\">Webinars</a></div>\n",
      "\n",
      "                        <div style=\"margin-bottom: 5px;\" class=\"menuHeader\">System Documentation</div>\n",
      "                        <div style=\"margin-left: 10px;\"><a class=\"navigationLinkOut\" href=\"https://www.bv-brc.org/docs/system_documentation/system_architecture.html\">System Architecture</a></div>\n",
      "                        <div style=\"margin-left: 10px;\"><a class=\"navigationLinkOut\" href=\"https://www.bv-brc.org/docs/test_procedures/index.html\">Test Procedures</a></div>\n",
      "                        <div style=\"margin-left: 10px;\"><a class=\"navigationLinkOut\" href=\"https://www.bv-brc.org/docs/usage_metrics.html\">Usage Metrics</a></div>\n",
      "\n",
      "                    </div>\n",
      "                </div>\n",
      "\n",
      "                <div class=\"topMenuButton showOnLoad dijitHidden\" data-dojo-type=\"dijit/form/DropDownButton\">\n",
      "                    <span>ABOUT</span>\n",
      "\n",
      "                    <div data-dojo-type=\"p3/widget/TooltipDialog\">\n",
      "                        <div><a class=\"TopMenuLink\" href=\"/about\">About BV-BRC</a></div>\n",
      "                        <div><a class=\"DialogButton\" rel=\"reportProblem:\">Contact Us</a></div>\n",
      "                        <div><a class=\"TopMenuLink\" href=\"/team\">Our Team</a></div>\n",
      "                        <div><a class=\"TopMenuLink\" href=\"https://www.bv-brc.org/docs/news/index.html\" target=\"_blank\">Announcements</a></div>\n",
      "                        <div><a class=\"TopMenuLink\" href=\"/brc-calendar\">BRC Calendar</a></div>\n",
      "                        <hr>\n",
      "                        <div><a class=\"TopMenuLink\" href=\"/publications\">Publications</a></div>\n",
      "                        <div><a class=\"TopMenuLink\" href=\"/citation\">Citation</a></div>\n",
      "                        <div><a class=\"TopMenuLink\" href=\"/related-resources\">Related Resources</a></div>\n",
      "                        <div><a class=\"TopMenuLink navigationLinkOut\" href=\"https://github.com/BV-BRC\" target=\"_blank\" rel=\"noopener\">BV-BRC GitHub</a></div>\n",
      "                        <div><a class=\"TopMenuLink\" href=\"/privacy-policy\">Privacy Policy</a></div>\n",
      "                    </div>\n",
      "                </div>\n",
      "            </div>\n",
      "\n",
      "            <div id=\"p3header-search\" style=\"padding: 6px 10px 2px 2px; width: 120px; flex-grow: 0.85\">\n",
      "                <input style=\"width:100%;font-size:1em;\" class=\"showOnLoad dijitHidden\"\n",
      "                       data-dojo-type=\"p3/widget/GlobalSearch\"/>\n",
      "            </div>\n",
      "\n",
      "            <!-- <div class=\"navbar-flex-space\" style=\"flex-grow: 0.85;\"></div> -->\n",
      "\n",
      "            <div class=\"HideWithAuth authLinks showOnLoad dijitHidden\" style=\"padding: 8px 10px 0 0;\">\n",
      "                <a class=\"LoginButton navigationLink\" href=\"/register\">Register</a>\n",
      "                <a class=\"LoginButton navigationLink\" href=\"/login\">Sign In</a>\n",
      "            </div>\n",
      "\n",
      "\n",
      "            <div class=\"showOnLoad ShowWithAuth dijitHidden\">\n",
      "                <div class=\"showOnLoad ShowWithAuth dijitHidden topMenuButtonIcon\"\n",
      "                        data-dojo-type=\"dijit/form/DropDownButton\" data-dojo-props=\"iconClass:'icon-user fa-2x login-btn'\">\n",
      "                    <span></span>\n",
      "                    <div class=\"warning_icon user_warning_icon icon-warning fa fa-1x\"> </div>\n",
      "                    <div data-dojo-type=\"p3/widget/TooltipDialog\">\n",
      "                        <div>Signed in as <span id=\"signedInAs\" style=\"font-weight:bold;\"></span></div>\n",
      "                        <hr>\n",
      "                        <div><a class=\"userProfile\">Account Settings <i class=\"warning_icon account-settings-warning icon-warning fa fa-1x\" style=\"vertical-align:middle;\"> </i></a></div>\n",
      "                        <div><a class=\"sulogin\">SU Login</a></div>\n",
      "                        <div><a class=\"suSwitchBack\" onclick=\"window.App.suSwitchBack()\" style=\"display:none\">SU Switch Back</a></div>\n",
      "                        <div><a class=\"logoutLink\" onclick=\"window.App.logout()\">Logout</a></div>\n",
      "                    </div>\n",
      "                </div>\n",
      "            </div>\n",
      "\n",
      "\n",
      "        </div>\n",
      "\n",
      "    </div>\n",
      "    <!-- -->\n",
      "    <!-- <div class=\"msg-box-below-topnav\"><strong>The BV-BRC site will be having a maintenance day this coming Thursday, March 14. A full outage is not planned but there will be intermittent interruptions in the execution of backend service jobs between 9:00 a.m. and 3:00 p.m. CDT. -- BV-BRC Team</strong></div>  -->\n",
      "    <!-- <div class=\"msg-box-below-topnav\"><strong>The BV-BRC scheduler will be undergoing maintenance on June 6, from 9:00AM-Noon (Chicago Time). Already running jobs will not be affected but job status and submission of new jobs will be unavailable. We are sorry for the inconvenience. -- BV-BRC Team</strong></div> -->\n",
      "    <!-- <div class=\"msg-box-below-topnav\"><strong>Welcome to the BV-BRC Alpha website. This site is used for active development and testing of the BV-BRC resource. Hence, the data and functionality may change frequently and without notice.</strong></div> -->\n",
      "    <!-- <div class=\"msg-box-below-topnav\"><strong>Welcome to the BV-BRC Beta website. This site is used to seek early community feedback for the BV-BRC resource, which is still under development. Please note that the data and functionality on this site may change frequently and without notice.</strong></div> -->\n",
      "</div>\n",
      "\n",
      "\n",
      "\t\t\t<div data-dojo-type=\"dijit/layout/ContentPane\" data-dojo-props=\"region:'center'\" style=\"padding:0px;border:0px solid #ddd;background:inherit;\">\n",
      "\n",
      "\t\t\t</div>\n",
      "\t\t\t<div data-dojo-type=\"p3/widget/WorkspaceController\" data-dojo-props=\"region:'bottom'\"></div>\n",
      "\t\t</div>\n",
      "\t\t<script>\n",
      "dojoConfig = {\n",
      "    parseOnLoad: false,\n",
      "    packages: [\n",
      "      {name: 'lazyload', location: \"../lazyload\", main: 'lazyload'},\n",
      "      {name: 'jDataView', location: '../jDataView/src', main: 'jdataview'},\n",
      "      {name: \"FileSaver\", location: \"../FileSaver\", main: \"FileSaver\"}\n",
      "    ]\n",
      "  }\n",
      "</script>\n",
      "\n",
      "<!-- Scripts -->\n",
      "\n",
      "\n",
      "\n",
      "  \n",
      "    <script type=\"text/javascript\" src=\"/js/3.35.5/mauve_viewer/dist/mauve-viewer.js\"></script>\n",
      "  \n",
      "    <script type=\"text/javascript\" src=\"/js/3.35.5/msa/dist/msa.min.js\"></script>\n",
      "  \n",
      "    <script type=\"text/javascript\" src=\"/js/3.35.5/bvbrc_js_client/dist/bvbrc_client.js\"></script>\n",
      "  \n",
      "    <script type=\"text/javascript\" src=\"/js/3.35.5/bundle/bundle.js\"></script>\n",
      "  \n",
      "    <script type=\"text/javascript\" src=\"/js/3.35.5/bundle/bundle2.js\"></script>\n",
      "  \n",
      "    <script type=\"text/javascript\" src=\"/js/3.35.5/clipboard-js/clipboard.min.js\"></script>\n",
      "  \n",
      "\n",
      "\n",
      "<script type=\"text/javascript\">\n",
      "  // LinkedIn Feed\n",
      "  let $linkedInFeedElm = $(\"#linkedin-feed\");\n",
      "\n",
      "  // Fetch the data only if feed element exists in the page\n",
      "  if ($linkedInFeedElm.length) {\n",
      "    $.getJSON({\n",
      "      url: '/linkedin/posts/?count=8',\n",
      "      success: function (data) {\n",
      "        for (let post of data.elements) {\n",
      "          if (post.id && post.id.includes('share')) {\n",
      "            $(\"#linkedin-slides\").append(\n",
      "              `<li class=\"slide\" style=\"height: 100%;\">\n",
      "                  <iframe src=\"https://www.linkedin.com/embed/feed/update/${post.id}\" style=\"height: 100%;width: 100%;\"/>\n",
      "              </li>`\n",
      "            );\n",
      "            $(\"#linkedin-dot-navigation\").append(\n",
      "              `<div class=\"dot\"></div>`\n",
      "            );\n",
      "          }\n",
      "        }\n",
      "\n",
      "        initializeSlide($('#linkedin-slides .slide'), $(\"#linkedin-dot-navigation .dot\"));\n",
      "      },\n",
      "      error: function( data ) {\n",
      "        console.log(data);\n",
      "      }\n",
      "    });\n",
      "  }\n",
      "\n",
      "  // News & Announcements Carousel\n",
      "  $.getJSON(\"https://www.bv-brc.org/docs/_static/carousel.json\", function (feed) {\n",
      "      for (var i = 0; i < feed.length; i++) {\n",
      "          var link = feed[i].link;\n",
      "          var title = feed[i].title;\n",
      "          var desc = feed[i].desc;\n",
      "          var word_limit = (feed[i].img == \"\") ? 60 : 20\n",
      "          var descSnippit = desc.split(\" \", word_limit).join(\" \");\n",
      "          var img = feed[i].img;\n",
      "          $(\"#announcements-slides\").append(\n",
      "              `<li class=\"slide\">\n",
      "                  <a href=\"${link}\" class=\"feed-link\">\n",
      "                      <h4 class=\"feed-title\">${title}</h4>\n",
      "                  </a>\n",
      "                  <p class=\"feed-description\">${descSnippit} ... <a href=\"${link}\">read more</a></p>\n",
      "                  <img src=\"${img}\" class=\"feed-img\">\n",
      "              </li>`\n",
      "          );\n",
      "          $(\"#announcements-dot-navigation\").append(\n",
      "              `<div class=\"dot\"></div>`\n",
      "          );\n",
      "      }\n",
      "\n",
      "      initializeSlide($('#announcements-slides .slide'), $(\"#announcements-dot-navigation .dot\"));\n",
      "  });\n",
      "\n",
      "  function initializeSlide($slides, $dots) {\n",
      "    $slides.first().addClass(\"showing\");\n",
      "    $dots.first().addClass(\"selected\");\n",
      "\n",
      "    var index = 0;\n",
      "    $slides.eq(0).addClass(\"current-slide\");\n",
      "    $dots.eq(0).addClass(\"current-dot\");\n",
      "    $dots.click(function () {\n",
      "      $dots.eq(index).removeClass(\"current-dot\");\n",
      "      $slides.eq(index).removeClass(\"current-slide\");\n",
      "      $(this).addClass(\"current-dot\");\n",
      "      index = $(this).index();\n",
      "      $slides.eq(index).addClass(\"current-slide\");\n",
      "    });\n",
      "  }\n",
      "</script>\n",
      "\n",
      "<!-- // use bundle.js\n",
      "<script type=\"text/javascript\" src=\"/js/cytoscape/dist/cytoscape.min.js\"></script>\n",
      "<script type=\"text/javascript\" src=\"/js/jquery/dist/jquery.min.js\"></script>\n",
      "<script type=\"text/javascript\" src=\"/js/webcola/WebCola/cola.min.js\"></script>\n",
      "<script type=\"text/javascript\" src=\"/js/dagre/dist/dagre.min.js\"></script>\n",
      "<script type=\"text/javascript\" src=\"/js/cytoscape-cose-bilkent/cytoscape-cose-bilkent.js\"></script>\n",
      "-->\n",
      "\n",
      "\n",
      "  <script src=\"/js/3.35.5/dojo/dojo.js\"></script>\n",
      "  <script>\n",
      "  require([\"p3/layer/core\"], function(){\n",
      "    \n",
      "      require([\"p3/app/p3app\"], function(App){\n",
      "        appOpts = {\"version\":\"3.0\",\"gaID\":\"G-GS8F60MG56\",\"uaID\":\"UA-165638318-1\",\"probModelSeedServiceURL\":\"https://p3.theseed.org/services/ProbModelSEED\",\"shockServiceURL\":\"https://p3.theseed.org/services/shock_api\",\"workspaceServiceURL\":\"https://p3.theseed.org/services/Workspace\",\"appBaseURL\":\"https://www.bv-brc.org\",\"appServiceURL\":\"https://p3.theseed.org/services/app_service\",\"dataServiceURL\":\"https://www.bv-brc.org/api\",\"homologyServiceURL\":\"https://p3.theseed.org/services/homology_service\",\"mailinglistURL\":\"https://lists.brcgateway.org/\",\"genomedistanceServiceURL\":\"https://p3.theseed.org/services/minhash_service\",\"compareregionServiceURL\":\"https://p3.theseed.org/services/compare_regions\",\"docsServiceURL\":\"https://www.bv-brc.org/docs/\",\"enableDevTools\":false,\"accountURL\":\"https://user.bv-brc.org\",\"appLabel\":\"\",\"jiraLabel\":\"PROD\",\"appVersion\":\"3.35.5\",\"userServiceURL\":\"https://user.bv-brc.org\",\"localStorageCheckInterval\":86400};\n",
      "        appOpts.workspaceAPI = \"https://p3.theseed.org/services/Workspace\";\n",
      "        appOpts.serviceAPI = \"https://p3.theseed.org/services/app_service\";\n",
      "        appOpts.dataAPI = \"https://www.bv-brc.org/api\";\n",
      "        appOpts.userServiceURL = \"https://user.bv-brc.org\";\n",
      "        appOpts.mailinglistURL = \"https://lists.brcgateway.org/\";\n",
      "        appOpts.localStorageCheckInterval = \"86400\";\n",
      "        window.App = new App(appOpts);\n",
      "      });\n",
      "      \n",
      "      });\n",
      "      \n",
      "      </script>\n",
      "      <!-- Google Analytics 4 -->\n",
      "      \n",
      "        <!-- Global site tag (gtag.js) - Google Analytics -->\n",
      "        <script async src=\"https://www.googletagmanager.com/gtag/js?id=G-GS8F60MG56\"></script>\n",
      "        <script>\n",
      "        window.dataLayer = window.dataLayer || [];\n",
      "        function gtag(){dataLayer.push(arguments);}\n",
      "        gtag('js', new Date());\n",
      "\n",
      "        gtag('config', 'G-GS8F60MG56', {\n",
      "          'link_attribution': true\n",
      "        });\n",
      "        </script>\n",
      "        \n",
      "\n",
      "      <!-- Google UA (EOL) -->\n",
      "      \n",
      "          <!-- Global site tag (gtag.js) - Google Analytics -->\n",
      "          <script async src=\"https://www.googletagmanager.com/gtag/js?id=UA-165638318-1\"></script>\n",
      "          <script>\n",
      "          window.dataLayer = window.dataLayer || [];\n",
      "          function gtag(){dataLayer.push(arguments);}\n",
      "          gtag('js', new Date());\n",
      "\n",
      "          gtag('config', 'UA-165638318-1', {\n",
      "            'link_attribution': true\n",
      "          });\n",
      "          </script>\n",
      "          \n",
      "\t</body>\n",
      "</html>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#now getting the gene names out of patric\n",
    "\n",
    "genome_id=\"195.2069\"\n",
    "AMR_db=\"CARD\"\n",
    "drug=\"ciprofloxacin\"\n",
    "link=f\"https://www.bv-brc.org/view/Genome/{genome_id}#view_tab=specialtyGenes&filter=and(eq(property,%22Antibiotic%20Resistance%22),eq(source,%22{AMR_db}%22),eq(antibiotics,%22{drug}%22))\"\n",
    "\n",
    "response=requests.get(link)\n",
    "print(response.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PATRIC ID to product name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_representative_products(clstr_fasta_path:str)->pd.DataFrame:\n",
    "    '''\n",
    "    Takes a CD-HIT fasta output and returns a dataframe that matches each CDS PATRIC ID with its product name\n",
    "\n",
    "    the fasta output is of this form:\n",
    "\n",
    "    >fig|195.2024.peg.83 Putative oxidoreductase ferredoxin-type protein, clusters with CPO\n",
    "    MNFSQISDACVKCGKCIPVCTIHEVNRDETTSPRGFLDLLAAYKEEKLELDKEAKKIFES\n",
    "    CFLCTNCVEVCPSKLRVDNVIEEVRYDIAKKFGIAWYKKIIFFFLRRRKILDLVAKLGYV\n",
    "    FQSCAFKIQSQDQNVGMKAKFSMPFVKKGRLLTSFNKKSFLNSNPDFIDNGGEKTVGFFV\n",
    "    GCLANYFYIDTANAVLKIAKEVKINVDLMKEQVCCGAPQFFTGDFKSVEILAKKNIEYFE\n",
    "    KKLEKLDAIIIPEATCSAMLKIDLEHFFNMQNEPEWAKRAQKISSRIYMASEYFYKFTNL\n",
    "    KELLESKKKLNYSITYHDPCHARKMQGVFKEPRELLKANYHFVEMSNPNACCGFGGVSMQ\n",
    "    TDYYDRALSVGLKKASMIDESKACVVSAECSACRMQISNALEQNSSKAIFASPLELIAKA\n",
    "    L\n",
    "    >fig|195.2024.peg.542 Septum-associated rare lipoprotein A\n",
    "    MKPYTINGKTYYPTVVSVGETADGIASWYGPGFHGKKTSNGETYNQNGLTAAHKTLPMNT\n",
    "    ILKVTNLNNNRQVTVRVNDRGPFVNNRIIDLSKGAASQIDMIAAGTAPVRLEVIGFGSAN\n",
    "    SGNNVVHSNINYGASGGIANNGQIYEGGNFMVQIGAFKNPSGAQTIASRYKTYRTYSSTI\n",
    "    RKSSVDGLSRVFLTGFRSEEEARDFAASGAFAGAFVVRE\n",
    "\n",
    "    The aim is to match 195.2024.peg.542 with Septum-associated rare lipoprotein A in teh dataframe\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - clstr_fasta_path: str, path of the fasta output\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    - df: pd.DataFrame, df of columns gene_representative and product_name\n",
    "    '''\n",
    "\n",
    "    dict={}\n",
    "\n",
    "    with open(clstr_fasta_path,'r') as f:\n",
    "        for line in f.readlines():\n",
    "            if line.startswith(\">\"):\n",
    "                pattern=\">fig.(\\d{3}\\.\\d{4}.peg.\\d+) (.+)\"\n",
    "                gene_representative=re.match(string=line, pattern=pattern).group(1)\n",
    "                product_name=re.match(string=line, pattern=pattern).group(2)\n",
    "                dict[gene_representative]=product_name\n",
    "    \n",
    "\n",
    "    df = pd.DataFrame(dict.items(), columns=['gene_representative', 'product_name'])\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_representative</th>\n",
       "      <th>product_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>195.2024.peg.575</td>\n",
       "      <td>Putative periplasmic protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>195.2024.peg.1455</td>\n",
       "      <td>Fe2+ ABC transporter, permease protein 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>195.2024.peg.1693</td>\n",
       "      <td>4-hydroxybenzoyl-CoA thioesterase family activ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>195.2024.peg.935</td>\n",
       "      <td>Cinnamyl alcohol dehydrogenase/reductase (EC 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>195.2024.peg.24</td>\n",
       "      <td>Two-component transcriptional response regulat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8016</th>\n",
       "      <td>195.2317.peg.807</td>\n",
       "      <td>Beta-1,3-glucosyltransferase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8017</th>\n",
       "      <td>195.2317.peg.1426</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8018</th>\n",
       "      <td>195.2317.peg.1246</td>\n",
       "      <td>ATP synthase gamma chain (EC 3.6.3.14)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8019</th>\n",
       "      <td>195.2317.peg.1018</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8020</th>\n",
       "      <td>195.2317.peg.1780</td>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8021 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_representative                                       product_name\n",
       "0       195.2024.peg.575                       Putative periplasmic protein\n",
       "1      195.2024.peg.1455           Fe2+ ABC transporter, permease protein 2\n",
       "2      195.2024.peg.1693  4-hydroxybenzoyl-CoA thioesterase family activ...\n",
       "3       195.2024.peg.935  Cinnamyl alcohol dehydrogenase/reductase (EC 1...\n",
       "4        195.2024.peg.24  Two-component transcriptional response regulat...\n",
       "...                  ...                                                ...\n",
       "8016    195.2317.peg.807                       Beta-1,3-glucosyltransferase\n",
       "8017   195.2317.peg.1426                               hypothetical protein\n",
       "8018   195.2317.peg.1246             ATP synthase gamma chain (EC 3.6.3.14)\n",
       "8019   195.2317.peg.1018                               hypothetical protein\n",
       "8020   195.2317.peg.1780                               hypothetical protein\n",
       "\n",
       "[8021 rows x 2 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rep_prod_df=get_representative_products(clstr_fasta_path)\n",
    "rep_prod_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_cluster_product(clstr_rep_df:pd.DataFrame, rep_prod_df:pd.DataFrame)->pd.DataFrame:\n",
    "    '''\n",
    "    Combines the cluster representative dataframe with the product name dataframe\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - clstr_rep_df: pd.DataFrame, dataframe with cluster number and gene representative (get_cluster_representatives() output)\n",
    "    - rep_prod_df: pd.DataFrame, dataframe with gene representative and product name (get_representative_products() output)\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    - df: pd.DataFrame, dataframe with cluster number, gene representative and product name\n",
    "    '''\n",
    "    # if clstr rep df doesnt have a column named Cluster, create one out of index\n",
    "    if 'Cluster' not in clstr_rep_df.columns:\n",
    "        clstr_rep_df['Cluster']=clstr_rep_df.index\n",
    "\n",
    "    df=pd.merge(clstr_rep_df, rep_prod_df, on='gene_representative', how='left')\n",
    "    df.drop(columns='gene_representative', inplace=True)\n",
    "\n",
    "    df.set_index('Cluster', inplace=True)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>product_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cluster 0</th>\n",
       "      <td>helicase, Snf2 family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1</th>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2</th>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 3</th>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4</th>\n",
       "      <td>DNA-directed RNA polymerase beta' subunit (EC ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8016</th>\n",
       "      <td>Endonuclease I precursor (EC 3.1.21.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8017</th>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8018</th>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8019</th>\n",
       "      <td>hypothetical protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8020</th>\n",
       "      <td>Oxygen-insensitive NAD(P)H nitroreductase (EC ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8021 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   product_name\n",
       "Cluster                                                        \n",
       "Cluster 0                                 helicase, Snf2 family\n",
       "Cluster 1                                  hypothetical protein\n",
       "Cluster 2                                  hypothetical protein\n",
       "Cluster 3                                  hypothetical protein\n",
       "Cluster 4     DNA-directed RNA polymerase beta' subunit (EC ...\n",
       "...                                                         ...\n",
       "Cluster 8016             Endonuclease I precursor (EC 3.1.21.1)\n",
       "Cluster 8017                               hypothetical protein\n",
       "Cluster 8018                     conserved hypothetical protein\n",
       "Cluster 8019                               hypothetical protein\n",
       "Cluster 8020  Oxygen-insensitive NAD(P)H nitroreductase (EC ...\n",
       "\n",
       "[8021 rows x 1 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clstr_prod_df=combine_cluster_product(get_cluster_representatives(clstr_path), get_representative_products(clstr_fasta_path))\n",
    "clstr_prod_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=combine_cluster_product(get_cluster_representatives(clstr_path), get_representative_products(clstr_fasta_path))\n",
    "df.to_csv('../data/cluster_descriptions/cluster_product.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## cluster pan annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cluster_pan_gene_class(pan_annot_path:str)->pd.DataFrame:\n",
    "    '''\n",
    "    Takes a pangenome annotation file and returns a dataframe with cluster number and gene class\n",
    "    This annotation file is a <species>_pangenome.csv file output from the pangenome analysis pipeline\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - pan_annot_path: str, path to the pangenome annotation file\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    - pan_df: pd.DataFrame, dataframe with cluster number and gene class\n",
    "    '''\n",
    "    with open(pan_annot_path, 'r') as f:\n",
    "        pan_df=pd.read_csv(f, index_col=0)\n",
    "        pan_df.drop(pan_df.columns[[1]], axis=1, inplace=True)\n",
    "        pan_df.columns=['Cluster', 'pan_gene_class']\n",
    "\n",
    "    pan_df.set_index('Cluster', inplace=True)\n",
    "    return pan_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pan_gene_class</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cluster 0</th>\n",
       "      <td>accessory</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1</th>\n",
       "      <td>accessory</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2</th>\n",
       "      <td>unique</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 3</th>\n",
       "      <td>unique</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4</th>\n",
       "      <td>core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8016</th>\n",
       "      <td>unique</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8017</th>\n",
       "      <td>unique</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8018</th>\n",
       "      <td>unique</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8019</th>\n",
       "      <td>unique</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8020</th>\n",
       "      <td>unique</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8021 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             pan_gene_class\n",
       "Cluster                    \n",
       "Cluster 0         accessory\n",
       "Cluster 1         accessory\n",
       "Cluster 2            unique\n",
       "Cluster 3            unique\n",
       "Cluster 4              core\n",
       "...                     ...\n",
       "Cluster 8016         unique\n",
       "Cluster 8017         unique\n",
       "Cluster 8018         unique\n",
       "Cluster 8019         unique\n",
       "Cluster 8020         unique\n",
       "\n",
       "[8021 rows x 1 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clstr_gene_class_df=get_cluster_pan_gene_class(pan_annot_path)\n",
    "clstr_gene_class_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=get_cluster_pan_gene_class(pan_annot_path)\n",
    "df.to_csv('../data/cluster_descriptions/cluster_pan_gene_class.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cluster_attributes(clstr_path, freq_path, pan_annot_path, clstr_fasta_path):\n",
    "\n",
    "\n",
    "    clstr_patric_id_df=get_cluster_representatives(clstr_path)\n",
    "    clstr_gene_class_df= get_cluster_pan_gene_class(pan_annot_path)\n",
    "    _patric_id_product_df=get_representative_products(clstr_fasta_path)\n",
    "    clstr_product_df=combine_cluster_product(clstr_patric_id_df, _patric_id_product_df)\n",
    "    clstr_freq_df=get_cluster_frequency(freq_path)\n",
    "\n",
    "    df=pd.concat([clstr_patric_id_df, clstr_gene_class_df, clstr_product_df, clstr_freq_df], axis=1)\n",
    "\n",
    "    #if there exist a Cluster col remove\n",
    "    if 'Cluster' in df.columns:\n",
    "        df.drop('Cluster', axis=1, inplace=True)\n",
    "    \n",
    "    return df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_representative</th>\n",
       "      <th>pan_gene_class</th>\n",
       "      <th>product_name</th>\n",
       "      <th>Frequency</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cluster 0</th>\n",
       "      <td>195.2029.peg.1780</td>\n",
       "      <td>accessory</td>\n",
       "      <td>helicase, Snf2 family</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1</th>\n",
       "      <td>195.2024.peg.1611</td>\n",
       "      <td>accessory</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2</th>\n",
       "      <td>195.2136.peg.1150</td>\n",
       "      <td>unique</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 3</th>\n",
       "      <td>195.2145.peg.407</td>\n",
       "      <td>unique</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4</th>\n",
       "      <td>195.2024.peg.685</td>\n",
       "      <td>core</td>\n",
       "      <td>DNA-directed RNA polymerase beta' subunit (EC ...</td>\n",
       "      <td>268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8016</th>\n",
       "      <td>195.2201.peg.1720</td>\n",
       "      <td>unique</td>\n",
       "      <td>Endonuclease I precursor (EC 3.1.21.1)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8017</th>\n",
       "      <td>195.2201.peg.1791</td>\n",
       "      <td>unique</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8018</th>\n",
       "      <td>195.2210.peg.964</td>\n",
       "      <td>unique</td>\n",
       "      <td>conserved hypothetical protein</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8019</th>\n",
       "      <td>195.2230.peg.1815</td>\n",
       "      <td>unique</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 8020</th>\n",
       "      <td>195.2314.peg.347</td>\n",
       "      <td>unique</td>\n",
       "      <td>Oxygen-insensitive NAD(P)H nitroreductase (EC ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8021 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_representative pan_gene_class  \\\n",
       "Cluster                                           \n",
       "Cluster 0      195.2029.peg.1780      accessory   \n",
       "Cluster 1      195.2024.peg.1611      accessory   \n",
       "Cluster 2      195.2136.peg.1150         unique   \n",
       "Cluster 3       195.2145.peg.407         unique   \n",
       "Cluster 4       195.2024.peg.685           core   \n",
       "...                          ...            ...   \n",
       "Cluster 8016   195.2201.peg.1720         unique   \n",
       "Cluster 8017   195.2201.peg.1791         unique   \n",
       "Cluster 8018    195.2210.peg.964         unique   \n",
       "Cluster 8019   195.2230.peg.1815         unique   \n",
       "Cluster 8020    195.2314.peg.347         unique   \n",
       "\n",
       "                                                   product_name  Frequency  \n",
       "Cluster                                                                     \n",
       "Cluster 0                                 helicase, Snf2 family        103  \n",
       "Cluster 1                                  hypothetical protein        239  \n",
       "Cluster 2                                  hypothetical protein          3  \n",
       "Cluster 3                                  hypothetical protein          1  \n",
       "Cluster 4     DNA-directed RNA polymerase beta' subunit (EC ...        268  \n",
       "...                                                         ...        ...  \n",
       "Cluster 8016             Endonuclease I precursor (EC 3.1.21.1)          1  \n",
       "Cluster 8017                               hypothetical protein          1  \n",
       "Cluster 8018                     conserved hypothetical protein          1  \n",
       "Cluster 8019                               hypothetical protein          1  \n",
       "Cluster 8020  Oxygen-insensitive NAD(P)H nitroreductase (EC ...          1  \n",
       "\n",
       "[8021 rows x 4 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_cluster_attributes(clstr_path, freq_path, pan_annot_path, clstr_fasta_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df=pd.read_csv(\"../data/cluster_descriptions/cluster_product.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8021"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(list(df[\"product_name\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict={}\n",
    "for i in list(df[\"product_name\"]):\n",
    "    dict[i]=0\n",
    "\n",
    "for i in list(df[\"product_name\"]):\n",
    "    dict[i]+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2860"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count=0\n",
    "for item in list(df['product_name']):\n",
    "    if 'hypothetical' in item:\n",
    "        count+=1\n",
    "count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "presence_df=pd.read_csv('../data/presence_matrices/Campylobacter_coli_presence_absence_T.csv',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#replace index clusters by their matching product_name from df\n",
    "my_dict=df.set_index('cluster').to_dict()['product_name']\n",
    "my_dict\n",
    "\n",
    "presence_df.rename(index=my_dict, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_node_attributes(G:nx.Graph, clstr_attribute_df):\n",
    "    '''\n",
    "    Set node attributes from a dataframe to a networkx graph\n",
    "\n",
    "    param:\n",
    "    ------\n",
    "    - G: nx.Graph, networkx graph\n",
    "    - clstr_attribute_df: pd.DataFrame, dataframe with cluster number and attributes\n",
    "\n",
    "    return:\n",
    "    -------\n",
    "    - G: nx.Graph, networkx graph with attributes set\n",
    "    '''\n",
    "    #make the df a dict of index clusters\n",
    "    my_dict=clstr_attribute_df.set_index('cluster').to_dict()\n",
    "\n",
    "    attribute=clstr_attribute_df.columns[1]\n",
    "    nx.set_node_attributes(G, my_dict, name=attribute)\n",
    "    return G"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
