{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/env/python\n",
    "\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "\n",
    "\n",
    "df = pd.read_csv('../test/roary/gene_presence_absence.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make the 1st column the row name\n",
    "df.set_index(df.columns[0], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene\n",
       "thrS          PBIEAMFA_00384\n",
       "rplC          PBIEAMFA_00193\n",
       "rpsM          PBIEAMFA_00217\n",
       "rplQ          PBIEAMFA_00220\n",
       "rpmA          PBIEAMFA_00428\n",
       "                   ...      \n",
       "group_9995               NaN\n",
       "group_9996               NaN\n",
       "group_9997               NaN\n",
       "group_9998               NaN\n",
       "group_9999               NaN\n",
       "Name: gordonii_USA_FDA, Length: 14944, dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check the 14th col\n",
    "df.iloc[:, 13]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "threshold = 0.5\n",
    "NO_ISOLATES = 31\n",
    "df = df[df['No. isolates'] > (threshold * NO_ISOLATES)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Non-unique Gene name</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>No. isolates</th>\n",
       "      <th>No. sequences</th>\n",
       "      <th>Avg sequences per isolate</th>\n",
       "      <th>Genome Fragment</th>\n",
       "      <th>Order within Fragment</th>\n",
       "      <th>Accessory Fragment</th>\n",
       "      <th>Accessory Order with Fragment</th>\n",
       "      <th>QC</th>\n",
       "      <th>...</th>\n",
       "      <th>oralis_USA_Baltimore</th>\n",
       "      <th>oralis_USA_Chicago</th>\n",
       "      <th>oralis_USA_FDA</th>\n",
       "      <th>oralis_Zealand_A16000295</th>\n",
       "      <th>oralis_Zealand_CF7_Ac3-15</th>\n",
       "      <th>pneumoniae_Lebanon_AUB_475</th>\n",
       "      <th>pneumoniae_Lebanon_AUB_521</th>\n",
       "      <th>pneumoniae_Lebanon_AUB_566</th>\n",
       "      <th>sanguinis_Texas</th>\n",
       "      <th>thermophilus_France</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>thrS</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Threonine--tRNA ligase</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30</td>\n",
       "      <td>4867</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>MKIIHIBE_01004</td>\n",
       "      <td>ICLACKFK_00729</td>\n",
       "      <td>CKAMCEFP_00392</td>\n",
       "      <td>AKDALEJM_00650</td>\n",
       "      <td>LAMCJNAC_01409</td>\n",
       "      <td>BGLEDEHL_00628</td>\n",
       "      <td>JOPHFHME_00627</td>\n",
       "      <td>ENNPKMBI_01540</td>\n",
       "      <td>LNMLLMIC_00398</td>\n",
       "      <td>GGJOBPCC_00616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rplC</th>\n",
       "      <td>NaN</td>\n",
       "      <td>50S ribosomal protein L3</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30</td>\n",
       "      <td>3549</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>MKIIHIBE_01292</td>\n",
       "      <td>ICLACKFK_01019</td>\n",
       "      <td>CKAMCEFP_00993</td>\n",
       "      <td>AKDALEJM_00308</td>\n",
       "      <td>LAMCJNAC_01022</td>\n",
       "      <td>BGLEDEHL_01997</td>\n",
       "      <td>JOPHFHME_01942</td>\n",
       "      <td>ENNPKMBI_00209</td>\n",
       "      <td>LNMLLMIC_00140</td>\n",
       "      <td>GGJOBPCC_01863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpsM</th>\n",
       "      <td>NaN</td>\n",
       "      <td>30S ribosomal protein S13</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30</td>\n",
       "      <td>3512</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>MKIIHIBE_01268</td>\n",
       "      <td>ICLACKFK_00994</td>\n",
       "      <td>CKAMCEFP_01017</td>\n",
       "      <td>AKDALEJM_00332</td>\n",
       "      <td>LAMCJNAC_00997</td>\n",
       "      <td>BGLEDEHL_01973</td>\n",
       "      <td>JOPHFHME_01918</td>\n",
       "      <td>ENNPKMBI_00233</td>\n",
       "      <td>LNMLLMIC_00163</td>\n",
       "      <td>GGJOBPCC_01839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rplQ</th>\n",
       "      <td>NaN</td>\n",
       "      <td>50S ribosomal protein L17</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30</td>\n",
       "      <td>3509</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>MKIIHIBE_01265</td>\n",
       "      <td>ICLACKFK_00991</td>\n",
       "      <td>CKAMCEFP_01020</td>\n",
       "      <td>AKDALEJM_00335</td>\n",
       "      <td>LAMCJNAC_00994</td>\n",
       "      <td>BGLEDEHL_01970</td>\n",
       "      <td>JOPHFHME_01915</td>\n",
       "      <td>ENNPKMBI_00236</td>\n",
       "      <td>LNMLLMIC_00166</td>\n",
       "      <td>GGJOBPCC_01836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpmA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>50S ribosomal protein L27</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30</td>\n",
       "      <td>13641</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>MKIIHIBE_00774</td>\n",
       "      <td>ICLACKFK_00509</td>\n",
       "      <td>CKAMCEFP_00155</td>\n",
       "      <td>AKDALEJM_00891</td>\n",
       "      <td>LAMCJNAC_00075</td>\n",
       "      <td>BGLEDEHL_01124</td>\n",
       "      <td>JOPHFHME_01069</td>\n",
       "      <td>ENNPKMBI_01088</td>\n",
       "      <td>LNMLLMIC_00883</td>\n",
       "      <td>GGJOBPCC_00466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_664</th>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30</td>\n",
       "      <td>14013</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2120.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>MKIIHIBE_00325</td>\n",
       "      <td>ICLACKFK_00074</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LAMCJNAC_01802</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_679</th>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30</td>\n",
       "      <td>11421</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13442.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>MKIIHIBE_01159</td>\n",
       "      <td>ICLACKFK_00890</td>\n",
       "      <td>CKAMCEFP_00553</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LAMCJNAC_01267</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_775</th>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30</td>\n",
       "      <td>9501</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7680.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>MKIIHIBE_00206</td>\n",
       "      <td>ICLACKFK_01311</td>\n",
       "      <td>CKAMCEFP_01394</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LAMCJNAC_00908</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_796</th>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30</td>\n",
       "      <td>10739</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12706.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>MKIIHIBE_01204</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LAMCJNAC_00516</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_999</th>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30</td>\n",
       "      <td>11143</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13018.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ICLACKFK_01372</td>\n",
       "      <td>CKAMCEFP_01136</td>\n",
       "      <td>AKDALEJM_01914</td>\n",
       "      <td>LAMCJNAC_01681</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1354 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Non-unique Gene name                 Annotation  No. isolates  \\\n",
       "Gene                                                                      \n",
       "thrS                       NaN     Threonine--tRNA ligase            31   \n",
       "rplC                       NaN   50S ribosomal protein L3            31   \n",
       "rpsM                       NaN  30S ribosomal protein S13            31   \n",
       "rplQ                       NaN  50S ribosomal protein L17            31   \n",
       "rpmA                       NaN  50S ribosomal protein L27            31   \n",
       "...                        ...                        ...           ...   \n",
       "group_664                  NaN       hypothetical protein            16   \n",
       "group_679                  NaN       hypothetical protein            16   \n",
       "group_775                  NaN       hypothetical protein            16   \n",
       "group_796                  NaN       hypothetical protein            16   \n",
       "group_999                  NaN       hypothetical protein            16   \n",
       "\n",
       "           No. sequences  Avg sequences per isolate  Genome Fragment  \\\n",
       "Gene                                                                   \n",
       "thrS                  31                        1.0               30   \n",
       "rplC                  31                        1.0               30   \n",
       "rpsM                  31                        1.0               30   \n",
       "rplQ                  31                        1.0               30   \n",
       "rpmA                  31                        1.0               30   \n",
       "...                  ...                        ...              ...   \n",
       "group_664             16                        1.0               30   \n",
       "group_679             16                        1.0               30   \n",
       "group_775             16                        1.0               30   \n",
       "group_796             16                        1.0               30   \n",
       "group_999             16                        1.0               30   \n",
       "\n",
       "           Order within Fragment  Accessory Fragment  \\\n",
       "Gene                                                   \n",
       "thrS                        4867                 NaN   \n",
       "rplC                        3549                 NaN   \n",
       "rpsM                        3512                 NaN   \n",
       "rplQ                        3509                 NaN   \n",
       "rpmA                       13641                 NaN   \n",
       "...                          ...                 ...   \n",
       "group_664                  14013                 1.0   \n",
       "group_679                  11421                 1.0   \n",
       "group_775                   9501                 1.0   \n",
       "group_796                  10739                 1.0   \n",
       "group_999                  11143                 1.0   \n",
       "\n",
       "           Accessory Order with Fragment   QC  ...  oralis_USA_Baltimore  \\\n",
       "Gene                                           ...                         \n",
       "thrS                                 NaN  NaN  ...        MKIIHIBE_01004   \n",
       "rplC                                 NaN  NaN  ...        MKIIHIBE_01292   \n",
       "rpsM                                 NaN  NaN  ...        MKIIHIBE_01268   \n",
       "rplQ                                 NaN  NaN  ...        MKIIHIBE_01265   \n",
       "rpmA                                 NaN  NaN  ...        MKIIHIBE_00774   \n",
       "...                                  ...  ...  ...                   ...   \n",
       "group_664                         2120.0  NaN  ...        MKIIHIBE_00325   \n",
       "group_679                        13442.0  NaN  ...        MKIIHIBE_01159   \n",
       "group_775                         7680.0  NaN  ...        MKIIHIBE_00206   \n",
       "group_796                        12706.0  NaN  ...        MKIIHIBE_01204   \n",
       "group_999                        13018.0  NaN  ...                   NaN   \n",
       "\n",
       "           oralis_USA_Chicago  oralis_USA_FDA oralis_Zealand_A16000295  \\\n",
       "Gene                                                                     \n",
       "thrS           ICLACKFK_00729  CKAMCEFP_00392           AKDALEJM_00650   \n",
       "rplC           ICLACKFK_01019  CKAMCEFP_00993           AKDALEJM_00308   \n",
       "rpsM           ICLACKFK_00994  CKAMCEFP_01017           AKDALEJM_00332   \n",
       "rplQ           ICLACKFK_00991  CKAMCEFP_01020           AKDALEJM_00335   \n",
       "rpmA           ICLACKFK_00509  CKAMCEFP_00155           AKDALEJM_00891   \n",
       "...                       ...             ...                      ...   \n",
       "group_664      ICLACKFK_00074             NaN                      NaN   \n",
       "group_679      ICLACKFK_00890  CKAMCEFP_00553                      NaN   \n",
       "group_775      ICLACKFK_01311  CKAMCEFP_01394                      NaN   \n",
       "group_796                 NaN             NaN                      NaN   \n",
       "group_999      ICLACKFK_01372  CKAMCEFP_01136           AKDALEJM_01914   \n",
       "\n",
       "          oralis_Zealand_CF7_Ac3-15 pneumoniae_Lebanon_AUB_475  \\\n",
       "Gene                                                             \n",
       "thrS                 LAMCJNAC_01409             BGLEDEHL_00628   \n",
       "rplC                 LAMCJNAC_01022             BGLEDEHL_01997   \n",
       "rpsM                 LAMCJNAC_00997             BGLEDEHL_01973   \n",
       "rplQ                 LAMCJNAC_00994             BGLEDEHL_01970   \n",
       "rpmA                 LAMCJNAC_00075             BGLEDEHL_01124   \n",
       "...                             ...                        ...   \n",
       "group_664            LAMCJNAC_01802                        NaN   \n",
       "group_679            LAMCJNAC_01267                        NaN   \n",
       "group_775            LAMCJNAC_00908                        NaN   \n",
       "group_796            LAMCJNAC_00516                        NaN   \n",
       "group_999            LAMCJNAC_01681                        NaN   \n",
       "\n",
       "          pneumoniae_Lebanon_AUB_521 pneumoniae_Lebanon_AUB_566  \\\n",
       "Gene                                                              \n",
       "thrS                  JOPHFHME_00627             ENNPKMBI_01540   \n",
       "rplC                  JOPHFHME_01942             ENNPKMBI_00209   \n",
       "rpsM                  JOPHFHME_01918             ENNPKMBI_00233   \n",
       "rplQ                  JOPHFHME_01915             ENNPKMBI_00236   \n",
       "rpmA                  JOPHFHME_01069             ENNPKMBI_01088   \n",
       "...                              ...                        ...   \n",
       "group_664                        NaN                        NaN   \n",
       "group_679                        NaN                        NaN   \n",
       "group_775                        NaN                        NaN   \n",
       "group_796                        NaN                        NaN   \n",
       "group_999                        NaN                        NaN   \n",
       "\n",
       "          sanguinis_Texas thermophilus_France  \n",
       "Gene                                           \n",
       "thrS       LNMLLMIC_00398      GGJOBPCC_00616  \n",
       "rplC       LNMLLMIC_00140      GGJOBPCC_01863  \n",
       "rpsM       LNMLLMIC_00163      GGJOBPCC_01839  \n",
       "rplQ       LNMLLMIC_00166      GGJOBPCC_01836  \n",
       "rpmA       LNMLLMIC_00883      GGJOBPCC_00466  \n",
       "...                   ...                 ...  \n",
       "group_664             NaN                 NaN  \n",
       "group_679             NaN                 NaN  \n",
       "group_775             NaN                 NaN  \n",
       "group_796             NaN                 NaN  \n",
       "group_999             NaN                 NaN  \n",
       "\n",
       "[1354 rows x 44 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# removing col 1 to 14\n",
    "df=df.drop(df.columns[0:13], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gordonii_USA_FDA</th>\n",
       "      <th>mitis_USA_NYC</th>\n",
       "      <th>oralis_Canada</th>\n",
       "      <th>oralis_China</th>\n",
       "      <th>oralis_ChinaCP018338</th>\n",
       "      <th>oralis_China_SOD</th>\n",
       "      <th>oralis_Denmark</th>\n",
       "      <th>oralis_Germany_DD25</th>\n",
       "      <th>oralis_Germany_SN51445</th>\n",
       "      <th>oralis_Germany_Uo5</th>\n",
       "      <th>...</th>\n",
       "      <th>oralis_USA_Baltimore</th>\n",
       "      <th>oralis_USA_Chicago</th>\n",
       "      <th>oralis_USA_FDA</th>\n",
       "      <th>oralis_Zealand_A16000295</th>\n",
       "      <th>oralis_Zealand_CF7_Ac3-15</th>\n",
       "      <th>pneumoniae_Lebanon_AUB_475</th>\n",
       "      <th>pneumoniae_Lebanon_AUB_521</th>\n",
       "      <th>pneumoniae_Lebanon_AUB_566</th>\n",
       "      <th>sanguinis_Texas</th>\n",
       "      <th>thermophilus_France</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>thrS</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rplC</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpsM</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rplQ</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpmA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_664</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_679</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_775</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_796</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_999</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1354 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gordonii_USA_FDA mitis_USA_NYC oralis_Canada oralis_China  \\\n",
       "Gene                                                                  \n",
       "thrS                     1             1             1            1   \n",
       "rplC                     1             1             1            1   \n",
       "rpsM                     1             1             1            1   \n",
       "rplQ                     1             1             1            1   \n",
       "rpmA                     1             1             1            1   \n",
       "...                    ...           ...           ...          ...   \n",
       "group_664                0             0             1            1   \n",
       "group_679                0             0             1            1   \n",
       "group_775                0             0             1            1   \n",
       "group_796                0             0             1            1   \n",
       "group_999                0             0             1            1   \n",
       "\n",
       "          oralis_ChinaCP018338 oralis_China_SOD oralis_Denmark  \\\n",
       "Gene                                                             \n",
       "thrS                         1                1              1   \n",
       "rplC                         1                1              1   \n",
       "rpsM                         1                1              1   \n",
       "rplQ                         1                1              1   \n",
       "rpmA                         1                1              1   \n",
       "...                        ...              ...            ...   \n",
       "group_664                    1                1              0   \n",
       "group_679                    1                0              0   \n",
       "group_775                    0                0              0   \n",
       "group_796                    0                1              1   \n",
       "group_999                    1                0              0   \n",
       "\n",
       "          oralis_Germany_DD25 oralis_Germany_SN51445 oralis_Germany_Uo5  ...  \\\n",
       "Gene                                                                     ...   \n",
       "thrS                        1                      1                  1  ...   \n",
       "rplC                        1                      1                  1  ...   \n",
       "rpsM                        1                      1                  1  ...   \n",
       "rplQ                        1                      1                  1  ...   \n",
       "rpmA                        1                      1                  1  ...   \n",
       "...                       ...                    ...                ...  ...   \n",
       "group_664                   0                      1                  0  ...   \n",
       "group_679                   0                      1                  0  ...   \n",
       "group_775                   1                      1                  0  ...   \n",
       "group_796                   0                      0                  1  ...   \n",
       "group_999                   0                      1                  1  ...   \n",
       "\n",
       "          oralis_USA_Baltimore oralis_USA_Chicago oralis_USA_FDA  \\\n",
       "Gene                                                               \n",
       "thrS                         1                  1              1   \n",
       "rplC                         1                  1              1   \n",
       "rpsM                         1                  1              1   \n",
       "rplQ                         1                  1              1   \n",
       "rpmA                         1                  1              1   \n",
       "...                        ...                ...            ...   \n",
       "group_664                    1                  1              0   \n",
       "group_679                    1                  1              1   \n",
       "group_775                    1                  1              1   \n",
       "group_796                    1                  0              0   \n",
       "group_999                    0                  1              1   \n",
       "\n",
       "          oralis_Zealand_A16000295 oralis_Zealand_CF7_Ac3-15  \\\n",
       "Gene                                                           \n",
       "thrS                             1                         1   \n",
       "rplC                             1                         1   \n",
       "rpsM                             1                         1   \n",
       "rplQ                             1                         1   \n",
       "rpmA                             1                         1   \n",
       "...                            ...                       ...   \n",
       "group_664                        0                         1   \n",
       "group_679                        0                         1   \n",
       "group_775                        0                         1   \n",
       "group_796                        0                         1   \n",
       "group_999                        1                         1   \n",
       "\n",
       "          pneumoniae_Lebanon_AUB_475 pneumoniae_Lebanon_AUB_521  \\\n",
       "Gene                                                              \n",
       "thrS                               1                          1   \n",
       "rplC                               1                          1   \n",
       "rpsM                               1                          1   \n",
       "rplQ                               1                          1   \n",
       "rpmA                               1                          1   \n",
       "...                              ...                        ...   \n",
       "group_664                          0                          0   \n",
       "group_679                          0                          0   \n",
       "group_775                          0                          0   \n",
       "group_796                          0                          0   \n",
       "group_999                          0                          0   \n",
       "\n",
       "          pneumoniae_Lebanon_AUB_566 sanguinis_Texas thermophilus_France  \n",
       "Gene                                                                      \n",
       "thrS                               1               1                   1  \n",
       "rplC                               1               1                   1  \n",
       "rpsM                               1               1                   1  \n",
       "rplQ                               1               1                   1  \n",
       "rpmA                               1               1                   1  \n",
       "...                              ...             ...                 ...  \n",
       "group_664                          0               0                   0  \n",
       "group_679                          0               0                   0  \n",
       "group_775                          0               0                   0  \n",
       "group_796                          0               0                   0  \n",
       "group_999                          0               0                   0  \n",
       "\n",
       "[1354 rows x 31 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#make the NaN 0s and the rest 1s\n",
    "df = df.fillna(0)\n",
    "\n",
    "df.where(df == 0, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gordonii_USA_FDA</th>\n",
       "      <th>mitis_USA_NYC</th>\n",
       "      <th>oralis_Canada</th>\n",
       "      <th>oralis_China</th>\n",
       "      <th>oralis_ChinaCP018338</th>\n",
       "      <th>oralis_China_SOD</th>\n",
       "      <th>oralis_Denmark</th>\n",
       "      <th>oralis_Germany_DD25</th>\n",
       "      <th>oralis_Germany_SN51445</th>\n",
       "      <th>oralis_Germany_Uo5</th>\n",
       "      <th>oralis_Korea</th>\n",
       "      <th>oralis_Korea_HP01</th>\n",
       "      <th>oralis_Lebanon_LAU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>thrS</th>\n",
       "      <td>PBIEAMFA_00384</td>\n",
       "      <td>LBCDGOKB_00404</td>\n",
       "      <td>ANGOOEGF_00911</td>\n",
       "      <td>IACFKHBD_00866</td>\n",
       "      <td>BAJMDCKA_00482</td>\n",
       "      <td>LFMCCKHJ_00425</td>\n",
       "      <td>NLMLKOPB_00785</td>\n",
       "      <td>PEFKPAAP_00942</td>\n",
       "      <td>LANMFOCN_00896</td>\n",
       "      <td>FGGKMPOH_01456</td>\n",
       "      <td>OOKGLKIK_01172</td>\n",
       "      <td>JGEGMBEE_01525</td>\n",
       "      <td>CGKMLFJB_01138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rplC</th>\n",
       "      <td>PBIEAMFA_00193</td>\n",
       "      <td>LBCDGOKB_01537</td>\n",
       "      <td>ANGOOEGF_01175</td>\n",
       "      <td>IACFKHBD_01663</td>\n",
       "      <td>BAJMDCKA_00194</td>\n",
       "      <td>LFMCCKHJ_01766</td>\n",
       "      <td>NLMLKOPB_01228</td>\n",
       "      <td>PEFKPAAP_01330</td>\n",
       "      <td>LANMFOCN_00582</td>\n",
       "      <td>FGGKMPOH_01732</td>\n",
       "      <td>OOKGLKIK_01800</td>\n",
       "      <td>JGEGMBEE_01783</td>\n",
       "      <td>CGKMLFJB_00890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpsM</th>\n",
       "      <td>PBIEAMFA_00217</td>\n",
       "      <td>LBCDGOKB_01513</td>\n",
       "      <td>ANGOOEGF_01151</td>\n",
       "      <td>IACFKHBD_01688</td>\n",
       "      <td>BAJMDCKA_00219</td>\n",
       "      <td>LFMCCKHJ_01742</td>\n",
       "      <td>NLMLKOPB_01204</td>\n",
       "      <td>PEFKPAAP_01306</td>\n",
       "      <td>LANMFOCN_00607</td>\n",
       "      <td>FGGKMPOH_01708</td>\n",
       "      <td>OOKGLKIK_01824</td>\n",
       "      <td>JGEGMBEE_01759</td>\n",
       "      <td>CGKMLFJB_00915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rplQ</th>\n",
       "      <td>PBIEAMFA_00220</td>\n",
       "      <td>LBCDGOKB_01510</td>\n",
       "      <td>ANGOOEGF_01148</td>\n",
       "      <td>IACFKHBD_01691</td>\n",
       "      <td>BAJMDCKA_00222</td>\n",
       "      <td>LFMCCKHJ_01739</td>\n",
       "      <td>NLMLKOPB_01201</td>\n",
       "      <td>PEFKPAAP_01303</td>\n",
       "      <td>LANMFOCN_00610</td>\n",
       "      <td>FGGKMPOH_01705</td>\n",
       "      <td>OOKGLKIK_01827</td>\n",
       "      <td>JGEGMBEE_01756</td>\n",
       "      <td>CGKMLFJB_00918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpmA</th>\n",
       "      <td>PBIEAMFA_00428</td>\n",
       "      <td>LBCDGOKB_00629</td>\n",
       "      <td>ANGOOEGF_00076</td>\n",
       "      <td>IACFKHBD_00075</td>\n",
       "      <td>BAJMDCKA_01186</td>\n",
       "      <td>LFMCCKHJ_01263</td>\n",
       "      <td>NLMLKOPB_01588</td>\n",
       "      <td>PEFKPAAP_00723</td>\n",
       "      <td>LANMFOCN_01182</td>\n",
       "      <td>FGGKMPOH_01205</td>\n",
       "      <td>OOKGLKIK_00939</td>\n",
       "      <td>JGEGMBEE_01281</td>\n",
       "      <td>CGKMLFJB_00170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_664</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ANGOOEGF_00441</td>\n",
       "      <td>IACFKHBD_01801</td>\n",
       "      <td>BAJMDCKA_00672</td>\n",
       "      <td>LFMCCKHJ_00719</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>LANMFOCN_01639</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>JGEGMBEE_00787</td>\n",
       "      <td>CGKMLFJB_00635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_679</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ANGOOEGF_01061</td>\n",
       "      <td>IACFKHBD_00724</td>\n",
       "      <td>BAJMDCKA_00327</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>LANMFOCN_00745</td>\n",
       "      <td>0</td>\n",
       "      <td>OOKGLKIK_01338</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_775</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ANGOOEGF_00567</td>\n",
       "      <td>IACFKHBD_01269</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>PEFKPAAP_00539</td>\n",
       "      <td>LANMFOCN_01768</td>\n",
       "      <td>0</td>\n",
       "      <td>OOKGLKIK_00340</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_796</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ANGOOEGF_01102</td>\n",
       "      <td>IACFKHBD_00512</td>\n",
       "      <td>0</td>\n",
       "      <td>LFMCCKHJ_00248</td>\n",
       "      <td>NLMLKOPB_01452</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>FGGKMPOH_01652</td>\n",
       "      <td>OOKGLKIK_00019</td>\n",
       "      <td>JGEGMBEE_01701</td>\n",
       "      <td>CGKMLFJB_00019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_999</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ANGOOEGF_01431</td>\n",
       "      <td>IACFKHBD_01426</td>\n",
       "      <td>BAJMDCKA_01572</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>LANMFOCN_00013</td>\n",
       "      <td>FGGKMPOH_00349</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1354 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gordonii_USA_FDA   mitis_USA_NYC   oralis_Canada    oralis_China  \\\n",
       "Gene                                                                         \n",
       "thrS        PBIEAMFA_00384  LBCDGOKB_00404  ANGOOEGF_00911  IACFKHBD_00866   \n",
       "rplC        PBIEAMFA_00193  LBCDGOKB_01537  ANGOOEGF_01175  IACFKHBD_01663   \n",
       "rpsM        PBIEAMFA_00217  LBCDGOKB_01513  ANGOOEGF_01151  IACFKHBD_01688   \n",
       "rplQ        PBIEAMFA_00220  LBCDGOKB_01510  ANGOOEGF_01148  IACFKHBD_01691   \n",
       "rpmA        PBIEAMFA_00428  LBCDGOKB_00629  ANGOOEGF_00076  IACFKHBD_00075   \n",
       "...                    ...             ...             ...             ...   \n",
       "group_664                0               0  ANGOOEGF_00441  IACFKHBD_01801   \n",
       "group_679                0               0  ANGOOEGF_01061  IACFKHBD_00724   \n",
       "group_775                0               0  ANGOOEGF_00567  IACFKHBD_01269   \n",
       "group_796                0               0  ANGOOEGF_01102  IACFKHBD_00512   \n",
       "group_999                0               0  ANGOOEGF_01431  IACFKHBD_01426   \n",
       "\n",
       "          oralis_ChinaCP018338 oralis_China_SOD  oralis_Denmark  \\\n",
       "Gene                                                              \n",
       "thrS            BAJMDCKA_00482   LFMCCKHJ_00425  NLMLKOPB_00785   \n",
       "rplC            BAJMDCKA_00194   LFMCCKHJ_01766  NLMLKOPB_01228   \n",
       "rpsM            BAJMDCKA_00219   LFMCCKHJ_01742  NLMLKOPB_01204   \n",
       "rplQ            BAJMDCKA_00222   LFMCCKHJ_01739  NLMLKOPB_01201   \n",
       "rpmA            BAJMDCKA_01186   LFMCCKHJ_01263  NLMLKOPB_01588   \n",
       "...                        ...              ...             ...   \n",
       "group_664       BAJMDCKA_00672   LFMCCKHJ_00719               0   \n",
       "group_679       BAJMDCKA_00327                0               0   \n",
       "group_775                    0                0               0   \n",
       "group_796                    0   LFMCCKHJ_00248  NLMLKOPB_01452   \n",
       "group_999       BAJMDCKA_01572                0               0   \n",
       "\n",
       "          oralis_Germany_DD25 oralis_Germany_SN51445 oralis_Germany_Uo5  \\\n",
       "Gene                                                                      \n",
       "thrS           PEFKPAAP_00942         LANMFOCN_00896     FGGKMPOH_01456   \n",
       "rplC           PEFKPAAP_01330         LANMFOCN_00582     FGGKMPOH_01732   \n",
       "rpsM           PEFKPAAP_01306         LANMFOCN_00607     FGGKMPOH_01708   \n",
       "rplQ           PEFKPAAP_01303         LANMFOCN_00610     FGGKMPOH_01705   \n",
       "rpmA           PEFKPAAP_00723         LANMFOCN_01182     FGGKMPOH_01205   \n",
       "...                       ...                    ...                ...   \n",
       "group_664                   0         LANMFOCN_01639                  0   \n",
       "group_679                   0         LANMFOCN_00745                  0   \n",
       "group_775      PEFKPAAP_00539         LANMFOCN_01768                  0   \n",
       "group_796                   0                      0     FGGKMPOH_01652   \n",
       "group_999                   0         LANMFOCN_00013     FGGKMPOH_00349   \n",
       "\n",
       "             oralis_Korea oralis_Korea_HP01 oralis_Lebanon_LAU  \n",
       "Gene                                                            \n",
       "thrS       OOKGLKIK_01172    JGEGMBEE_01525     CGKMLFJB_01138  \n",
       "rplC       OOKGLKIK_01800    JGEGMBEE_01783     CGKMLFJB_00890  \n",
       "rpsM       OOKGLKIK_01824    JGEGMBEE_01759     CGKMLFJB_00915  \n",
       "rplQ       OOKGLKIK_01827    JGEGMBEE_01756     CGKMLFJB_00918  \n",
       "rpmA       OOKGLKIK_00939    JGEGMBEE_01281     CGKMLFJB_00170  \n",
       "...                   ...               ...                ...  \n",
       "group_664               0    JGEGMBEE_00787     CGKMLFJB_00635  \n",
       "group_679  OOKGLKIK_01338                 0                  0  \n",
       "group_775  OOKGLKIK_00340                 0                  0  \n",
       "group_796  OOKGLKIK_00019    JGEGMBEE_01701     CGKMLFJB_00019  \n",
       "group_999               0                 0                  0  \n",
       "\n",
       "[1354 rows x 13 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_df = df.iloc[:, 0:13]\n",
    "metadata_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gordonii_USA_FDA</th>\n",
       "      <th>mitis_USA_NYC</th>\n",
       "      <th>oralis_Canada</th>\n",
       "      <th>oralis_China</th>\n",
       "      <th>oralis_ChinaCP018338</th>\n",
       "      <th>oralis_China_SOD</th>\n",
       "      <th>oralis_Denmark</th>\n",
       "      <th>oralis_Germany_DD25</th>\n",
       "      <th>oralis_Germany_SN51445</th>\n",
       "      <th>oralis_Germany_Uo5</th>\n",
       "      <th>...</th>\n",
       "      <th>oralis_USA_Baltimore</th>\n",
       "      <th>oralis_USA_Chicago</th>\n",
       "      <th>oralis_USA_FDA</th>\n",
       "      <th>oralis_Zealand_A16000295</th>\n",
       "      <th>oralis_Zealand_CF7_Ac3-15</th>\n",
       "      <th>pneumoniae_Lebanon_AUB_475</th>\n",
       "      <th>pneumoniae_Lebanon_AUB_521</th>\n",
       "      <th>pneumoniae_Lebanon_AUB_566</th>\n",
       "      <th>sanguinis_Texas</th>\n",
       "      <th>thermophilus_France</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>thrS</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rplC</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpsM</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rplQ</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpmA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_664</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_679</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_775</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_796</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_999</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1354 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gordonii_USA_FDA mitis_USA_NYC oralis_Canada oralis_China  \\\n",
       "Gene                                                                  \n",
       "thrS                     1             1             1            1   \n",
       "rplC                     1             1             1            1   \n",
       "rpsM                     1             1             1            1   \n",
       "rplQ                     1             1             1            1   \n",
       "rpmA                     1             1             1            1   \n",
       "...                    ...           ...           ...          ...   \n",
       "group_664                0             0             1            1   \n",
       "group_679                0             0             1            1   \n",
       "group_775                0             0             1            1   \n",
       "group_796                0             0             1            1   \n",
       "group_999                0             0             1            1   \n",
       "\n",
       "          oralis_ChinaCP018338 oralis_China_SOD oralis_Denmark  \\\n",
       "Gene                                                             \n",
       "thrS                         1                1              1   \n",
       "rplC                         1                1              1   \n",
       "rpsM                         1                1              1   \n",
       "rplQ                         1                1              1   \n",
       "rpmA                         1                1              1   \n",
       "...                        ...              ...            ...   \n",
       "group_664                    1                1              0   \n",
       "group_679                    1                0              0   \n",
       "group_775                    0                0              0   \n",
       "group_796                    0                1              1   \n",
       "group_999                    1                0              0   \n",
       "\n",
       "          oralis_Germany_DD25 oralis_Germany_SN51445 oralis_Germany_Uo5  ...  \\\n",
       "Gene                                                                     ...   \n",
       "thrS                        1                      1                  1  ...   \n",
       "rplC                        1                      1                  1  ...   \n",
       "rpsM                        1                      1                  1  ...   \n",
       "rplQ                        1                      1                  1  ...   \n",
       "rpmA                        1                      1                  1  ...   \n",
       "...                       ...                    ...                ...  ...   \n",
       "group_664                   0                      1                  0  ...   \n",
       "group_679                   0                      1                  0  ...   \n",
       "group_775                   1                      1                  0  ...   \n",
       "group_796                   0                      0                  1  ...   \n",
       "group_999                   0                      1                  1  ...   \n",
       "\n",
       "          oralis_USA_Baltimore oralis_USA_Chicago oralis_USA_FDA  \\\n",
       "Gene                                                               \n",
       "thrS                         1                  1              1   \n",
       "rplC                         1                  1              1   \n",
       "rpsM                         1                  1              1   \n",
       "rplQ                         1                  1              1   \n",
       "rpmA                         1                  1              1   \n",
       "...                        ...                ...            ...   \n",
       "group_664                    1                  1              0   \n",
       "group_679                    1                  1              1   \n",
       "group_775                    1                  1              1   \n",
       "group_796                    1                  0              0   \n",
       "group_999                    0                  1              1   \n",
       "\n",
       "          oralis_Zealand_A16000295 oralis_Zealand_CF7_Ac3-15  \\\n",
       "Gene                                                           \n",
       "thrS                             1                         1   \n",
       "rplC                             1                         1   \n",
       "rpsM                             1                         1   \n",
       "rplQ                             1                         1   \n",
       "rpmA                             1                         1   \n",
       "...                            ...                       ...   \n",
       "group_664                        0                         1   \n",
       "group_679                        0                         1   \n",
       "group_775                        0                         1   \n",
       "group_796                        0                         1   \n",
       "group_999                        1                         1   \n",
       "\n",
       "          pneumoniae_Lebanon_AUB_475 pneumoniae_Lebanon_AUB_521  \\\n",
       "Gene                                                              \n",
       "thrS                               1                          1   \n",
       "rplC                               1                          1   \n",
       "rpsM                               1                          1   \n",
       "rplQ                               1                          1   \n",
       "rpmA                               1                          1   \n",
       "...                              ...                        ...   \n",
       "group_664                          0                          0   \n",
       "group_679                          0                          0   \n",
       "group_775                          0                          0   \n",
       "group_796                          0                          0   \n",
       "group_999                          0                          0   \n",
       "\n",
       "          pneumoniae_Lebanon_AUB_566 sanguinis_Texas thermophilus_France  \n",
       "Gene                                                                      \n",
       "thrS                               1               1                   1  \n",
       "rplC                               1               1                   1  \n",
       "rpsM                               1               1                   1  \n",
       "rplQ                               1               1                   1  \n",
       "rpmA                               1               1                   1  \n",
       "...                              ...             ...                 ...  \n",
       "group_664                          0               0                   0  \n",
       "group_679                          0               0                   0  \n",
       "group_775                          0               0                   0  \n",
       "group_796                          0               0                   0  \n",
       "group_999                          0               0                   0  \n",
       "\n",
       "[1354 rows x 31 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#!/bin/env/python\n",
    "\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "from typing import Tuple\n",
    "import os\n",
    "\n",
    "def read_df(file_path: str) -> pd.DataFrame:\n",
    "    '''this function reads a csv file and returns a dataframe\n",
    "    input:\n",
    "        - file_path: (str) path to the csv file\n",
    "    output:\n",
    "        - df: (pd.DataFrame) dataframe read from the csv file\n",
    "    '''\n",
    "\n",
    "    if not os.path.exists(file_path):\n",
    "        print(f'file {file_path} does not exist')\n",
    "        return None\n",
    "    df = pd.read_csv(file_path)\n",
    "    return df\n",
    "\n",
    "def collect_data(df: pd.DataFrame, sample_col: int=13, threshold: float = 0) -> Tuple[pd.DataFrame, pd.DataFrame]:\n",
    "    '''this function takes a raw dataframe and returns only the genes as rows and samples as columns in a binary matrix\n",
    "    - saves all the info in a seperate metadata dataframe\n",
    "    - removes all the genes that are not present in mroe than a threshhold percentage of samples\n",
    "    - returns the binary matrix and the metadata dataframe\n",
    "\n",
    "    input:\n",
    "        - df: (pd.DataFrame) raw dataframe read from csv\n",
    "        - sample_col: (int) the col number whre samples start\n",
    "        - threshold: (float) the percentage of samples a gene needs to be present in to be kept\n",
    "    output:\n",
    "        - df: (pd.DataFrame) binary matrix of genes and samples\n",
    "        - metadata_df: (pd.DataFrame) metadata dataframe with all the info from the raw dataframe\n",
    "    \n",
    "    '''\n",
    "    df.set_index(df.columns[0], inplace=True)\n",
    "    SAMPLE_NO = df.shape[1] - sample_col\n",
    "\n",
    "    df = df[df['No. isolates'] > (threshold * SAMPLE_NO)]\n",
    "\n",
    "    metadata_df = df.iloc[:, 0:sample_col]\n",
    "    df=df.drop(df.columns[0:sample_col], axis=1)\n",
    "    df = df.fillna(0)\n",
    "    df=df.where(df == 0, 1)\n",
    "\n",
    "    return df, metadata_df\n",
    "\n",
    "\n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "test_df=read_df('../test/roary/gene_presence_absence.csv')\n",
    "test_df, test_metadata_df = collect_data(test_df, threshold=0.5)\n",
    "test_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df= test_df\n",
    "adj_matrix = df.dot(df.T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['thrS', 'rplC', 'rpsM', 'rplQ', 'rpmA', 'rpsI', 'rpsL', 'rpmH', 'rpmF',\n",
       "       'rpsJ',\n",
       "       ...\n",
       "       'group_575', 'group_5776', 'smc_2', 'aaaT', 'group_660', 'group_664',\n",
       "       'group_679', 'group_775', 'group_796', 'group_999'],\n",
       "      dtype='object', name='Gene', length=1354)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adj_matrix.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene</th>\n",
       "      <th>thrS</th>\n",
       "      <th>rplC</th>\n",
       "      <th>rpsM</th>\n",
       "      <th>rplQ</th>\n",
       "      <th>rpmA</th>\n",
       "      <th>rpsI</th>\n",
       "      <th>rpsL</th>\n",
       "      <th>rpmH</th>\n",
       "      <th>rpmF</th>\n",
       "      <th>rpsJ</th>\n",
       "      <th>...</th>\n",
       "      <th>group_575</th>\n",
       "      <th>group_5776</th>\n",
       "      <th>smc_2</th>\n",
       "      <th>aaaT</th>\n",
       "      <th>group_660</th>\n",
       "      <th>group_664</th>\n",
       "      <th>group_679</th>\n",
       "      <th>group_775</th>\n",
       "      <th>group_796</th>\n",
       "      <th>group_999</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>thrS</th>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rplC</th>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpsM</th>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rplQ</th>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpmA</th>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_664</th>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_679</th>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>16</td>\n",
       "      <td>14</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_775</th>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "      <td>16</td>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_796</th>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_999</th>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1354 rows × 1354 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene      thrS rplC rpsM rplQ rpmA rpsI rpsL rpmH rpmF rpsJ  ... group_575  \\\n",
       "Gene                                                         ...             \n",
       "thrS        31   31   31   31   31   31   31   31   31   31  ...        16   \n",
       "rplC        31   31   31   31   31   31   31   31   31   31  ...        16   \n",
       "rpsM        31   31   31   31   31   31   31   31   31   31  ...        16   \n",
       "rplQ        31   31   31   31   31   31   31   31   31   31  ...        16   \n",
       "rpmA        31   31   31   31   31   31   31   31   31   31  ...        16   \n",
       "...        ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...       ...   \n",
       "group_664   16   16   16   16   16   16   16   16   16   16  ...        11   \n",
       "group_679   16   16   16   16   16   16   16   16   16   16  ...        14   \n",
       "group_775   16   16   16   16   16   16   16   16   16   16  ...        14   \n",
       "group_796   16   16   16   16   16   16   16   16   16   16  ...        11   \n",
       "group_999   16   16   16   16   16   16   16   16   16   16  ...        13   \n",
       "\n",
       "Gene      group_5776 smc_2 aaaT group_660 group_664 group_679 group_775  \\\n",
       "Gene                                                                      \n",
       "thrS              16    16   16        16        16        16        16   \n",
       "rplC              16    16   16        16        16        16        16   \n",
       "rpsM              16    16   16        16        16        16        16   \n",
       "rplQ              16    16   16        16        16        16        16   \n",
       "rpmA              16    16   16        16        16        16        16   \n",
       "...              ...   ...  ...       ...       ...       ...       ...   \n",
       "group_664         10    10    9        10        16        12        12   \n",
       "group_679         14    11   11        13        12        16        14   \n",
       "group_775         14    12   11        13        12        14        16   \n",
       "group_796         11     9    8        10        12        11        10   \n",
       "group_999         12    11   11        13        12        13        13   \n",
       "\n",
       "Gene      group_796 group_999  \n",
       "Gene                           \n",
       "thrS             16        16  \n",
       "rplC             16        16  \n",
       "rpsM             16        16  \n",
       "rplQ             16        16  \n",
       "rpmA             16        16  \n",
       "...             ...       ...  \n",
       "group_664        12        12  \n",
       "group_679        11        13  \n",
       "group_775        10        13  \n",
       "group_796        16         9  \n",
       "group_999         9        16  \n",
       "\n",
       "[1354 rows x 1354 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import networkx as nx\n",
    "#make a [][] matrix out of adj_matrix\n",
    "# adj_matrix = adj_matrix.values\n",
    "adj_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make dataype int\n",
    "adj_matrix = adj_matrix.astype(int)\n",
    "#cretae a graph out of it\n",
    "G = nx.from_pandas_adjacency(adj_matrix)\n",
    "G.remove_edges_from(nx.selfloop_edges(G))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'scipy'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[22], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m#visualize G\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mmatplotlib\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpyplot\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mplt\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m \u001b[43mnx\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdraw\u001b[49m\u001b[43m(\u001b[49m\u001b[43mG\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mwith_labels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.12/site-packages/networkx/drawing/nx_pylab.py:121\u001b[0m, in \u001b[0;36mdraw\u001b[0;34m(G, pos, ax, **kwds)\u001b[0m\n\u001b[1;32m    118\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwith_labels\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m kwds:\n\u001b[1;32m    119\u001b[0m     kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwith_labels\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlabels\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m kwds\n\u001b[0;32m--> 121\u001b[0m \u001b[43mdraw_networkx\u001b[49m\u001b[43m(\u001b[49m\u001b[43mG\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpos\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpos\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43max\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43max\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    122\u001b[0m ax\u001b[38;5;241m.\u001b[39mset_axis_off()\n\u001b[1;32m    123\u001b[0m plt\u001b[38;5;241m.\u001b[39mdraw_if_interactive()\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.12/site-packages/networkx/drawing/nx_pylab.py:302\u001b[0m, in \u001b[0;36mdraw_networkx\u001b[0;34m(G, pos, arrows, with_labels, **kwds)\u001b[0m\n\u001b[1;32m    299\u001b[0m label_kwds \u001b[38;5;241m=\u001b[39m {k: v \u001b[38;5;28;01mfor\u001b[39;00m k, v \u001b[38;5;129;01min\u001b[39;00m kwds\u001b[38;5;241m.\u001b[39mitems() \u001b[38;5;28;01mif\u001b[39;00m k \u001b[38;5;129;01min\u001b[39;00m valid_label_kwds}\n\u001b[1;32m    301\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m pos \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 302\u001b[0m     pos \u001b[38;5;241m=\u001b[39m \u001b[43mnx\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdrawing\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mspring_layout\u001b[49m\u001b[43m(\u001b[49m\u001b[43mG\u001b[49m\u001b[43m)\u001b[49m  \u001b[38;5;66;03m# default to spring layout\u001b[39;00m\n\u001b[1;32m    304\u001b[0m draw_networkx_nodes(G, pos, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mnode_kwds)\n\u001b[1;32m    305\u001b[0m draw_networkx_edges(G, pos, arrows\u001b[38;5;241m=\u001b[39marrows, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39medge_kwds)\n",
      "File \u001b[0;32m<class 'networkx.utils.decorators.argmap'> compilation 8:4\u001b[0m, in \u001b[0;36margmap_spring_layout_5\u001b[0;34m(G, k, pos, fixed, iterations, threshold, weight, scale, center, dim, seed)\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mcollections\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mgzip\u001b[39;00m\n\u001b[0;32m----> 4\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01minspect\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mitertools\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mre\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.12/site-packages/networkx/drawing/layout.py:477\u001b[0m, in \u001b[0;36mspring_layout\u001b[0;34m(G, k, pos, fixed, iterations, threshold, weight, scale, center, dim, seed)\u001b[0m\n\u001b[1;32m    475\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(G) \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m500\u001b[39m:  \u001b[38;5;66;03m# sparse solver for large graphs\u001b[39;00m\n\u001b[1;32m    476\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m\n\u001b[0;32m--> 477\u001b[0m A \u001b[38;5;241m=\u001b[39m \u001b[43mnx\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_scipy_sparse_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43mG\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mweight\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mweight\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    478\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m k \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m fixed \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    479\u001b[0m     \u001b[38;5;66;03m# We must adjust k by domain size for layouts not near 1x1\u001b[39;00m\n\u001b[1;32m    480\u001b[0m     nnodes, _ \u001b[38;5;241m=\u001b[39m A\u001b[38;5;241m.\u001b[39mshape\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.12/site-packages/networkx/utils/backends.py:412\u001b[0m, in \u001b[0;36m_dispatch.__call__\u001b[0;34m(self, backend, *args, **kwargs)\u001b[0m\n\u001b[1;32m    409\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__call__\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m/\u001b[39m, \u001b[38;5;241m*\u001b[39margs, backend\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    410\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m backends:\n\u001b[1;32m    411\u001b[0m         \u001b[38;5;66;03m# Fast path if no backends are installed\u001b[39;00m\n\u001b[0;32m--> 412\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43morig_func\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    414\u001b[0m     \u001b[38;5;66;03m# Use `backend_name` in this function instead of `backend`\u001b[39;00m\n\u001b[1;32m    415\u001b[0m     backend_name \u001b[38;5;241m=\u001b[39m backend\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.12/site-packages/networkx/convert_matrix.py:567\u001b[0m, in \u001b[0;36mto_scipy_sparse_array\u001b[0;34m(G, nodelist, dtype, weight, format)\u001b[0m\n\u001b[1;32m    486\u001b[0m \u001b[38;5;129m@nx\u001b[39m\u001b[38;5;241m.\u001b[39m_dispatch(edge_attrs\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mweight\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    487\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mto_scipy_sparse_array\u001b[39m(G, nodelist\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, dtype\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, weight\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mweight\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28mformat\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcsr\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m    488\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Returns the graph adjacency matrix as a SciPy sparse array.\u001b[39;00m\n\u001b[1;32m    489\u001b[0m \n\u001b[1;32m    490\u001b[0m \u001b[38;5;124;03m    Parameters\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    565\u001b[0m \u001b[38;5;124;03m       https://docs.scipy.org/doc/scipy/reference/sparse.html\u001b[39;00m\n\u001b[1;32m    566\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 567\u001b[0m     \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mscipy\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01msp\u001b[39;00m\n\u001b[1;32m    569\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(G) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    570\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m nx\u001b[38;5;241m.\u001b[39mNetworkXError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mGraph has no nodes or edges\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'scipy'"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAr4AAAIRCAYAAACszb5OAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/H5lhTAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAf1klEQVR4nO3df2zX1b348VehtFXvbRdhVhDsyq5ubGTu0gZGuWSZV2vQuJDsRhYXUa8ma7ZdhF69g3Gjg5g0283MnZvgNkGzBL3En/GPXkf/uBdR2L0XblmWQeIiXAtbKynGFnW3CLy/f/il32/XgnxKC7LX45F8/ujxnM/nfHaGe+7tu2/LiqIoAgAA/sRNON8bAACAc0H4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJBCyeH78ssvx8033xzTpk2LsrKyeOGFFz50zdatW6OhoSGqqqpi5syZ8eijj45mrwAAMGolh++7774b11xzTfz4xz8+o/n79++PG2+8MRYuXBidnZ3xne98J5YtWxbPPvtsyZsFAIDRKiuKohj14rKyeP7552Px4sWnnPPtb387Xnzxxdi7d+/gWEtLS/zqV7+KHTt2jPajAQCgJOXj/QE7duyI5ubmIWM33HBDbNiwId5///2YNGnSsDUDAwMxMDAw+POJEyfirbfeismTJ0dZWdl4bxkAgPOsKIo4cuRITJs2LSZMGJtfSxv38O3p6Yna2tohY7W1tXHs2LHo7e2NqVOnDlvT1tYWa9asGe+tAQDwEXfgwIGYPn36mLzXuIdvRAy7Snvy7opTXb1dtWpVtLa2Dv7c19cXV155ZRw4cCCqq6vHb6MAAHwk9Pf3x4wZM+LP//zPx+w9xz18L7/88ujp6RkydujQoSgvL4/JkyePuKaysjIqKyuHjVdXVwtfAIBExvI213F/ju/8+fOjo6NjyNiWLVuisbFxxPt7AQBgPJQcvu+8807s3r07du/eHREfPK5s9+7d0dXVFREf3KawdOnSwfktLS3xxhtvRGtra+zduzc2btwYGzZsiHvvvXdsvgEAAJyBkm912LlzZ3zpS18a/Pnkvbi33357PPHEE9Hd3T0YwRER9fX10d7eHitWrIhHHnkkpk2bFg8//HB85StfGYPtAwDAmTmr5/ieK/39/VFTUxN9fX3u8QUASGA8+m/c7/EFAICPAuELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJCC8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJCC8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJCC8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIYVfiuW7cu6uvro6qqKhoaGmLbtm2nnb9p06a45ppr4uKLL46pU6fGnXfeGYcPHx7VhgEAYDRKDt/NmzfH8uXLY/Xq1dHZ2RkLFy6MRYsWRVdX14jzX3nllVi6dGncdddd8Zvf/Caefvrp+K//+q+4++67z3rzAABwpkoO34ceeijuuuuuuPvuu2PWrFnxz//8zzFjxoxYv379iPN/+ctfxic+8YlYtmxZ1NfXx1/91V/F17/+9di5c+dZbx4AAM5USeF79OjR2LVrVzQ3Nw8Zb25uju3bt4+4pqmpKQ4ePBjt7e1RFEW8+eab8cwzz8RNN910ys8ZGBiI/v7+IS8AADgbJYVvb29vHD9+PGpra4eM19bWRk9Pz4hrmpqaYtOmTbFkyZKoqKiIyy+/PD72sY/Fj370o1N+TltbW9TU1Ay+ZsyYUco2AQBgmFH9cltZWdmQn4uiGDZ20p49e2LZsmVx//33x65du+Kll16K/fv3R0tLyynff9WqVdHX1zf4OnDgwGi2CQAAg8pLmTxlypSYOHHisKu7hw4dGnYV+KS2trZYsGBB3HfffRER8bnPfS4uueSSWLhwYTz44IMxderUYWsqKyujsrKylK0BAMBplXTFt6KiIhoaGqKjo2PIeEdHRzQ1NY245r333osJE4Z+zMSJEyPigyvFAABwLpR8q0Nra2s89thjsXHjxti7d2+sWLEiurq6Bm9dWLVqVSxdunRw/s033xzPPfdcrF+/Pvbt2xevvvpqLFu2LObOnRvTpk0bu28CAACnUdKtDhERS5YsicOHD8fatWuju7s7Zs+eHe3t7VFXVxcREd3d3UOe6XvHHXfEkSNH4sc//nH8/d//fXzsYx+La6+9Nr73ve+N3bcAAIAPUVZcAPcb9Pf3R01NTfT19UV1dfX53g4AAONsPPpvVE91AACAC43wBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJCC8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJCC8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJCC8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApjCp8161bF/X19VFVVRUNDQ2xbdu2084fGBiI1atXR11dXVRWVsYnP/nJ2Lhx46g2DAAAo1Fe6oLNmzfH8uXLY926dbFgwYL4yU9+EosWLYo9e/bElVdeOeKaW265Jd58883YsGFD/MVf/EUcOnQojh07dtabBwCAM1VWFEVRyoJ58+bFnDlzYv369YNjs2bNisWLF0dbW9uw+S+99FJ89atfjX379sWll146qk329/dHTU1N9PX1RXV19ajeAwCAC8d49F9JtzocPXo0du3aFc3NzUPGm5ubY/v27SOuefHFF6OxsTG+//3vxxVXXBFXX3113HvvvfGHP/zhlJ8zMDAQ/f39Q14AAHA2SrrVobe3N44fPx61tbVDxmtra6Onp2fENfv27YtXXnklqqqq4vnnn4/e3t74xje+EW+99dYp7/Nta2uLNWvWlLI1AAA4rVH9cltZWdmQn4uiGDZ20okTJ6KsrCw2bdoUc+fOjRtvvDEeeuiheOKJJ0551XfVqlXR19c3+Dpw4MBotgkAAINKuuI7ZcqUmDhx4rCru4cOHRp2FfikqVOnxhVXXBE1NTWDY7NmzYqiKOLgwYNx1VVXDVtTWVkZlZWVpWwNAABOq6QrvhUVFdHQ0BAdHR1Dxjs6OqKpqWnENQsWLIjf//738c477wyOvfbaazFhwoSYPn36KLYMAAClK/lWh9bW1njsscdi48aNsXfv3lixYkV0dXVFS0tLRHxwm8LSpUsH5996660xefLkuPPOO2PPnj3x8ssvx3333Rd/+7d/GxdddNHYfRMAADiNkp/ju2TJkjh8+HCsXbs2uru7Y/bs2dHe3h51dXUREdHd3R1dXV2D8//sz/4sOjo64u/+7u+isbExJk+eHLfccks8+OCDY/ctAADgQ5T8HN/zwXN8AQByOe/P8QUAgAuV8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJCC8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJCC8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJCC8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKYwqfNetWxf19fVRVVUVDQ0NsW3btjNa9+qrr0Z5eXl8/vOfH83HAgDAqJUcvps3b47ly5fH6tWro7OzMxYuXBiLFi2Krq6u067r6+uLpUuXxl//9V+PerMAADBaZUVRFKUsmDdvXsyZMyfWr18/ODZr1qxYvHhxtLW1nXLdV7/61bjqqqti4sSJ8cILL8Tu3bvP+DP7+/ujpqYm+vr6orq6upTtAgBwARqP/ivpiu/Ro0dj165d0dzcPGS8ubk5tm/ffsp1jz/+eLz++uvxwAMPnNHnDAwMRH9//5AXAACcjZLCt7e3N44fPx61tbVDxmtra6Onp2fENb/97W9j5cqVsWnTpigvLz+jz2lra4uamprB14wZM0rZJgAADDOqX24rKysb8nNRFMPGIiKOHz8et956a6xZsyauvvrqM37/VatWRV9f3+DrwIEDo9kmAAAMOrNLsP/XlClTYuLEicOu7h46dGjYVeCIiCNHjsTOnTujs7MzvvWtb0VExIkTJ6IoiigvL48tW7bEtddeO2xdZWVlVFZWlrI1AAA4rZKu+FZUVERDQ0N0dHQMGe/o6IimpqZh86urq+PXv/517N69e/DV0tISn/rUp2L37t0xb968s9s9AACcoZKu+EZEtLa2xm233RaNjY0xf/78+OlPfxpdXV3R0tISER/cpvC73/0ufv7zn8eECRNi9uzZQ9ZfdtllUVVVNWwcAADGU8nhu2TJkjh8+HCsXbs2uru7Y/bs2dHe3h51dXUREdHd3f2hz/QFAIBzreTn+J4PnuMLAJDLeX+OLwAAXKiELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJCC8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJCC8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJCC8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIYVThu27duqivr4+qqqpoaGiIbdu2nXLuc889F9dff318/OMfj+rq6pg/f3784he/GPWGAQBgNEoO382bN8fy5ctj9erV0dnZGQsXLoxFixZFV1fXiPNffvnluP7666O9vT127doVX/rSl+Lmm2+Ozs7Os948AACcqbKiKIpSFsybNy/mzJkT69evHxybNWtWLF68ONra2s7oPT772c/GkiVL4v777z+j+f39/VFTUxN9fX1RXV1dynYBALgAjUf/lXTF9+jRo7Fr165obm4eMt7c3Bzbt28/o/c4ceJEHDlyJC699NJTzhkYGIj+/v4hLwAAOBslhW9vb28cP348amtrh4zX1tZGT0/PGb3HD37wg3j33XfjlltuOeWctra2qKmpGXzNmDGjlG0CAMAwo/rltrKysiE/F0UxbGwkTz31VHz3u9+NzZs3x2WXXXbKeatWrYq+vr7B14EDB0azTQAAGFReyuQpU6bExIkTh13dPXTo0LCrwH9s8+bNcdddd8XTTz8d11133WnnVlZWRmVlZSlbAwCA0yrpim9FRUU0NDRER0fHkPGOjo5oamo65bqnnnoq7rjjjnjyySfjpptuGt1OAQDgLJR0xTciorW1NW677bZobGyM+fPnx09/+tPo6uqKlpaWiPjgNoXf/e538fOf/zwiPojepUuXxg9/+MP4whe+MHi1+KKLLoqampox/CoAAHBqJYfvkiVL4vDhw7F27dro7u6O2bNnR3t7e9TV1UVERHd395Bn+v7kJz+JY8eOxTe/+c345je/OTh+++23xxNPPHH23wAAAM5Ayc/xPR88xxcAIJfz/hxfAAC4UAlfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJCC8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJCC8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJCC8AUAIAXhCwBACsIXAIAUhC8AACkIXwAAUhC+AACkIHwBAEhB+AIAkILwBQAgBeELAEAKwhcAgBSELwAAKQhfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBIQfgCAJDCqMJ33bp1UV9fH1VVVdHQ0BDbtm077fytW7dGQ0NDVFVVxcyZM+PRRx8d1WYBAGC0Sg7fzZs3x/Lly2P16tXR2dkZCxcujEWLFkVXV9eI8/fv3x833nhjLFy4MDo7O+M73/lOLFu2LJ599tmz3jwAAJypsqIoilIWzJs3L+bMmRPr168fHJs1a1YsXrw42trahs3/9re/HS+++GLs3bt3cKylpSV+9atfxY4dO87oM/v7+6Ompib6+vqiurq6lO0CAHABGo/+Ky9l8tGjR2PXrl2xcuXKIePNzc2xffv2Edfs2LEjmpubh4zdcMMNsWHDhnj//fdj0qRJw9YMDAzEwMDA4M99fX0R8cF/AAAA/Ok72X0lXqM9rZLCt7e3N44fPx61tbVDxmtra6Onp2fENT09PSPOP3bsWPT29sbUqVOHrWlra4s1a9YMG58xY0Yp2wUA4AJ3+PDhqKmpGZP3Kil8TyorKxvyc1EUw8Y+bP5I4yetWrUqWltbB39+++23o66uLrq6usbsi3Ph6O/vjxkzZsSBAwfc6pKQ88/N+efm/HPr6+uLK6+8Mi699NIxe8+SwnfKlCkxceLEYVd3Dx06NOyq7kmXX375iPPLy8tj8uTJI66prKyMysrKYeM1NTX+i59YdXW180/M+efm/HNz/rlNmDB2T98t6Z0qKiqioaEhOjo6hox3dHREU1PTiGvmz58/bP6WLVuisbFxxPt7AQBgPJSc0K2trfHYY4/Fxo0bY+/evbFixYro6uqKlpaWiPjgNoWlS5cOzm9paYk33ngjWltbY+/evbFx48bYsGFD3HvvvWP3LQAA4EOUfI/vkiVL4vDhw7F27dro7u6O2bNnR3t7e9TV1UVERHd395Bn+tbX10d7e3usWLEiHnnkkZg2bVo8/PDD8ZWvfOWMP7OysjIeeOCBEW9/4E+f88/N+efm/HNz/rmNx/mX/BxfAAC4EI3d3cIAAPARJnwBAEhB+AIAkILwBQAghY9M+K5bty7q6+ujqqoqGhoaYtu2baedv3Xr1mhoaIiqqqqYOXNmPProo+dop4yHUs7/ueeei+uvvz4+/vGPR3V1dcyfPz9+8YtfnMPdMtZK/fN/0quvvhrl5eXx+c9/fnw3yLgq9fwHBgZi9erVUVdXF5WVlfHJT34yNm7ceI52y1gr9fw3bdoU11xzTVx88cUxderUuPPOO+Pw4cPnaLeMlZdffjluvvnmmDZtWpSVlcULL7zwoWvGpP2Kj4B/+Zd/KSZNmlT87Gc/K/bs2VPcc889xSWXXFK88cYbI87ft29fcfHFFxf33HNPsWfPnuJnP/tZMWnSpOKZZ545xztnLJR6/vfcc0/xve99r/jP//zP4rXXXitWrVpVTJo0qfjv//7vc7xzxkKp53/S22+/XcycObNobm4urrnmmnOzWcbcaM7/y1/+cjFv3ryio6Oj2L9/f/Ef//EfxauvvnoOd81YKfX8t23bVkyYMKH44Q9/WOzbt6/Ytm1b8dnPfrZYvHjxOd45Z6u9vb1YvXp18eyzzxYRUTz//POnnT9W7feRCN+5c+cWLS0tQ8Y+/elPFytXrhxx/j/8wz8Un/70p4eMff3rXy++8IUvjNseGT+lnv9IPvOZzxRr1qwZ661xDoz2/JcsWVL84z/+Y/HAAw8I3wtYqef/r//6r0VNTU1x+PDhc7E9xlmp5/9P//RPxcyZM4eMPfzww8X06dPHbY+MvzMJ37Fqv/N+q8PRo0dj165d0dzcPGS8ubk5tm/fPuKaHTt2DJt/ww03xM6dO+P9998ft70y9kZz/n/sxIkTceTIkbj00kvHY4uMo9Ge/+OPPx6vv/56PPDAA+O9RcbRaM7/xRdfjMbGxvj+978fV1xxRVx99dVx7733xh/+8IdzsWXG0GjOv6mpKQ4ePBjt7e1RFEW8+eab8cwzz8RNN910LrbMeTRW7Vfyv7ltrPX29sbx48ejtrZ2yHhtbW309PSMuKanp2fE+ceOHYve3t6YOnXquO2XsTWa8/9jP/jBD+Ldd9+NW265ZTy2yDgazfn/9re/jZUrV8a2bduivPy8/y2MszCa89+3b1+88sorUVVVFc8//3z09vbGN77xjXjrrbfc53uBGc35NzU1xaZNm2LJkiXxv//7v3Hs2LH48pe/HD/60Y/OxZY5j8aq/c77Fd+TysrKhvxcFMWwsQ+bP9I4F4ZSz/+kp556Kr773e/G5s2b47LLLhuv7THOzvT8jx8/HrfeemusWbMmrr766nO1PcZZKX/+T5w4EWVlZbFp06aYO3du3HjjjfHQQw/FE0884arvBaqU89+zZ08sW7Ys7r///ti1a1e89NJLsX///mhpaTkXW+U8G4v2O++XS6ZMmRITJ04c9v/uDh06NKzsT7r88stHnF9eXh6TJ08et70y9kZz/idt3rw57rrrrnj66afjuuuuG89tMk5KPf8jR47Ezp07o7OzM771rW9FxAchVBRFlJeXx5YtW+Laa689J3vn7I3mz//UqVPjiiuuiJqamsGxWbNmRVEUcfDgwbjqqqvGdc+MndGcf1tbWyxYsCDuu+++iIj43Oc+F5dcckksXLgwHnzwQf/E90/YWLXfeb/iW1FREQ0NDdHR0TFkvKOjI5qamkZcM3/+/GHzt2zZEo2NjTFp0qRx2ytjbzTnH/HBld477rgjnnzySfd2XcBKPf/q6ur49a9/Hbt37x58tbS0xKc+9anYvXt3zJs371xtnTEwmj//CxYsiN///vfxzjvvDI699tprMWHChJg+ffq47pexNZrzf++992LChKHpMnHixIj4f1f/+NM0Zu1X0q/CjZOTjzPZsGFDsWfPnmL58uXFJZdcUvzP//xPURRFsXLlyuK2224bnH/ykRYrVqwo9uzZU2zYsMHjzC5gpZ7/k08+WZSXlxePPPJI0d3dPfh6++23z9dX4CyUev5/zFMdLmylnv+RI0eK6dOnF3/zN39T/OY3vym2bt1aXHXVVcXdd999vr4CZ6HU83/88ceL8vLyYt26dcXrr79evPLKK0VjY2Mxd+7c8/UVGKUjR44UnZ2dRWdnZxERxUMPPVR0dnYOPspuvNrvIxG+RVEUjzzySFFXV1dUVFQUc+bMKbZu3Tr4126//fbii1/84pD5//7v/1785V/+ZVFRUVF84hOfKNavX3+Od8xYKuX8v/jFLxYRMex1++23n/uNMyZK/fP//xO+F75Sz3/v3r3FddddV1x00UXF9OnTi9bW1uK99947x7tmrJR6/g8//HDxmc98prjooouKqVOnFl/72teKgwcPnuNdc7b+7d/+7bT/Wz5e7VdWFP7ZAAAAf/rO+z2+AABwLghfAABSEL4AAKQgfAEASEH4AgCQgvAFACAF4QsAQArCFwCAFIQvAAApCF8AAFIQvgAApCB8AQBI4f8AaCkHRuWfkuQAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#visualize G\n",
    "import matplotlib.pyplot as plt\n",
    "nx.draw(G, with_labels=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'numpy.ndarray' object has no attribute 'index'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/lib/python3.11/site-packages/networkx/convert_matrix.py:199\u001b[0m, in \u001b[0;36mfrom_pandas_adjacency\u001b[0;34m(df, create_using)\u001b[0m\n\u001b[1;32m    198\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 199\u001b[0m     df \u001b[38;5;241m=\u001b[39m df[\u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m]\n\u001b[1;32m    200\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'numpy.ndarray' object has no attribute 'index'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[14], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m graph \u001b[38;5;241m=\u001b[39m \u001b[43mnx\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_pandas_adjacency\u001b[49m\u001b[43m(\u001b[49m\u001b[43madj_matrix\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m graph\u001b[38;5;241m.\u001b[39mremove_edges_from(nx\u001b[38;5;241m.\u001b[39mselfloop_edges(graph))\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.11/site-packages/networkx/convert_matrix.py:201\u001b[0m, in \u001b[0;36mfrom_pandas_adjacency\u001b[0;34m(df, create_using)\u001b[0m\n\u001b[1;32m    199\u001b[0m     df \u001b[38;5;241m=\u001b[39m df[df\u001b[38;5;241m.\u001b[39mindex]\n\u001b[1;32m    200\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m--> 201\u001b[0m     missing \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(\u001b[38;5;28mset\u001b[39m(\u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m)\u001b[38;5;241m.\u001b[39mdifference(\u001b[38;5;28mset\u001b[39m(df\u001b[38;5;241m.\u001b[39mcolumns)))\n\u001b[1;32m    202\u001b[0m     msg \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmissing\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not in columns\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    203\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m nx\u001b[38;5;241m.\u001b[39mNetworkXError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mColumns must match Indices.\u001b[39m\u001b[38;5;124m\"\u001b[39m, msg) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'numpy.ndarray' object has no attribute 'index'"
     ]
    }
   ],
   "source": [
    "# graph = nx.from_pandas_adjacency(adj_matrix)\n",
    "graph.remove_edges_from(nx.selfloop_edges(graph))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['thrS', 'rplC', 'rpsM', 'rplQ', 'rpmA', 'rpsI', 'rpsL', 'rpmH', 'rpmF',\n",
       "       'rpsJ',\n",
       "       ...\n",
       "       'group_575', 'group_5776', 'smc_2', 'aaaT', 'group_660', 'group_664',\n",
       "       'group_679', 'group_775', 'group_796', 'group_999'],\n",
       "      dtype='object', name='Gene', length=1354)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>thrS</th>\n",
       "      <th>rplC</th>\n",
       "      <th>rpsM</th>\n",
       "      <th>rplQ</th>\n",
       "      <th>rpmA</th>\n",
       "      <th>rpsI</th>\n",
       "      <th>rpsL</th>\n",
       "      <th>rpmH</th>\n",
       "      <th>rpmF</th>\n",
       "      <th>rpsJ</th>\n",
       "      <th>...</th>\n",
       "      <th>group_575</th>\n",
       "      <th>group_5776</th>\n",
       "      <th>smc_2</th>\n",
       "      <th>aaaT</th>\n",
       "      <th>group_660</th>\n",
       "      <th>group_664</th>\n",
       "      <th>group_679</th>\n",
       "      <th>group_775</th>\n",
       "      <th>group_796</th>\n",
       "      <th>group_999</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>thrS</th>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>...</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rplC</th>\n",
       "      <td>31.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>...</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpsM</th>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>...</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rplQ</th>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>...</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpmA</th>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>...</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_664</th>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_679</th>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_775</th>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_796</th>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>group_999</th>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>...</td>\n",
       "      <td>13.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1354 rows × 1354 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           thrS  rplC  rpsM  rplQ  rpmA  rpsI  rpsL  rpmH  rpmF  rpsJ  ...  \\\n",
       "thrS        0.0  31.0  31.0  31.0  31.0  31.0  31.0  31.0  31.0  31.0  ...   \n",
       "rplC       31.0   0.0  31.0  31.0  31.0  31.0  31.0  31.0  31.0  31.0  ...   \n",
       "rpsM       31.0  31.0   0.0  31.0  31.0  31.0  31.0  31.0  31.0  31.0  ...   \n",
       "rplQ       31.0  31.0  31.0   0.0  31.0  31.0  31.0  31.0  31.0  31.0  ...   \n",
       "rpmA       31.0  31.0  31.0  31.0   0.0  31.0  31.0  31.0  31.0  31.0  ...   \n",
       "...         ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   \n",
       "group_664  16.0  16.0  16.0  16.0  16.0  16.0  16.0  16.0  16.0  16.0  ...   \n",
       "group_679  16.0  16.0  16.0  16.0  16.0  16.0  16.0  16.0  16.0  16.0  ...   \n",
       "group_775  16.0  16.0  16.0  16.0  16.0  16.0  16.0  16.0  16.0  16.0  ...   \n",
       "group_796  16.0  16.0  16.0  16.0  16.0  16.0  16.0  16.0  16.0  16.0  ...   \n",
       "group_999  16.0  16.0  16.0  16.0  16.0  16.0  16.0  16.0  16.0  16.0  ...   \n",
       "\n",
       "           group_575  group_5776  smc_2  aaaT  group_660  group_664  \\\n",
       "thrS            16.0        16.0   16.0  16.0       16.0       16.0   \n",
       "rplC            16.0        16.0   16.0  16.0       16.0       16.0   \n",
       "rpsM            16.0        16.0   16.0  16.0       16.0       16.0   \n",
       "rplQ            16.0        16.0   16.0  16.0       16.0       16.0   \n",
       "rpmA            16.0        16.0   16.0  16.0       16.0       16.0   \n",
       "...              ...         ...    ...   ...        ...        ...   \n",
       "group_664       11.0        10.0   10.0   9.0       10.0        0.0   \n",
       "group_679       14.0        14.0   11.0  11.0       13.0       12.0   \n",
       "group_775       14.0        14.0   12.0  11.0       13.0       12.0   \n",
       "group_796       11.0        11.0    9.0   8.0       10.0       12.0   \n",
       "group_999       13.0        12.0   11.0  11.0       13.0       12.0   \n",
       "\n",
       "           group_679  group_775  group_796  group_999  \n",
       "thrS            16.0       16.0       16.0       16.0  \n",
       "rplC            16.0       16.0       16.0       16.0  \n",
       "rpsM            16.0       16.0       16.0       16.0  \n",
       "rplQ            16.0       16.0       16.0       16.0  \n",
       "rpmA            16.0       16.0       16.0       16.0  \n",
       "...              ...        ...        ...        ...  \n",
       "group_664       12.0       12.0       12.0       12.0  \n",
       "group_679        0.0       14.0       11.0       13.0  \n",
       "group_775       14.0        0.0       10.0       13.0  \n",
       "group_796       11.0       10.0        0.0        9.0  \n",
       "group_999       13.0       13.0        9.0        0.0  \n",
       "\n",
       "[1354 rows x 1354 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#show G as an adjacency matrix\n",
    "nx.to_pandas_adjacency(G)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/rayane/capstone-project/src'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "create_gml = lambda G, file_path: nx.write_gml(G, file_path)\n",
    "create_gml(G, '../data/graphs/1.1_gene-gene_presence_0.5_strepto_31.gml') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30148.437223042834"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean(list(dict(G.degree(weight='weight')).values()))\n",
    "# np.mean(list(dict(G.degree()).values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Network': 'gene presence network',\n",
       " '|N|': 1354,\n",
       " '|V|': 915981,\n",
       " 'Density': 1.0,\n",
       " '<k>': 1353.0,\n",
       " '<k> weighted': 30148.437223042834,\n",
       " '<cc>': 1.0,\n",
       " '<spath>': 1.0,\n",
       " 'd': 1}"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import networkx as nx\n",
    "   \n",
    "def compute_network_stats(G, network_name):\n",
    "    largest_cc = max(nx.connected_components(G), key=len)\n",
    "    largest_cc_subgraph = G.subgraph(largest_cc)\n",
    "    stats = {\n",
    "        \"Network\": network_name,\n",
    "        \"|N|\": G.number_of_nodes(),\n",
    "        \"|V|\": G.number_of_edges(),\n",
    "        \"Density\": nx.density(G),\n",
    "        \"<k>\": np.mean(list(dict(G.degree()).values())),\n",
    "        \"<k> weighted\": np.mean(list(dict(G.degree(weight='weight')).values())),\n",
    "        \"<cc>\": nx.average_clustering(G),\n",
    "        \"<spath>\": nx.average_shortest_path_length(largest_cc_subgraph),\n",
    "        \"d\": nx.diameter(largest_cc_subgraph)\n",
    "    }\n",
    "    \n",
    "    return stats\n",
    "\n",
    "gene_presence_50 = compute_network_stats(G, 'gene presence network')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "G=nx.read_gml('../data/graphs/1.1_gene-gene_presence_0.5_strepto_31.gml')\n",
    "g=nx.read_gml('../data/graphs/1.1_test.gml')\n",
    "g==G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "test=compute_network_stats(g, 'test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Network': 'test',\n",
       " '|N|': 1354,\n",
       " '|V|': 915981,\n",
       " 'Density': 1.0,\n",
       " '<k>': 1353.0,\n",
       " '<k> weighted': 30148.437223042834,\n",
       " '<cc>': 1.0,\n",
       " '<spath>': 1.0,\n",
       " 'd': 1}"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
